{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Seq import MutableSeq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "from Bio import motifs\n",
    "from papa_helpers import get_terminal_regions\n",
    "import os\n",
    "import sys\n",
    "import random"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reference transcripts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>3400</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_ref_gene_1</td>\n",
       "      <td>ref_g1_tr_1</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_1</td>\n",
       "      <td>ref_g1_tr_1</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_1</td>\n",
       "      <td>ref_g1_tr_1</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2000</td>\n",
       "      <td>2200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_1</td>\n",
       "      <td>ref_g1_tr_1</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3000</td>\n",
       "      <td>3400</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_1</td>\n",
       "      <td>ref_g1_tr_1</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+---------------+-----------------+-------------+---------------+----------------+\n",
       "| Chromosome   |     Start |       End | Strand       | Feature    | gene_id       | transcript_id   | gene_name   |   exon_number | gene_type      |\n",
       "| (category)   |   (int32) |   (int32) | (category)   | (object)   | (object)      | (object)        | (object)    |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+---------------+-----------------+-------------+---------------+----------------|\n",
       "| chr1         |       100 |      3400 | +            | transcript | id_ref_gene_1 | ref_g1_tr_1     | ref_gene_1  |           nan | protein_coding |\n",
       "| chr1         |       100 |       300 | +            | exon       | id_ref_gene_1 | ref_g1_tr_1     | ref_gene_1  |             1 | protein_coding |\n",
       "| chr1         |      1000 |      1200 | +            | exon       | id_ref_gene_1 | ref_g1_tr_1     | ref_gene_1  |             2 | protein_coding |\n",
       "| chr1         |      2000 |      2200 | +            | exon       | id_ref_gene_1 | ref_g1_tr_1     | ref_gene_1  |             3 | protein_coding |\n",
       "| chr1         |      3000 |      3400 | +            | exon       | id_ref_gene_1 | ref_g1_tr_1     | ref_gene_1  |             4 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+---------------+-----------------+-------------+---------------+----------------+\n",
       "Stranded PyRanges object has 5 rows and 10 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A plus strand transcript\n",
    "test_ref_tr1 = {\"Chromosome\": [\"chr1\"]*5,\n",
    "                \"Start\": [100,100,1000,2000,3000],\n",
    "                \"End\": [3400,300,1200,2200,3400],\n",
    "                \"Strand\": [\"+\"]*5,\n",
    "                \"Feature\": [\"transcript\"] +[\"exon\"]*4,\n",
    "                \"gene_id\": [\"id_ref_gene_1\"]*5,\n",
    "                \"transcript_id\": [\"ref_g1_tr_1\"] * 5,\n",
    "                \"gene_name\": [\"ref_gene_1\"]*5,\n",
    "                \"exon_number\": [None,1,2,3,4],\n",
    "                \"gene_type\": [\"protein_coding\"] * 5}\n",
    "\n",
    "\n",
    "pr.from_dict(test_ref_tr1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>3400</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_ref_gene_1</td>\n",
       "      <td>ref_g1_tr_1</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_1</td>\n",
       "      <td>ref_g1_tr_1</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_1</td>\n",
       "      <td>ref_g1_tr_1</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2000</td>\n",
       "      <td>2200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_1</td>\n",
       "      <td>ref_g1_tr_1</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3000</td>\n",
       "      <td>3400</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_1</td>\n",
       "      <td>ref_g1_tr_1</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr2</td>\n",
       "      <td>100</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_ref_gene_2</td>\n",
       "      <td>ref_g2_tr_1</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr2</td>\n",
       "      <td>100</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_2</td>\n",
       "      <td>ref_g2_tr_1</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr2</td>\n",
       "      <td>700</td>\n",
       "      <td>800</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_2</td>\n",
       "      <td>ref_g2_tr_1</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr2</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_2</td>\n",
       "      <td>ref_g2_tr_1</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+---------------+-----------------+-------------+---------------+----------------+\n",
       "| Chromosome   | Start     | End       | Strand       | Feature    | gene_id       | transcript_id   | gene_name   | exon_number   | gene_type      |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | (object)   | (object)      | (object)        | (object)    | (float64)     | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+---------------+-----------------+-------------+---------------+----------------|\n",
       "| chr1         | 100       | 3400      | +            | transcript | id_ref_gene_1 | ref_g1_tr_1     | ref_gene_1  | nan           | protein_coding |\n",
       "| chr1         | 100       | 300       | +            | exon       | id_ref_gene_1 | ref_g1_tr_1     | ref_gene_1  | 1.0           | protein_coding |\n",
       "| chr1         | 1000      | 1200      | +            | exon       | id_ref_gene_1 | ref_g1_tr_1     | ref_gene_1  | 2.0           | protein_coding |\n",
       "| chr1         | 2000      | 2200      | +            | exon       | id_ref_gene_1 | ref_g1_tr_1     | ref_gene_1  | 3.0           | protein_coding |\n",
       "| ...          | ...       | ...       | ...          | ...        | ...           | ...             | ...         | ...           | ...            |\n",
       "| chr2         | 100       | 1200      | -            | transcript | id_ref_gene_2 | ref_g2_tr_1     | ref_gene_2  | nan           | protein_coding |\n",
       "| chr2         | 100       | 200       | -            | exon       | id_ref_gene_2 | ref_g2_tr_1     | ref_gene_2  | 3.0           | protein_coding |\n",
       "| chr2         | 700       | 800       | -            | exon       | id_ref_gene_2 | ref_g2_tr_1     | ref_gene_2  | 2.0           | protein_coding |\n",
       "| chr2         | 1000      | 1200      | -            | exon       | id_ref_gene_2 | ref_g2_tr_1     | ref_gene_2  | 1.0           | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+---------------+-----------------+-------------+---------------+----------------+\n",
       "Stranded PyRanges object has 9 rows and 10 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A minus strand transcript\n",
    "test_ref_tr2 = {\"Chromosome\": [\"chr2\"]*4,\n",
    "                \"Start\": [100,100,700,1000],\n",
    "                \"End\": [1200,200,800,1200],\n",
    "                \"Strand\": [\"-\"]*4,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*3,\n",
    "                \"gene_id\": [\"id_ref_gene_2\"]*4,\n",
    "                \"transcript_id\": [\"ref_g2_tr_1\"]*4,\n",
    "                \"gene_name\": [\"ref_gene_2\"]*4,\n",
    "                \"exon_number\": [None,3,2,1],\n",
    "                \"gene_type\": [\"protein_coding\"] * 4\n",
    "               }\n",
    "\n",
    "\n",
    "test_ref = pr.concat([pr.from_dict(test_ref_tr1), pr.from_dict(test_ref_tr2)])\n",
    "test_ref"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test novel transcripts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Chromosome': [], 'Start': [], 'End': [], 'Strand': []}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Collecting 3'ends for a dummy PolyASite atlas\n",
    "pas_ends = {\"Chromosome\": [],\n",
    "           \"Start\": [],\n",
    "           \"End\": [],\n",
    "           \"Strand\": []}\n",
    "\n",
    "# Collect test novel transcripts\n",
    "test_novel_trs = []\n",
    "\n",
    "\n",
    "pas_ends"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_pas_ends(pas_dict, chrom, start, end, strand):\n",
    "    '''\n",
    "    '''\n",
    "    \n",
    "    pas_d = pas_dict.copy()\n",
    "    \n",
    "    pas_d[\"Chromosome\"].append(chrom)\n",
    "    pas_d[\"Start\"].append(start)\n",
    "    pas_d[\"End\"].append(end)\n",
    "    pas_d[\"Strand\"].append(strand)\n",
    "    \n",
    "    return pas_d\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>500</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>350</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>500</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+---------------+--------------------+-------------+---------------+----------------+\n",
       "| Chromosome   |     Start |       End | Strand       | Feature    | gene_id       | transcript_id      | gene_name   |   exon_number | gene_type      |\n",
       "| (category)   |   (int32) |   (int32) | (category)   | (object)   | (object)      | (object)           | (object)    |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+---------------+--------------------+-------------+---------------+----------------|\n",
       "| chr1         |       100 |       700 | +            | transcript | id_nov_gene_1 | nov_gene_1_tx_fi_p | nov_gene_1  |           nan | protein_coding |\n",
       "| chr1         |       100 |       300 | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_fi_p | nov_gene_1  |             1 | protein_coding |\n",
       "| chr1         |       500 |       700 | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_fi_p | nov_gene_1  |             2 | protein_coding |\n",
       "| chr1         |       100 |       700 | +            | transcript | id_nov_gene_1 | nov_gene_1_tx_fi_f | nov_gene_1  |           nan | protein_coding |\n",
       "| chr1         |       100 |       350 | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_fi_f | nov_gene_1  |             1 | protein_coding |\n",
       "| chr1         |       500 |       700 | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_fi_f | nov_gene_1  |             2 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+---------------+--------------------+-------------+---------------+----------------+\n",
       "Stranded PyRanges object has 6 rows and 10 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now make test novel transcripts to cover my test cases\n",
    "\n",
    "# 1. Novel last exons in first intron of annotated transcript (e.g. STMN2)\n",
    "# For these to pass, they should share an identical 3'end with a first exon of a known transcript\n",
    "\n",
    "test_novel_fi = {\"Chromosome\": [\"chr1\"]*6,\n",
    "                \"Start\": [100] + [100,500] + [100] + [100,500],\n",
    "                \"End\": [700,300,700,700,350,700],\n",
    "                \"Strand\": [\"+\"]*6,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*2 + [\"transcript\"] + [\"exon\"]*2,\n",
    "                \"gene_id\": [\"id_nov_gene_1\"]*6,\n",
    "                \"transcript_id\": [\"nov_gene_1_tx_fi_p\"]*3 + [\"nov_gene_1_tx_fi_f\"]*3,\n",
    "                 \"gene_name\": [\"nov_gene_1\"]*6,\n",
    "                \"exon_number\": [None,1,2,None,1,2],\n",
    "                 \"gene_type\": [\"protein_coding\"] * 6\n",
    "               }\n",
    "\n",
    "test_novel_trs.append(pr.from_dict(test_novel_fi))\n",
    "update_pas_ends(pas_ends, \"chr1\", 699, 700, \"+\")\n",
    "pr.from_dict(test_novel_fi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Chromosome': ['chr1'], 'Start': [699], 'End': [700], 'Strand': ['+']}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "pas_ends"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>880</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_fi_ext_p</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>880</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_fi_ext_p</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>880</td>\n",
       "      <td>1500</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_fi_ext_f</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>880</td>\n",
       "      <td>1500</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_fi_ext_f</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+---------------+------------------------+-------------+---------------+-------+\n",
       "| Chromosome   |     Start |       End | Strand       | Feature    | gene_id       | transcript_id          | gene_name   |   exon_number | +1    |\n",
       "| (category)   |   (int32) |   (int32) | (category)   | (object)   | (object)      | (object)               | (object)    |     (float64) | ...   |\n",
       "|--------------+-----------+-----------+--------------+------------+---------------+------------------------+-------------+---------------+-------|\n",
       "| chr2         |       880 |      1200 | -            | transcript | id_nov_gene_2 | nov_gene_2_tx_fi_ext_p | nov_gene_2  |           nan | ...   |\n",
       "| chr2         |       880 |      1200 | -            | exon       | id_nov_gene_2 | nov_gene_2_tx_fi_ext_p | nov_gene_2  |             1 | ...   |\n",
       "| chr2         |       880 |      1500 | -            | transcript | id_nov_gene_2 | nov_gene_2_tx_fi_ext_f | nov_gene_2  |           nan | ...   |\n",
       "| chr2         |       880 |      1500 | -            | exon       | id_nov_gene_2 | nov_gene_2_tx_fi_ext_f | nov_gene_2  |             1 | ...   |\n",
       "+--------------+-----------+-----------+--------------+------------+---------------+------------------------+-------------+---------------+-------+\n",
       "Stranded PyRanges object has 4 rows and 10 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: gene_type"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#1b. First intron extensions\n",
    "# 1 should fall within x bp of the 5'end of the annotated first exon (to account for diffs in annotation)\n",
    "# To fail should fall outside x bp (e.g. 500bp away) - easier to do this on minus strand as starts at 1200 (1999)\n",
    "\n",
    "test_novel_fi_ext = {\"Chromosome\": [\"chr2\"]*4,\n",
    "                     \"Start\": [880,880] + [880,880],\n",
    "                \"End\": [1200,1200] + [1500,1500],\n",
    "                \"Strand\": [\"-\"]*4,\n",
    "                \"Feature\": [\"transcript\", \"exon\"]*2,\n",
    "                \"gene_id\": [\"id_nov_gene_2\"]*4,\n",
    "                \"transcript_id\": [\"nov_gene_2_tx_fi_ext_p\"]*2 + [\"nov_gene_2_tx_fi_ext_f\"]*2,\n",
    "                     \"gene_name\": [\"nov_gene_2\"]*4,\n",
    "                     \"exon_number\": [None,1,None,1],\n",
    "                     \"gene_type\": [\"protein_coding\"] * 4\n",
    "                    }\n",
    "\n",
    "test_novel_trs.append(pr.from_dict(test_novel_fi_ext))\n",
    "update_pas_ends(pas_ends, \"chr2\", 880, 881, \"-\")\n",
    "pr.from_dict(test_novel_fi_ext)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Chromosome': ['chr1', 'chr2'],\n",
       " 'Start': [699, 880],\n",
       " 'End': [700, 881],\n",
       " 'Strand': ['+', '-']}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pas_ends"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>1600</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1400</td>\n",
       "      <td>1600</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>1600</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1400</td>\n",
       "      <td>1600</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>1850</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_no_pas_no_motif_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_no_pas_no_motif_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_no_pas_no_motif_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1750</td>\n",
       "      <td>1850</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_no_pas_no_motif_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+---------------+--------------------------------------+-------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | Feature    | gene_id       | transcript_id                        | gene_name   | +2    |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | (object)   | (object)      | (object)                             | (object)    | ...   |\n",
       "|--------------+-----------+-----------+--------------+------------+---------------+--------------------------------------+-------------+-------|\n",
       "| chr1         | 100       | 1600      | +            | transcript | id_nov_gene_1 | nov_gene_1_tx_mi_s_p                 | nov_gene_1  | ...   |\n",
       "| chr1         | 100       | 300       | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_mi_s_p                 | nov_gene_1  | ...   |\n",
       "| chr1         | 1000      | 1200      | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_mi_s_p                 | nov_gene_1  | ...   |\n",
       "| chr1         | 1400      | 1600      | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_mi_s_p                 | nov_gene_1  | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...        | ...           | ...                                  | ...         | ...   |\n",
       "| chr1         | 100       | 1850      | +            | transcript | id_nov_gene_1 | nov_gene_1_tx_mi_s_no_pas_no_motif_f | nov_gene_1  | ...   |\n",
       "| chr1         | 100       | 300       | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_mi_s_no_pas_no_motif_f | nov_gene_1  | ...   |\n",
       "| chr1         | 1000      | 1200      | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_mi_s_no_pas_no_motif_f | nov_gene_1  | ...   |\n",
       "| chr1         | 1750      | 1850      | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_mi_s_no_pas_no_motif_f | nov_gene_1  | ...   |\n",
       "+--------------+-----------+-----------+--------------+------------+---------------+--------------------------------------+-------------+-------+\n",
       "Stranded PyRanges object has 12 rows and 10 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: exon_number, gene_type"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2. Internal intron, spliced in last exon (e.g. ONECUT1)\n",
    "# For these to pass, the 5'end of the novel last exon must exactly match a reference intron\n",
    "# Also include an event that passes the 5'end filter but won't have a nearby annotated pas\n",
    "\n",
    "test_novel_mi_s = {\"Chromosome\": [\"chr1\"]*12,\n",
    "                \"Start\": [100,100,1000,1400] + [100,100,1000,1400] + [100, 100, 1000, 1750],\n",
    "                \"End\": [1600,300,1200,1600] + [1600,300,1300,1600] + [1850, 300, 1200, 1850],\n",
    "                \"Strand\": [\"+\"]*12,\n",
    "                \"Feature\": ([\"transcript\"] + [\"exon\"]*3)*3,\n",
    "                \"gene_id\": [\"id_nov_gene_1\"]*12,\n",
    "                \"transcript_id\": [\"nov_gene_1_tx_mi_s_p\"]*4 + [\"nov_gene_1_tx_mi_s_f\"]*4 + [\"nov_gene_1_tx_mi_s_no_pas_no_motif_f\"]*4,\n",
    "                 \"gene_name\": [\"nov_gene_1\"] * 12,\n",
    "                \"exon_number\": [None,1,2,3,None,1,2,3, None,1,2,3],\n",
    "                 \"gene_type\": [\"protein_coding\"] * 12\n",
    "               }\n",
    "test_novel_trs.append(pr.from_dict(test_novel_mi_s))\n",
    "update_pas_ends(pas_ends, \"chr1\", 1599, 1600, \"+\")\n",
    "pr.from_dict(test_novel_mi_s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>2400</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2000</td>\n",
       "      <td>2400</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>2400</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1950</td>\n",
       "      <td>2400</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>2600</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_no_pas_motif_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_no_pas_motif_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_no_pas_motif_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2000</td>\n",
       "      <td>2600</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_no_pas_motif_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+---------------+-------------------------------------+-------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | Feature    | gene_id       | transcript_id                       | gene_name   | +2    |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | (object)   | (object)      | (object)                            | (object)    | ...   |\n",
       "|--------------+-----------+-----------+--------------+------------+---------------+-------------------------------------+-------------+-------|\n",
       "| chr1         | 100       | 2400      | +            | transcript | id_nov_gene_1 | nov_gene_1_tx_mi_ext_p              | nov_gene_1  | ...   |\n",
       "| chr1         | 100       | 300       | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_mi_ext_p              | nov_gene_1  | ...   |\n",
       "| chr1         | 1000      | 1200      | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_mi_ext_p              | nov_gene_1  | ...   |\n",
       "| chr1         | 2000      | 2400      | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_mi_ext_p              | nov_gene_1  | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...        | ...           | ...                                 | ...         | ...   |\n",
       "| chr1         | 100       | 2600      | +            | transcript | id_nov_gene_1 | nov_gene_1_tx_mi_ext_no_pas_motif_p | nov_gene_1  | ...   |\n",
       "| chr1         | 100       | 300       | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_mi_ext_no_pas_motif_p | nov_gene_1  | ...   |\n",
       "| chr1         | 1000      | 1200      | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_mi_ext_no_pas_motif_p | nov_gene_1  | ...   |\n",
       "| chr1         | 2000      | 2600      | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_mi_ext_no_pas_motif_p | nov_gene_1  | ...   |\n",
       "+--------------+-----------+-----------+--------------+------------+---------------+-------------------------------------+-------------+-------+\n",
       "Stranded PyRanges object has 12 rows and 10 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: exon_number, gene_type"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3. Internal intron bleedthrough (e.g. SIN3B)\n",
    "# For these events to pass, the 5'end of the last exon matches the 5'end of the overlapping ref exon\n",
    "\n",
    "test_novel_mi_ext = {\"Chromosome\": [\"chr1\"]*12,\n",
    "                \"Start\": [100,100,1000,2000] + [100,100,1000,1950] + [100,100,1000,2000],\n",
    "                \"End\": [2400,300,1200,2400] + [2400,300,1200,2400] + [2600,300,1200,2600],\n",
    "                \"Strand\": [\"+\"]*12,\n",
    "                \"Feature\": ([\"transcript\"] + [\"exon\"]*3)*3,\n",
    "                \"gene_id\": [\"id_nov_gene_1\"]*12,\n",
    "                \"transcript_id\": [\"nov_gene_1_tx_mi_ext_p\"]*4 + [\"nov_gene_1_tx_mi_ext_f\"]*4 + [\"nov_gene_1_tx_mi_ext_no_pas_motif_p\"]*4,\n",
    "                     \"gene_name\": [\"nov_gene_1\"]*12,\n",
    "                \"exon_number\": [None,1,2,3]*3,\n",
    "                 \"gene_type\": [\"protein_coding\"] * 12\n",
    "               }\n",
    "\n",
    "test_novel_trs.append(pr.from_dict(test_novel_mi_ext))\n",
    "update_pas_ends(pas_ends, \"chr1\", 2399, 2400, \"+\")\n",
    "pr.from_dict(test_novel_mi_ext)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Chromosome': ['chr1', 'chr2', 'chr1', 'chr1'],\n",
       " 'Start': [699, 880, 1599, 2399],\n",
       " 'End': [700, 881, 1600, 2400],\n",
       " 'Strand': ['+', '-', '+', '+']}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pas_ends"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>3800</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_li_spl_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_li_spl_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_li_spl_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2000</td>\n",
       "      <td>2200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_li_spl_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3800</td>\n",
       "      <td>4000</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_li_spl_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+---------------+---------------------+-------------+---------------+-------+\n",
       "| Chromosome   |     Start |       End | Strand       | Feature    | gene_id       | transcript_id       | gene_name   |   exon_number | +1    |\n",
       "| (category)   |   (int32) |   (int32) | (category)   | (object)   | (object)      | (object)            | (object)    |     (float64) | ...   |\n",
       "|--------------+-----------+-----------+--------------+------------+---------------+---------------------+-------------+---------------+-------|\n",
       "| chr1         |       100 |      3800 | +            | transcript | id_nov_gene_1 | nov_gene_1_li_spl_p | nov_gene_1  |           nan | ...   |\n",
       "| chr1         |       100 |       300 | +            | exon       | id_nov_gene_1 | nov_gene_1_li_spl_p | nov_gene_1  |             1 | ...   |\n",
       "| chr1         |      1000 |      1200 | +            | exon       | id_nov_gene_1 | nov_gene_1_li_spl_p | nov_gene_1  |             2 | ...   |\n",
       "| chr1         |      2000 |      2200 | +            | exon       | id_nov_gene_1 | nov_gene_1_li_spl_p | nov_gene_1  |             3 | ...   |\n",
       "| chr1         |      3800 |      4000 | +            | exon       | id_nov_gene_1 | nov_gene_1_li_spl_p | nov_gene_1  |             4 | ...   |\n",
       "+--------------+-----------+-----------+--------------+------------+---------------+---------------------+-------------+---------------+-------+\n",
       "Stranded PyRanges object has 5 rows and 10 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: gene_type"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#4. Last exon spliced (same rules as prev spliced apply - for simplicity will just make 1 that passes)\n",
    "test_novel_li_spl = {\"Chromosome\": [\"chr1\"]*5,\n",
    "                \"Start\": [100,100,1000,2000,3800],\n",
    "                \"End\": [3800,300,1200,2200,4000],\n",
    "                \"Strand\": [\"+\"]*5,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*4,\n",
    "                \"gene_id\": [\"id_nov_gene_1\"]*5,\n",
    "                \"transcript_id\": [\"nov_gene_1_li_spl_p\"]*5,\n",
    "                     \"gene_name\": [\"nov_gene_1\"]*5,\n",
    "                       \"exon_number\": [None,1,2,3,4],\n",
    "                       \"gene_type\": [\"protein_coding\"] * 5\n",
    "               }\n",
    "\n",
    "test_novel_trs.append(pr.from_dict(test_novel_li_spl))\n",
    "update_pas_ends(pas_ends, \"chr1\", 3999,4000, \"+\")\n",
    "pr.from_dict(test_novel_li_spl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>3600</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_novel_tx_le_ext_plus_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_novel_tx_le_ext_plus_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_novel_tx_le_ext_plus_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2000</td>\n",
       "      <td>2200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_novel_tx_le_ext_plus_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3000</td>\n",
       "      <td>3600</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_novel_tx_le_ext_plus_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+---------------+-----------------------------------+-------------+-------+\n",
       "| Chromosome   |     Start |       End | Strand       | Feature    | gene_id       | transcript_id                     | gene_name   | +2    |\n",
       "| (category)   |   (int32) |   (int32) | (category)   | (object)   | (object)      | (object)                          | (object)    | ...   |\n",
       "|--------------+-----------+-----------+--------------+------------+---------------+-----------------------------------+-------------+-------|\n",
       "| chr1         |      1000 |      3600 | +            | transcript | id_nov_gene_1 | nov_gene_1_novel_tx_le_ext_plus_p | nov_gene_1  | ...   |\n",
       "| chr1         |       100 |       300 | +            | exon       | id_nov_gene_1 | nov_gene_1_novel_tx_le_ext_plus_p | nov_gene_1  | ...   |\n",
       "| chr1         |      1000 |      1200 | +            | exon       | id_nov_gene_1 | nov_gene_1_novel_tx_le_ext_plus_p | nov_gene_1  | ...   |\n",
       "| chr1         |      2000 |      2200 | +            | exon       | id_nov_gene_1 | nov_gene_1_novel_tx_le_ext_plus_p | nov_gene_1  | ...   |\n",
       "| chr1         |      3000 |      3600 | +            | exon       | id_nov_gene_1 | nov_gene_1_novel_tx_le_ext_plus_p | nov_gene_1  | ...   |\n",
       "+--------------+-----------+-----------+--------------+------------+---------------+-----------------------------------+-------------+-------+\n",
       "Stranded PyRanges object has 5 rows and 10 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: exon_number, gene_type"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#5. Last exon extension (same rules as prev extensions apply - for simplicity will just make 1 that passes)\n",
    "test_novel_le_ext_plus = {\"Chromosome\": [\"chr1\"]*5,\n",
    "                \"Start\": [1000,100,1000,2000,3000],\n",
    "                \"End\": [3600,300,1200,2200,3600],\n",
    "                \"Strand\": [\"+\"]*5,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*4,\n",
    "                \"gene_id\": [\"id_nov_gene_1\"]*5,\n",
    "                \"transcript_id\": [\"nov_gene_1_novel_tx_le_ext_plus_p\"] * 5,\n",
    "                          \"gene_name\": [\"nov_gene_1\"]*5,\n",
    "                \"exon_number\": [None,1,2,3,4],\n",
    "                           \"gene_type\": [\"protein_coding\"] * 5}\n",
    "\n",
    "test_novel_trs.append(pr.from_dict(test_novel_le_ext_plus))\n",
    "update_pas_ends(pas_ends, \"chr1\", 3599, 3600, \"+\")\n",
    "pr.from_dict(test_novel_le_ext_plus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>50</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_le_ext_minus_f</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>50</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_le_ext_minus_f</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>700</td>\n",
       "      <td>800</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_le_ext_minus_f</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_le_ext_minus_f</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+---------------+------------------------------+-------------+-------+\n",
       "| Chromosome   |     Start |       End | Strand       | Feature    | gene_id       | transcript_id                | gene_name   | +2    |\n",
       "| (category)   |   (int32) |   (int32) | (category)   | (object)   | (object)      | (object)                     | (object)    | ...   |\n",
       "|--------------+-----------+-----------+--------------+------------+---------------+------------------------------+-------------+-------|\n",
       "| chr2         |        50 |      1200 | -            | transcript | id_nov_gene_2 | nov_gene_2_tx_le_ext_minus_f | nov_gene_2  | ...   |\n",
       "| chr2         |        50 |       200 | -            | exon       | id_nov_gene_2 | nov_gene_2_tx_le_ext_minus_f | nov_gene_2  | ...   |\n",
       "| chr2         |       700 |       800 | -            | exon       | id_nov_gene_2 | nov_gene_2_tx_le_ext_minus_f | nov_gene_2  | ...   |\n",
       "| chr2         |      1000 |      1200 | -            | exon       | id_nov_gene_2 | nov_gene_2_tx_le_ext_minus_f | nov_gene_2  | ...   |\n",
       "+--------------+-----------+-----------+--------------+------------+---------------+------------------------------+-------------+-------+\n",
       "Stranded PyRanges object has 4 rows and 10 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: exon_number, gene_type"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# last exon extension on the minus strand\n",
    "# Can only get it to fail the distance cut-off as not enough 'genomic space' for a 100nt extension\n",
    "test_novel_le_ext_minus = {\"Chromosome\": [\"chr2\"]*4,\n",
    "                \"Start\": [50,50,700,1000],\n",
    "                \"End\": [1200,200,800,1200],\n",
    "                \"Strand\": [\"-\"]*4,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*3,\n",
    "                \"gene_id\": [\"id_nov_gene_2\"]*4,\n",
    "                \"transcript_id\": [\"nov_gene_2_tx_le_ext_minus_f\"]*4,\n",
    "                           \"gene_name\": [\"nov_gene_2\"]*4,\n",
    "                \"exon_number\": [None,3,2,1],\n",
    "                             \"gene_type\": [\"protein_coding\"] * 4\n",
    "               }\n",
    "\n",
    "test_novel_trs.append(pr.from_dict(test_novel_le_ext_minus))\n",
    "pr.from_dict(test_novel_le_ext_minus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>400</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_mi_spl_minus_no_pas_motif_p</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>400</td>\n",
       "      <td>550</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_mi_spl_minus_no_pas_motif_p</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>700</td>\n",
       "      <td>800</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_mi_spl_minus_no_pas_motif_p</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_mi_spl_minus_no_pas_motif_p</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+---------------+-------------------------------------------+-------+\n",
       "| Chromosome   |     Start |       End | Strand       | Feature    | gene_id       | transcript_id                             | +3    |\n",
       "| (category)   |   (int32) |   (int32) | (category)   | (object)   | (object)      | (object)                                  | ...   |\n",
       "|--------------+-----------+-----------+--------------+------------+---------------+-------------------------------------------+-------|\n",
       "| chr2         |       400 |      1200 | -            | transcript | id_nov_gene_2 | nov_gene_2_tx_mi_spl_minus_no_pas_motif_p | ...   |\n",
       "| chr2         |       400 |       550 | -            | exon       | id_nov_gene_2 | nov_gene_2_tx_mi_spl_minus_no_pas_motif_p | ...   |\n",
       "| chr2         |       700 |       800 | -            | exon       | id_nov_gene_2 | nov_gene_2_tx_mi_spl_minus_no_pas_motif_p | ...   |\n",
       "| chr2         |      1000 |      1200 | -            | exon       | id_nov_gene_2 | nov_gene_2_tx_mi_spl_minus_no_pas_motif_p | ...   |\n",
       "+--------------+-----------+-----------+--------------+------------+---------------+-------------------------------------------+-------+\n",
       "Stranded PyRanges object has 4 rows and 10 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: gene_name, exon_number, gene_type"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1 last transcript on the minus strand that contains a PAS motif\n",
    "# (to check that motif matching on the opposite strand works correctly)\n",
    "test_novel_mi_spl_minus = {\"Chromosome\": [\"chr2\"]*4,\n",
    "                \"Start\": [400,400,700,1000],\n",
    "                \"End\": [1200,550,800,1200],\n",
    "                \"Strand\": [\"-\"]*4,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*3,\n",
    "                \"gene_id\": [\"id_nov_gene_2\"]*4,\n",
    "                \"transcript_id\": [\"nov_gene_2_tx_mi_spl_minus_no_pas_motif_p\"]*4,\n",
    "                           \"gene_name\": [\"nov_gene_2\"]*4,\n",
    "                \"exon_number\": [None,3,2,1],\n",
    "                             \"gene_type\": [\"protein_coding\"] * 4\n",
    "               }\n",
    "test_novel_trs.append(pr.from_dict(test_novel_mi_spl_minus))\n",
    "pr.from_dict(test_novel_mi_spl_minus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>500</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>350</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>500</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>1600</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1400</td>\n",
       "      <td>1600</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>1600</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1400</td>\n",
       "      <td>1600</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>1850</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_no_pas_no_motif_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_no_pas_no_motif_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_no_pas_no_motif_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1750</td>\n",
       "      <td>1850</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_s_no_pas_no_motif_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>2400</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2000</td>\n",
       "      <td>2400</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>2400</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1950</td>\n",
       "      <td>2400</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>2600</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_no_pas_motif_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_no_pas_motif_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_no_pas_motif_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2000</td>\n",
       "      <td>2600</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_mi_ext_no_pas_motif_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>3800</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_li_spl_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_li_spl_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_li_spl_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2000</td>\n",
       "      <td>2200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_li_spl_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3800</td>\n",
       "      <td>4000</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_li_spl_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>3600</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_novel_tx_le_ext_plus_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_novel_tx_le_ext_plus_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_novel_tx_le_ext_plus_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2000</td>\n",
       "      <td>2200</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_novel_tx_le_ext_plus_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3000</td>\n",
       "      <td>3600</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_novel_tx_le_ext_plus_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr2</td>\n",
       "      <td>880</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_fi_ext_p</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chr2</td>\n",
       "      <td>880</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_fi_ext_p</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr2</td>\n",
       "      <td>880</td>\n",
       "      <td>1500</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_fi_ext_f</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr2</td>\n",
       "      <td>880</td>\n",
       "      <td>1500</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_fi_ext_f</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr2</td>\n",
       "      <td>50</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_le_ext_minus_f</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr2</td>\n",
       "      <td>50</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_le_ext_minus_f</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chr2</td>\n",
       "      <td>700</td>\n",
       "      <td>800</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_le_ext_minus_f</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr2</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_le_ext_minus_f</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr2</td>\n",
       "      <td>400</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_mi_spl_minus_no_pas_motif_p</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chr2</td>\n",
       "      <td>400</td>\n",
       "      <td>550</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_mi_spl_minus_no_pas_motif_p</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr2</td>\n",
       "      <td>700</td>\n",
       "      <td>800</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_mi_spl_minus_no_pas_motif_p</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr2</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_mi_spl_minus_no_pas_motif_p</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+---------------+-------------------------------------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | Feature    | gene_id       | transcript_id                             | +3    |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | (object)   | (object)      | (object)                                  | ...   |\n",
       "|--------------+-----------+-----------+--------------+------------+---------------+-------------------------------------------+-------|\n",
       "| chr1         | 100       | 700       | +            | transcript | id_nov_gene_1 | nov_gene_1_tx_fi_p                        | ...   |\n",
       "| chr1         | 100       | 300       | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_fi_p                        | ...   |\n",
       "| chr1         | 500       | 700       | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_fi_p                        | ...   |\n",
       "| chr1         | 100       | 700       | +            | transcript | id_nov_gene_1 | nov_gene_1_tx_fi_f                        | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...        | ...           | ...                                       | ...   |\n",
       "| chr2         | 400       | 1200      | -            | transcript | id_nov_gene_2 | nov_gene_2_tx_mi_spl_minus_no_pas_motif_p | ...   |\n",
       "| chr2         | 400       | 550       | -            | exon       | id_nov_gene_2 | nov_gene_2_tx_mi_spl_minus_no_pas_motif_p | ...   |\n",
       "| chr2         | 700       | 800       | -            | exon       | id_nov_gene_2 | nov_gene_2_tx_mi_spl_minus_no_pas_motif_p | ...   |\n",
       "| chr2         | 1000      | 1200      | -            | exon       | id_nov_gene_2 | nov_gene_2_tx_mi_spl_minus_no_pas_motif_p | ...   |\n",
       "+--------------+-----------+-----------+--------------+------------+---------------+-------------------------------------------+-------+\n",
       "Stranded PyRanges object has 52 rows and 10 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: gene_name, exon_number, gene_type"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_novel = pr.concat(test_novel_trs)\n",
    "test_novel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>500</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>350</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>chr2</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_mi_spl_minus_no_pas_motif_p</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chr2</td>\n",
       "      <td>100</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_ref_gene_2</td>\n",
       "      <td>ref_g2_tr_1</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chr2</td>\n",
       "      <td>100</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_2</td>\n",
       "      <td>ref_g2_tr_1</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chr2</td>\n",
       "      <td>700</td>\n",
       "      <td>800</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_2</td>\n",
       "      <td>ref_g2_tr_1</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr2</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_2</td>\n",
       "      <td>ref_g2_tr_1</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+---------------+--------------------+-------------+---------------+----------------+\n",
       "| Chromosome   | Start     | End       | Strand       | Feature    | gene_id       | transcript_id      | gene_name   | exon_number   | gene_type      |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | (object)   | (object)      | (object)           | (object)    | (float64)     | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+---------------+--------------------+-------------+---------------+----------------|\n",
       "| chr1         | 100       | 700       | +            | transcript | id_nov_gene_1 | nov_gene_1_tx_fi_p | nov_gene_1  | nan           | protein_coding |\n",
       "| chr1         | 100       | 300       | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_fi_p | nov_gene_1  | 1.0           | protein_coding |\n",
       "| chr1         | 500       | 700       | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_fi_p | nov_gene_1  | 2.0           | protein_coding |\n",
       "| chr1         | 100       | 700       | +            | transcript | id_nov_gene_1 | nov_gene_1_tx_fi_f | nov_gene_1  | nan           | protein_coding |\n",
       "| ...          | ...       | ...       | ...          | ...        | ...           | ...                | ...         | ...           | ...            |\n",
       "| chr2         | 100       | 1200      | -            | transcript | id_ref_gene_2 | ref_g2_tr_1        | ref_gene_2  | nan           | protein_coding |\n",
       "| chr2         | 100       | 200       | -            | exon       | id_ref_gene_2 | ref_g2_tr_1        | ref_gene_2  | 3.0           | protein_coding |\n",
       "| chr2         | 700       | 800       | -            | exon       | id_ref_gene_2 | ref_g2_tr_1        | ref_gene_2  | 2.0           | protein_coding |\n",
       "| chr2         | 1000      | 1200      | -            | exon       | id_ref_gene_2 | ref_g2_tr_1        | ref_gene_2  | 1.0           | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+---------------+--------------------+-------------+---------------+----------------+\n",
       "Stranded PyRanges object has 61 rows and 10 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_trs = pr.concat([test_novel, test_ref])\n",
    "test_trs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PAPA_tests</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PAPA_tests</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>500</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_p</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PAPA_tests</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PAPA_tests</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100</td>\n",
       "      <td>350</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_1</td>\n",
       "      <td>nov_gene_1_tx_fi_f</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PAPA_tests</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>chr2</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_nov_gene_2</td>\n",
       "      <td>nov_gene_2_tx_mi_spl_minus_no_pas_motif_p</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PAPA_tests</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chr2</td>\n",
       "      <td>100</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>id_ref_gene_2</td>\n",
       "      <td>ref_g2_tr_1</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PAPA_tests</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chr2</td>\n",
       "      <td>100</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_2</td>\n",
       "      <td>ref_g2_tr_1</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PAPA_tests</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chr2</td>\n",
       "      <td>700</td>\n",
       "      <td>800</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_2</td>\n",
       "      <td>ref_g2_tr_1</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PAPA_tests</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr2</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>id_ref_gene_2</td>\n",
       "      <td>ref_g2_tr_1</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PAPA_tests</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+---------------+--------------------+-------------+---------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | Feature    | gene_id       | transcript_id      | gene_name   | exon_number   | +2    |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | (object)   | (object)      | (object)           | (object)    | (float64)     | ...   |\n",
       "|--------------+-----------+-----------+--------------+------------+---------------+--------------------+-------------+---------------+-------|\n",
       "| chr1         | 100       | 700       | +            | transcript | id_nov_gene_1 | nov_gene_1_tx_fi_p | nov_gene_1  | nan           | ...   |\n",
       "| chr1         | 100       | 300       | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_fi_p | nov_gene_1  | 1.0           | ...   |\n",
       "| chr1         | 500       | 700       | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_fi_p | nov_gene_1  | 2.0           | ...   |\n",
       "| chr1         | 100       | 700       | +            | transcript | id_nov_gene_1 | nov_gene_1_tx_fi_f | nov_gene_1  | nan           | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...        | ...           | ...                | ...         | ...           | ...   |\n",
       "| chr2         | 100       | 1200      | -            | transcript | id_ref_gene_2 | ref_g2_tr_1        | ref_gene_2  | nan           | ...   |\n",
       "| chr2         | 100       | 200       | -            | exon       | id_ref_gene_2 | ref_g2_tr_1        | ref_gene_2  | 3.0           | ...   |\n",
       "| chr2         | 700       | 800       | -            | exon       | id_ref_gene_2 | ref_g2_tr_1        | ref_gene_2  | 2.0           | ...   |\n",
       "| chr2         | 1000      | 1200      | -            | exon       | id_ref_gene_2 | ref_g2_tr_1        | ref_gene_2  | 1.0           | ...   |\n",
       "+--------------+-----------+-----------+--------------+------------+---------------+--------------------+-------------+---------------+-------+\n",
       "Stranded PyRanges object has 61 rows and 11 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: gene_type, Source"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_trs.Source = \"PAPA_tests\"\n",
    "test_trs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_novel.to_gtf(\"../tests/test_annotations/test_transcripts_novel.gtf\")\n",
    "test_ref.to_gtf(\"../tests/test_annotations/test_transcripts_reference.gtf\")\n",
    "test_trs.to_gtf(\"../tests/test_annotations/test_transcripts_combined.gtf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate fake reference genome for test transcripts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chromosome\n",
       "chr1    4000\n",
       "chr2    1500\n",
       "Name: End, dtype: int32"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Each chrom will start at 0 - from there on need to define a large enough range for the region to get trs + a bit of intergenic sequence\n",
    "test_trs.as_df().groupby(\"Chromosome\")[\"End\"].max()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=MutableSeq('AATAGACTGATTAATACATTGAAGAGCTGGTGACCGTTGCCCCGTCTAGAGGCA...GAG'), id='chr1', name='chr1', description='<unknown description>', dbxrefs=[])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For simplicity let's make both chromosomes 10kb -\n",
    "# Need to generate random sequences for each chromosome\n",
    "random.seed(123)\n",
    "chr1 = \"\".join(random.choices(\"ATCG\", k=10000))\n",
    "chr2 = \"\".join(random.choices(\"ATCG\", k=10000))\n",
    "\n",
    "seq_chr1 = SeqRecord(seq=MutableSeq(chr1), id=\"chr1\", name=\"chr1\")\n",
    "seq_chr2 = SeqRecord(seq=MutableSeq(chr2), id=\"chr2\", name=\"chr2\")\n",
    "\n",
    "seq_chr1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now need to make sure specific last exons do / do not contain PAS motifs\n",
    "\n",
    "**To contain motifs**\n",
    "- nov_gene_2_tx_mi_spl_minus_no_pas_motif_p\n",
    "- nov_gene_1_tx_mi_ext_no_pas_motif_p\n",
    "\n",
    "**To not contain motifs**\n",
    "- nov_gene_1_tx_mi_s_no_pas_no_motif_f\n",
    "\n",
    "For each of these, need to extract last exons & check the corresponding sequence does not contain any PAS motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "gruber_pas_motifs = \"\"\"AATAAA\n",
    "ATTAAA\n",
    "TATAAA\n",
    "AGTAAA\n",
    "AATACA\n",
    "CATAAA\n",
    "AATATA\n",
    "GATAAA\n",
    "AATGAA\n",
    "AAGAAA\n",
    "ACTAAA\n",
    "AATAGA\n",
    "AATAAT\n",
    "AACAAA\n",
    "ATTACA\n",
    "ATTATA\n",
    "AACAAG\n",
    "AATAAG\"\"\".split(\"\\n\")\n",
    "\n",
    "# Generate motifs object\n",
    "# from generate a list of Seq objects storing motifs\n",
    "pas_seqs = [Seq(motif) for motif in gruber_pas_motifs]\n",
    "pas_motifs = motifs.create(pas_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# x = Seq(seq_chr2.seq[400:500])\n",
    "# print(x)\n",
    "# print(Seq.reverse_complement(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for idx, motif in pas_motifs.instances.search(Seq(seq_chr2.seq[400:500]).reverse_complement()):\n",
    "#     print(idx)\n",
    "#     print(motif)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------+-----------+-----------+--------------+------------+---------------+-------------------------------------------+-------+\n",
      "| Chromosome   |     Start |       End | Strand       | Feature    | gene_id       | transcript_id                             | +3    |\n",
      "| (category)   |   (int32) |   (int32) | (category)   | (object)   | (object)      | (object)                                  | ...   |\n",
      "|--------------+-----------+-----------+--------------+------------+---------------+-------------------------------------------+-------|\n",
      "| chr1         |      2000 |      2600 | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_mi_ext_no_pas_motif_p       | ...   |\n",
      "| chr2         |       400 |       550 | -            | exon       | id_nov_gene_2 | nov_gene_2_tx_mi_spl_minus_no_pas_motif_p | ...   |\n",
      "+--------------+-----------+-----------+--------------+------------+---------------+-------------------------------------------+-------+\n",
      "Stranded PyRanges object has 2 rows and 10 columns from 2 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
      "3 hidden columns: gene_name, exon_number, gene_type\n",
      "nov_gene_1_tx_mi_ext_no_pas_motif_p - found motifs - \n",
      "nov_gene_2_tx_mi_spl_minus_no_pas_motif_p - found motifs - 62_TATAAA\n"
     ]
    }
   ],
   "source": [
    "# These transcripts should have PAS motifs within the final 100bp\n",
    "tx_w_motifs = [\"nov_gene_1_tx_mi_ext_no_pas_motif_p\", \"nov_gene_2_tx_mi_spl_minus_no_pas_motif_p\"]\n",
    "\n",
    "\n",
    "tes_tx_w_motifs = get_terminal_regions(test_novel.subset(lambda df: (df[\"Feature\"] == \"exon\") & (df[\"transcript_id\"].isin(tx_w_motifs))))\n",
    "print(tes_tx_w_motifs)\n",
    "\n",
    "for k, v in tes_tx_w_motifs:\n",
    "    for idx, row in v.iterrows():\n",
    "        if row[\"Strand\"] == \"+\":\n",
    "            reg_start = row[\"End\"] - 100\n",
    "            reg_end = row[\"End\"]\n",
    "            reg_str = row[\"Strand\"]\n",
    "        \n",
    "        else:\n",
    "            reg_start = row[\"Start\"]\n",
    "            reg_end = row[\"Start\"] + 100\n",
    "            reg_str = row[\"Strand\"]\n",
    "            \n",
    "        if k[0] == \"chr1\":\n",
    "            if reg_str == \"+\":\n",
    "                found_motifs = \";\".join(\"_\".join([str(idx), str(motif)]) for idx, motif in pas_motifs.instances.search(seq_chr1.seq[reg_start:reg_end]))\n",
    "            \n",
    "            else:\n",
    "                reg_seq = Seq(seq_chr2.seq[reg_start:reg_end]).reverse_complement()                \n",
    "                found_motifs = \";\".join(\"_\".join([str(idx), str(motif)]) for idx, motif in pas_motifs.instances.search(Seq(seq_chr1.seq[reg_start:reg_end]).reverse_complement()))\n",
    "            \n",
    "        else:\n",
    "            if reg_str == \"+\":\n",
    "                found_motifs = \";\".join(\"_\".join([str(idx), str(motif)]) for idx, motif in pas_motifs.instances.search(seq_chr2.seq[reg_start:reg_end]))\n",
    "            \n",
    "            else:\n",
    "                #x = \"_\".join(idx, motif for idx, motif in pas_motifs.instances.search(Seq.reverse_complement(Seq(seq_chr2.seq[reg_start:reg_end]))))\n",
    "                #print(x)    \n",
    "                reg_seq = Seq(seq_chr2.seq[reg_start:reg_end]).reverse_complement()\n",
    "                found_motifs = \";\".join(\"_\".join([str(idx), str(motif)]) for idx, motif in pas_motifs.instances.search(reg_seq))\n",
    "            \n",
    "        \n",
    "        print(f\"{row['transcript_id']} - found motifs - {found_motifs}\")\n",
    "\n",
    "# first check if by random chance the motif has \n",
    "#\";\".join(\"_\".join(idx, motif) for idx, motif in pas_motifs.instances.search(seq_chr1.seq[2600-100:2600]))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MutableSeq('AATAAA')"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# minus strand tx -nov_gene_2_tx_mi_spl_minus_no_pas_motif_p - has a PAS motif within last 100bp\n",
    "# need to add motif for nov_gene_1_tx_mi_ext_no_pas_motif_p\n",
    "mut_tx_te = tes_tx_w_motifs.subset(lambda df: df[\"transcript_id\"] == \"nov_gene_1_tx_mi_ext_no_pas_motif_p\")\n",
    "# Set seq to mutate to ~ 25 bp upstream of end\n",
    "mut_reg_start = mut_tx_te.three_end().End[0] - 26\n",
    "mut_reg_end = mut_tx_te.three_end().End[0] - 20\n",
    "mut_reg_end - mut_reg_start\n",
    "\n",
    "seq_chr1.seq[mut_reg_start:mut_reg_end] = \"AATAAA\"\n",
    "\n",
    "seq_chr1.seq[mut_reg_start:mut_reg_end]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'74_AATAAA'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# double check it has a motif\n",
    "(\";\".join(\"_\".join([str(idx), str(motif)]) \n",
    "          for idx, motif in pas_motifs.instances.search(seq_chr1.seq[mut_tx_te.End.iloc[0] - 100:mut_tx_te.End.iloc[0]])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------+-----------+-----------+--------------+------------+---------------+--------------------------------------+-------------+-------+\n",
      "| Chromosome   |     Start |       End | Strand       | Feature    | gene_id       | transcript_id                        | gene_name   | +2    |\n",
      "| (category)   |   (int32) |   (int32) | (category)   | (object)   | (object)      | (object)                             | (object)    | ...   |\n",
      "|--------------+-----------+-----------+--------------+------------+---------------+--------------------------------------+-------------+-------|\n",
      "| chr1         |      1750 |      1850 | +            | exon       | id_nov_gene_1 | nov_gene_1_tx_mi_s_no_pas_no_motif_f | nov_gene_1  | ...   |\n",
      "+--------------+-----------+-----------+--------------+------------+---------------+--------------------------------------+-------------+-------+\n",
      "Stranded PyRanges object has 1 rows and 10 columns from 1 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
      "2 hidden columns: exon_number, gene_type\n",
      "nov_gene_1_tx_mi_s_no_pas_no_motif_f - found motifs - \n"
     ]
    }
   ],
   "source": [
    "# now check this tx - nov_gene_1_tx_mi_s_no_pas_no_motif_f - doesn't have any PAS motifs\n",
    "tx_w_no_motifs = [\"nov_gene_1_tx_mi_s_no_pas_no_motif_f\"]\n",
    "\n",
    "tes_tx_w_no_motifs = get_terminal_regions(test_novel.subset(lambda df: (df[\"Feature\"] == \"exon\") & (df[\"transcript_id\"].isin(tx_w_no_motifs))))\n",
    "print(tes_tx_w_no_motifs)\n",
    "\n",
    "for k, v in tes_tx_w_no_motifs:\n",
    "    for idx, row in v.iterrows():\n",
    "        if row[\"Strand\"] == \"+\":\n",
    "            reg_start = row[\"End\"] - 100\n",
    "            reg_end = row[\"End\"]\n",
    "            reg_str = row[\"Strand\"]\n",
    "        \n",
    "        else:\n",
    "            reg_start = row[\"Start\"]\n",
    "            reg_end = row[\"Start\"] + 100\n",
    "            reg_str = row[\"Strand\"]\n",
    "            \n",
    "        if k[0] == \"chr1\":\n",
    "            if reg_str == \"+\":\n",
    "                found_motifs = \";\".join(\"_\".join([str(idx), str(motif)]) for idx, motif in pas_motifs.instances.search(seq_chr1.seq[reg_start:reg_end]))\n",
    "            \n",
    "            else:\n",
    "                reg_seq = Seq(seq_chr2.seq[reg_start:reg_end]).reverse_complement()                \n",
    "                found_motifs = \";\".join(\"_\".join([str(idx), str(motif)]) for idx, motif in pas_motifs.instances.search(Seq(seq_chr1.seq[reg_start:reg_end]).reverse_complement()))\n",
    "            \n",
    "        else:\n",
    "            if reg_str == \"+\":\n",
    "                found_motifs = \";\".join(\"_\".join([str(idx), str(motif)]) for idx, motif in pas_motifs.instances.search(seq_chr2.seq[reg_start:reg_end]))\n",
    "            \n",
    "            else:\n",
    "                #x = \"_\".join(idx, motif for idx, motif in pas_motifs.instances.search(Seq.reverse_complement(Seq(seq_chr2.seq[reg_start:reg_end]))))\n",
    "                #print(x)    \n",
    "                reg_seq = Seq(seq_chr2.seq[reg_start:reg_end]).reverse_complement()\n",
    "                found_motifs = \";\".join(\"_\".join([str(idx), str(motif)]) for idx, motif in pas_motifs.instances.search(reg_seq))\n",
    "            \n",
    "        \n",
    "        print(f\"{row['transcript_id']} - found motifs - {found_motifs}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../tests/test_annotations/test_genome.fa\", \"w\") as out_fasta:\n",
    "    for seq in [seq_chr1, seq_chr2]:\n",
    "        SeqIO.write(seq,out_fasta, \"fasta\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now need to make transcripts FASTA\n",
    "os.system(\"which gffread\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf = \"../tests/test_annotations/test_transcripts_combined.gtf\"\n",
    "genome_fa = \"../tests/test_annotations/test_genome.fa\"\n",
    "out_tx_fa = \"../tests/test_annotations/test_transcripts.fa\"\n",
    "\n",
    "os.system(f\"gffread -w {out_tx_fa} -g {genome_fa} {gtf}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate BED file of PolyASites\n",
    "\n",
    "txipts that pass all isoform filters but don't havr he 'no_pas_motif' annotation (i.e. don't have an annotated PAS but pass due to motif)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>699</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:700:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1599</td>\n",
       "      <td>1600</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1600:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2399</td>\n",
       "      <td>2400</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:2400:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3799</td>\n",
       "      <td>3800</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:3800:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3599</td>\n",
       "      <td>3600</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:3600:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr2</td>\n",
       "      <td>880</td>\n",
       "      <td>881</td>\n",
       "      <td>-</td>\n",
       "      <td>chr2:881:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------------+\n",
       "| Chromosome   |     Start |       End | Strand       | Name        |\n",
       "| (category)   |   (int32) |   (int32) | (category)   | (object)    |\n",
       "|--------------+-----------+-----------+--------------+-------------|\n",
       "| chr1         |       699 |       700 | +            | chr1:700:+  |\n",
       "| chr1         |      1599 |      1600 | +            | chr1:1600:+ |\n",
       "| chr1         |      2399 |      2400 | +            | chr1:2400:+ |\n",
       "| chr1         |      3799 |      3800 | +            | chr1:3800:+ |\n",
       "| chr1         |      3599 |      3600 | +            | chr1:3600:+ |\n",
       "| chr2         |       880 |       881 | -            | chr2:881:-  |\n",
       "+--------------+-----------+-----------+--------------+-------------+\n",
       "Stranded PyRanges object has 6 rows and 5 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pas_bed = (test_novel.subset(lambda df: df.transcript_id.str.endswith(\"_p\"))\n",
    " .subset(lambda df: ~df.transcript_id.str.contains(\"no_pas_motif\", regex=False))\n",
    " .subset(lambda df: df[\"Feature\"] == \"transcript\")\n",
    " .three_end() # last templated coord as the rep for the PAS\n",
    " .assign(\"Name\", lambda df: df[\"Chromosome\"].str.cat(df[[\"End\", \"Strand\"]].astype(str), sep=\":\"))\n",
    " [[\"Name\"]]\n",
    " )\n",
    "\n",
    "pas_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "pas_bed.to_bed(\"../tests/test_annotations/test_pas.bed\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:papa_pipeline]",
   "language": "python",
   "name": "conda-env-papa_pipeline-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
