{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b5170bd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "import time\n",
    "from timeit import default_timer as timer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "fb15ee4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------|\n",
       "|            1 |        10 |       340 | +            | transcript | ref_gene_1 | ref_tr_1        |           nan | protein_coding |\n",
       "|            1 |        10 |        30 | +            | exon       | ref_gene_1 | ref_tr_1        |             1 | protein_coding |\n",
       "|            1 |       100 |       120 | +            | exon       | ref_gene_1 | ref_tr_1        |             2 | protein_coding |\n",
       "|            1 |       200 |       220 | +            | exon       | ref_gene_1 | ref_tr_1        |             3 | protein_coding |\n",
       "|            1 |       300 |       340 | +            | exon       | ref_gene_1 | ref_tr_1        |             4 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------+\n",
       "Stranded PyRanges object has 5 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A plus strand transcript\n",
    "test_ref_tr1 = {\"Chromosome\": [1]*5,\n",
    "                \"Start\": [10,10,100,200,300],\n",
    "                \"End\": [340,30,120,220,340],\n",
    "                \"Strand\": [\"+\"]*5,\n",
    "                \"Feature\": [\"transcript\"] +[\"exon\"]*4,\n",
    "                \"gene_id\": [\"ref_gene_1\"]*5,\n",
    "                \"transcript_id\": [\"ref_tr_1\"] * 5,\n",
    "                \"exon_number\": [None,1,2,3,4],\n",
    "                \"gene_type\": [\"protein_coding\"] * 5}\n",
    "\n",
    "\n",
    "pr.from_dict(test_ref_tr1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1f982b51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Chromosome': [],\n",
       " 'Start': [],\n",
       " 'End': [],\n",
       " 'Strand': [],\n",
       " 'Feature': [],\n",
       " 'gene_id': [],\n",
       " 'transcript_id': []}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{\"Chromosome\": [],\n",
    "                \"Start\": [],\n",
    "                \"End\": [],\n",
    "                \"Strand\": [],\n",
    "                \"Feature\": [],\n",
    "                \"gene_id\": [],\n",
    "                \"transcript_id\": []\n",
    "               }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "37648c16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>80</td>\n",
       "      <td>90</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------+\n",
       "| Chromosome   | Start     | End       | Strand       | Feature    | gene_id    | transcript_id   | exon_number   | gene_type      |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | (object)   | (object)   | (object)        | (float64)     | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------|\n",
       "| 1            | 10        | 340       | +            | transcript | ref_gene_1 | ref_tr_1        | nan           | protein_coding |\n",
       "| 1            | 10        | 30        | +            | exon       | ref_gene_1 | ref_tr_1        | 1.0           | protein_coding |\n",
       "| 1            | 100       | 120       | +            | exon       | ref_gene_1 | ref_tr_1        | 2.0           | protein_coding |\n",
       "| 1            | 200       | 220       | +            | exon       | ref_gene_1 | ref_tr_1        | 3.0           | protein_coding |\n",
       "| ...          | ...       | ...       | ...          | ...        | ...        | ...             | ...           | ...            |\n",
       "| 2            | 10        | 120       | -            | transcript | ref_gene_2 | ref_tr_2        | nan           | protein_coding |\n",
       "| 2            | 10        | 20        | -            | exon       | ref_gene_2 | ref_tr_2        | 3.0           | protein_coding |\n",
       "| 2            | 80        | 90        | -            | exon       | ref_gene_2 | ref_tr_2        | 2.0           | protein_coding |\n",
       "| 2            | 100       | 120       | -            | exon       | ref_gene_2 | ref_tr_2        | 1.0           | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------+\n",
       "Stranded PyRanges object has 9 rows and 9 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A minus strand transcript\n",
    "test_ref_tr2 = {\"Chromosome\": [2]*4,\n",
    "                \"Start\": [10,10,80,100],\n",
    "                \"End\": [120,20,90,120],\n",
    "                \"Strand\": [\"-\"]*4,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*3,\n",
    "                \"gene_id\": [\"ref_gene_2\"]*4,\n",
    "                \"transcript_id\": [\"ref_tr_2\"]*4,\n",
    "                \"exon_number\": [None,3,2,1],\n",
    "                \"gene_type\": [\"protein_coding\"] * 4\n",
    "               }\n",
    "\n",
    "\n",
    "test_ref = pr.concat([pr.from_dict(test_ref_tr1), pr.from_dict(test_ref_tr2)])\n",
    "test_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "fbba9f36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>35</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------|\n",
       "|            1 |        10 |        70 | +            | transcript | nov_gene_1 | nov_tx_fi_p     |           nan | protein_coding |\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_fi_p     |             1 | protein_coding |\n",
       "|            1 |        50 |        70 | +            | exon       | nov_gene_1 | nov_tx_fi_p     |             2 | protein_coding |\n",
       "|            1 |        10 |        70 | +            | transcript | nov_gene_1 | nov_tx_fi_f     |           nan | protein_coding |\n",
       "|            1 |        10 |        35 | +            | exon       | nov_gene_1 | nov_tx_fi_f     |             1 | protein_coding |\n",
       "|            1 |        50 |        70 | +            | exon       | nov_gene_1 | nov_tx_fi_f     |             2 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------+\n",
       "Stranded PyRanges object has 6 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now make test novel transcripts to cover my test cases\n",
    "\n",
    "# 1. Novel last exons in first intron of annotated transcript (e.g. STMN2)\n",
    "# For these to pass, they should share an identical 3'end with a first exon of a known transcript\n",
    "\n",
    "test_novel_fi = {\"Chromosome\": [1]*6,\n",
    "                \"Start\": [10] + [10,50] + [10] + [10,50],\n",
    "                \"End\": [70,30,70,70,35,70],\n",
    "                \"Strand\": [\"+\"]*6,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*2 + [\"transcript\"] + [\"exon\"]*2,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*6,\n",
    "                \"transcript_id\": [\"nov_tx_fi_p\"]*3 + [\"nov_tx_fi_f\"]*3,\n",
    "                \"exon_number\": [None,1,2,None,1,2],\n",
    "                 \"gene_type\": [\"protein_coding\"] * 6\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_fi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "213b2de9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------|\n",
       "|            1 |        10 |       160 | +            | transcript | nov_gene_1 | nov_tx_si_p     |           nan | protein_coding |\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_si_p     |             1 | protein_coding |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_si_p     |             2 | protein_coding |\n",
       "|            1 |       140 |       160 | +            | exon       | nov_gene_1 | nov_tx_si_p     |             3 | protein_coding |\n",
       "|            1 |        50 |       160 | +            | transcript | nov_gene_1 | nov_tx_si_f     |           nan | protein_coding |\n",
       "|            1 |        50 |        70 | +            | exon       | nov_gene_1 | nov_tx_si_f     |             1 | protein_coding |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_si_f     |             2 | protein_coding |\n",
       "|            1 |       140 |       160 | +            | exon       | nov_gene_1 | nov_tx_si_f     |             3 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------+\n",
       "Stranded PyRanges object has 8 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2. Internal intron, spliced in last exon (fully contained within last exon) (e.g. ONECUT1)\n",
    "# For these to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "\n",
    "test_novel_si = {\"Chromosome\": [1]*8,\n",
    "                \"Start\": [10,10,100,140] + [50,50,100,140],\n",
    "                \"End\": [160,30,120,160] + [160,70,120,160],\n",
    "                \"Strand\": [\"+\"]*8,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*3 + [\"transcript\"] + [\"exon\"]*3,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*8,\n",
    "                \"transcript_id\": [\"nov_tx_si_p\"]*4 + [\"nov_tx_si_f\"]*4,\n",
    "                \"exon_number\": [None,1,2,3,None,1,2,3],\n",
    "                 \"gene_type\": [\"protein_coding\"] * 8\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_si)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "2f34a0e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>240</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>240</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>240</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>240</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------|\n",
       "|            1 |        10 |       240 | +            | transcript | nov_gene_1 | nov_tx_bl_p     |           nan | protein_coding |\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_bl_p     |             1 | protein_coding |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_bl_p     |             2 | protein_coding |\n",
       "|            1 |       200 |       240 | +            | exon       | nov_gene_1 | nov_tx_bl_p     |             3 | protein_coding |\n",
       "|            1 |        10 |       240 | +            | transcript | nov_gene_1 | nov_tx_bl_f     |           nan | protein_coding |\n",
       "|            1 |        50 |        70 | +            | exon       | nov_gene_1 | nov_tx_bl_f     |             1 | protein_coding |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_bl_f     |             2 | protein_coding |\n",
       "|            1 |       200 |       240 | +            | exon       | nov_gene_1 | nov_tx_bl_f     |             3 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------+\n",
       "Stranded PyRanges object has 8 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3. Internal intron bleedthrough (e.g. SIN3B)\n",
    "# For these events to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "\n",
    "test_novel_bl = {\"Chromosome\": [1]*8,\n",
    "                \"Start\": [10,10,100,200] + [10,50,100,200],\n",
    "                \"End\": [240,30,120,240] + [240,70,120,240],\n",
    "                \"Strand\": [\"+\"]*8,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*3 + [\"transcript\"] + [\"exon\"]*3,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*8,\n",
    "                \"transcript_id\": [\"nov_tx_bl_p\"]*4 + [\"nov_tx_bl_f\"]*4,\n",
    "                \"exon_number\": [None,1,2,3,None,1,2,3],\n",
    "                 \"gene_type\": [\"protein_coding\"] * 8\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_bl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "f32f14c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>130</td>\n",
       "      <td>140</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>150</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>4.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------|\n",
       "|            1 |        10 |       160 | +            | transcript | nov_gene_1 | nov_tx_mult_p   |           nan | protein_coding |\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_mult_p   |             1 | protein_coding |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_mult_p   |             2 | protein_coding |\n",
       "|            1 |       130 |       140 | +            | exon       | nov_gene_1 | nov_tx_mult_p   |             3 | protein_coding |\n",
       "|            1 |       150 |       160 | +            | exon       | nov_gene_1 | nov_tx_mult_p   |             4 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------+\n",
       "Stranded PyRanges object has 5 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 4. Internal intron with novel internal and terminal exon (e.g.)\n",
    "# For these events to pass, they should match the intron chain of a known transcript,\n",
    "# but have a, continuous chain of length n of novel events at the 3'end of the transcript\n",
    "# (n can be varied)\n",
    "# Event know from NP is fully contained within annotated intron - also set this constraint5\n",
    "\n",
    "test_novel_mult = {\"Chromosome\": [1]*5, \n",
    "                   \"Start\": [10,10,100,130,150],\n",
    "                   \"End\": [160,30,120,140,160],\n",
    "                   \"Strand\": [\"+\"]*5,\n",
    "                   \"Feature\": [\"transcript\"] + [\"exon\"]*4,\n",
    "                   \"gene_id\": [\"nov_gene_1\"]*5,\n",
    "                   \"transcript_id\": [\"nov_tx_mult_p\"]*5,\n",
    "                   \"exon_number\": [None,1,2,3,4],\n",
    "                   \"gene_type\": [\"protein_coding\"] * 5\n",
    "                    }\n",
    "\n",
    "pr.from_dict(test_novel_mult)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "c68fdf5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>310</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>4.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>330</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>5.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------|\n",
       "|            1 |        10 |       340 | +            | transcript | nov_gene_1 | nov_tx_3ui_p    |           nan | protein_coding |\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |             1 | protein_coding |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |             2 | protein_coding |\n",
       "|            1 |       200 |       220 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |             3 | protein_coding |\n",
       "|            1 |       300 |       310 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |             4 | protein_coding |\n",
       "|            1 |       330 |       340 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |             5 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+---------------+----------------+\n",
       "Stranded PyRanges object has 6 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 5. 3'UTR intron fully contained within an annotated 3'UTR (e.g. TDP-43)\n",
    "# For this to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "# (Annotate as a 3'UTR intron (spliced out) after filtering for intron chain match)\n",
    "test_novel_3ui = {\"Chromosome\": [1]*6,\n",
    "                \"Start\": [10,10,100,200,300,330],\n",
    "                \"End\": [340,30,120,220,310,340],\n",
    "                \"Strand\": [\"+\"]*6,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*5,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*6,\n",
    "                \"transcript_id\": [\"nov_tx_3ui_p\"]*6,\n",
    "                  \"exon_number\": [None,1,2,3,4,5],\n",
    "                  \"gene_type\": [\"protein_coding\"] * 6\n",
    "                 }\n",
    "\n",
    "pr.from_dict(test_novel_3ui)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "883cf45e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>380</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>360</td>\n",
       "      <td>380</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>4.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-------------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id     |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)          |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-------------------+---------------+----------------|\n",
       "|            1 |        10 |       380 | +            | transcript | nov_gene_1 | nov_tx_exc_dist_p |           nan | protein_coding |\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_exc_dist_p |             1 | protein_coding |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_exc_dist_p |             2 | protein_coding |\n",
       "|            1 |       200 |       220 | +            | exon       | nov_gene_1 | nov_tx_exc_dist_p |             3 | protein_coding |\n",
       "|            1 |       360 |       380 | +            | exon       | nov_gene_1 | nov_tx_exc_dist_p |             4 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-------------------+---------------+----------------+\n",
       "Stranded PyRanges object has 5 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#6. Distal last exon spliced from penultimate exon (i.e. a mutually exclusive last exon) (e.g. SMC1A)\n",
    "# For this to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "# Will have to annotate more precisely later (i.e. differentiate from 7)\n",
    "\n",
    "test_novel_exc_dist = {\"Chromosome\": [1]*5,\n",
    "                \"Start\": [10,10,100,200,360],\n",
    "                \"End\": [380,30,120,220,380],\n",
    "                \"Strand\": [\"+\"]*5,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*4,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*5,\n",
    "                \"transcript_id\": [\"nov_tx_exc_dist_p\"]*5,\n",
    "                       \"exon_number\": [None,1,2,3,4],\n",
    "                       \"gene_type\": [\"protein_coding\"] * 5\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_exc_dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "44d51aea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>380</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>310</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>4.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>360</td>\n",
       "      <td>380</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>5.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+------------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id    |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)         |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+------------------+---------------+----------------|\n",
       "|            1 |        10 |       380 | +            | transcript | nov_gene_1 | nov_tx_le_dist_p |           nan | protein_coding |\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |             1 | protein_coding |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |             2 | protein_coding |\n",
       "|            1 |       200 |       220 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |             3 | protein_coding |\n",
       "|            1 |       300 |       310 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |             4 | protein_coding |\n",
       "|            1 |       360 |       380 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |             5 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+------------------+---------------+----------------+\n",
       "Stranded PyRanges object has 6 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#7. Distal last exon spliced from ann\n",
    "# For this to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "test_novel_le_dist = {\"Chromosome\": [1]*6,\n",
    "                      \"Start\": [10,10,100,200,300,360],\n",
    "                      \"End\": [380,30,120,220,310,380],\n",
    "                      \"Strand\": [\"+\"]*6,\n",
    "                      \"Feature\": [\"transcript\"] + [\"exon\"]*5,\n",
    "                      \"gene_id\": [\"nov_gene_1\"]*6,\n",
    "                      \"transcript_id\": [\"nov_tx_le_dist_p\"]*6,\n",
    "                      \"exon_number\": [None,1,2,3,4,5],\n",
    "                      \"gene_type\": [\"protein_coding\"] * 6\n",
    "                     }\n",
    "\n",
    "pr.from_dict(test_novel_le_dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "8d51550f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>80</td>\n",
       "      <td>90</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+------------------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id          |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)               |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+------------------------+---------------+----------------|\n",
       "|            2 |        40 |       120 | -            | transcript | nov_gene_2 | nov_tx_le_dist_p_minus |           nan | protein_coding |\n",
       "|            2 |        40 |        50 | -            | exon       | nov_gene_2 | nov_tx_le_dist_p_minus |             1 | protein_coding |\n",
       "|            2 |        80 |        90 | -            | exon       | nov_gene_2 | nov_tx_le_dist_p_minus |             2 | protein_coding |\n",
       "|            2 |       100 |       120 | -            | exon       | nov_gene_2 | nov_tx_le_dist_p_minus |             3 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+------------------------+---------------+----------------+\n",
       "Stranded PyRanges object has 4 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#8. Distal last exon spliced from annoatrd (minus strand)\n",
    "# For this to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "test_novel_le_dist_minus = {\"Chromosome\": [2]*4,\n",
    "                      \"Start\": [40,40,80,100],\n",
    "                      \"End\": [120,50,90,120],\n",
    "                      \"Strand\": [\"-\"]*4,\n",
    "                      \"Feature\": [\"transcript\"] + [\"exon\"]*3,\n",
    "                      \"gene_id\": [\"nov_gene_2\"]*4,\n",
    "                      \"transcript_id\": [\"nov_tx_le_dist_p_minus\"]*4,\n",
    "                            \"exon_number\": [None,1,2,3],\n",
    "                            \"gene_type\": [\"protein_coding\"] * 4\n",
    "                     }\n",
    "\n",
    "pr.from_dict(test_novel_le_dist_minus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "37d62a2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>350</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>novel_tx_3utr_ext_p_plus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>novel_tx_3utr_ext_p_plus</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>novel_tx_3utr_ext_p_plus</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>novel_tx_3utr_ext_p_plus</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>350</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>novel_tx_3utr_ext_p_plus</td>\n",
       "      <td>4.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+--------------------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id            |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)                 |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+--------------------------+---------------+----------------|\n",
       "|            1 |        10 |       350 | +            | transcript | nov_gene_1 | novel_tx_3utr_ext_p_plus |           nan | protein_coding |\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | novel_tx_3utr_ext_p_plus |             1 | protein_coding |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | novel_tx_3utr_ext_p_plus |             2 | protein_coding |\n",
       "|            1 |       200 |       220 | +            | exon       | nov_gene_1 | novel_tx_3utr_ext_p_plus |             3 | protein_coding |\n",
       "|            1 |       300 |       350 | +            | exon       | nov_gene_1 | novel_tx_3utr_ext_p_plus |             4 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+--------------------------+---------------+----------------+\n",
       "Stranded PyRanges object has 5 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#9 3'UTR extension (plus & minus strand)\n",
    "# Expect a complete intron chain match, just that last exon 3'end is downstream of annotated 3'end\n",
    "#A - Plus strand\n",
    "test_novel_3utr_ext_plus = {\"Chromosome\": [1]*5,\n",
    "                \"Start\": [10,10,100,200,300],\n",
    "                \"End\": [350,30,120,220,350],\n",
    "                \"Strand\": [\"+\"]*5,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*4,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*5,\n",
    "                \"transcript_id\": [\"novel_tx_3utr_ext_p_plus\"] * 5,\n",
    "                \"exon_number\": [None,1,2,3,4],\n",
    "                           \"gene_type\": [\"protein_coding\"] * 5}\n",
    "\n",
    "pr.from_dict(test_novel_3utr_ext_plus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "4b511e38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>novel_tx_3utr_ext_minus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>20</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>novel_tx_3utr_ext_minus</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>80</td>\n",
       "      <td>90</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>novel_tx_3utr_ext_minus</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>novel_tx_3utr_ext_minus</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-------------------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id           |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)                |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-------------------------+---------------+----------------|\n",
       "|            2 |         5 |       120 | -            | transcript | nov_gene_2 | novel_tx_3utr_ext_minus |           nan | protein_coding |\n",
       "|            2 |         5 |        20 | -            | exon       | nov_gene_2 | novel_tx_3utr_ext_minus |             3 | protein_coding |\n",
       "|            2 |        80 |        90 | -            | exon       | nov_gene_2 | novel_tx_3utr_ext_minus |             2 | protein_coding |\n",
       "|            2 |       100 |       120 | -            | exon       | nov_gene_2 | novel_tx_3utr_ext_minus |             1 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-------------------------+---------------+----------------+\n",
       "Stranded PyRanges object has 4 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#B - Minus strand\n",
    "test_novel_3utr_ext_minus = {\"Chromosome\": [2]*4,\n",
    "                \"Start\": [5,5,80,100],\n",
    "                \"End\": [120,20,90,120],\n",
    "                \"Strand\": [\"-\"]*4,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*3,\n",
    "                \"gene_id\": [\"nov_gene_2\"]*4,\n",
    "                \"transcript_id\": [\"novel_tx_3utr_ext_minus\"]*4,\n",
    "                \"exon_number\": [None,3,2,1],\n",
    "                             \"gene_type\": [\"protein_coding\"] * 4\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_3utr_ext_minus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "1fb25cb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>70</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>novel_bl_p_minus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>70</td>\n",
       "      <td>90</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>novel_bl_p_minus</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>novel_bl_p_minus</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+------------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id    |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)         |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+------------------+---------------+----------------|\n",
       "|            2 |        70 |       120 | -            | transcript | nov_gene_2 | novel_bl_p_minus |           nan | protein_coding |\n",
       "|            2 |        70 |        90 | -            | exon       | nov_gene_2 | novel_bl_p_minus |             2 | protein_coding |\n",
       "|            2 |       100 |       120 | -            | exon       | nov_gene_2 | novel_bl_p_minus |             1 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+------------------+---------------+----------------+\n",
       "Stranded PyRanges object has 3 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 10. Bleedthrough but on minus strand\n",
    "# Made one b4 but as doing filtering of complete match isoforms\n",
    "# want to also cover the minus strand scenario (where look for start smaller than ref start )\n",
    "test_novel_bl_minus = {\"Chromosome\": [2]*3,\n",
    "                \"Start\": [70,70,100],\n",
    "                \"End\": [120,90,120],\n",
    "                \"Strand\": [\"-\"]*3,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*2,\n",
    "                \"gene_id\": [\"nov_gene_2\"]*3,\n",
    "                \"transcript_id\": [\"novel_bl_p_minus\"]*3,\n",
    "                \"exon_number\": [None,2,1],\n",
    "                       \"gene_type\": [\"protein_coding\"] * 3\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_bl_minus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "271d7cac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>93</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>novel_fi_p_minus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>93</td>\n",
       "      <td>96</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>novel_fi_p_minus</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>novel_fi_p_minus</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+------------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id    |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)         |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+------------------+---------------+----------------|\n",
       "|            2 |        93 |       120 | -            | transcript | nov_gene_2 | novel_fi_p_minus |           nan | protein_coding |\n",
       "|            2 |        93 |        96 | -            | exon       | nov_gene_2 | novel_fi_p_minus |             2 | protein_coding |\n",
       "|            2 |       100 |       120 | -            | exon       | nov_gene_2 | novel_fi_p_minus |             1 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+------------------+---------------+----------------+\n",
       "Stranded PyRanges object has 3 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 11. first intron spliced in event on minus strand\n",
    "\n",
    "test_novel_fi_minus = {\"Chromosome\": [2]*3,\n",
    "                \"Start\": [93,93,100],\n",
    "                \"End\": [120,96,120],\n",
    "                \"Strand\": [\"-\"]*3,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*2,\n",
    "                \"gene_id\": [\"nov_gene_2\"]*3,\n",
    "                \"transcript_id\": [\"novel_fi_p_minus\"]*3,\n",
    "                \"exon_number\": [None,2,1],\n",
    "                       \"gene_type\": [\"protein_coding\"] * 3\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_fi_minus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "522586e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>280</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>novel_tx_any_not_first_intron_plus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>novel_tx_any_not_first_intron_plus</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>novel_tx_any_not_first_intron_plus</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>260</td>\n",
       "      <td>280</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>novel_tx_any_not_first_intron_plus</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+------------------------------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id                      |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)                           |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+------------------------------------+---------------+----------------|\n",
       "|            1 |       100 |       280 | +            | transcript | nov_gene_1 | novel_tx_any_not_first_intron_plus |           nan | protein_coding |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | novel_tx_any_not_first_intron_plus |             1 | protein_coding |\n",
       "|            1 |       200 |       220 | +            | exon       | nov_gene_1 | novel_tx_any_not_first_intron_plus |             2 | protein_coding |\n",
       "|            1 |       260 |       280 | +            | exon       | nov_gene_1 | novel_tx_any_not_first_intron_plus |             3 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+------------------------------------+---------------+----------------+\n",
       "Stranded PyRanges object has 4 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 12. 'any' complete match, but matching reference intron for novel's first intron is not a first intron \n",
    "test_novel_any_not_fi_plus = {\"Chromosome\": [1]*4,\n",
    "                              \"Start\": [100,100,200,260],\n",
    "                              \"End\": [280,120,220,280],\n",
    "                              \"Strand\": [\"+\"]*4,\n",
    "                              \"Feature\": [\"transcript\"] + [\"exon\"]*3,\n",
    "                              \"gene_id\": [\"nov_gene_1\"]*4,\n",
    "                              \"transcript_id\": [\"novel_tx_any_not_first_intron_plus\"] * 4,\n",
    "                              \"exon_number\": [None,1,2,3],\n",
    "                              \"gene_type\": [\"protein_coding\"] * 4\n",
    "                             }\n",
    "\n",
    "\n",
    "pr.from_dict(test_novel_any_not_fi_plus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "d4a97f62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>90</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>novel_tx_any_not_first_intron_minus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>novel_tx_any_not_first_intron_minus</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>novel_tx_any_not_first_intron_minus</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>80</td>\n",
       "      <td>90</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>novel_tx_any_not_first_intron_minus</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-------------------------------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id                       |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)                            |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-------------------------------------+---------------+----------------|\n",
       "|            2 |         4 |        90 | -            | transcript | nov_gene_2 | novel_tx_any_not_first_intron_minus |           nan | protein_coding |\n",
       "|            2 |         4 |         6 | -            | exon       | nov_gene_2 | novel_tx_any_not_first_intron_minus |             3 | protein_coding |\n",
       "|            2 |        10 |        20 | -            | exon       | nov_gene_2 | novel_tx_any_not_first_intron_minus |             2 | protein_coding |\n",
       "|            2 |        80 |        90 | -            | exon       | nov_gene_2 | novel_tx_any_not_first_intron_minus |             1 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-------------------------------------+---------------+----------------+\n",
       "Stranded PyRanges object has 4 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 12B - minus\n",
    "test_novel_any_not_fi_minus =  {\"Chromosome\": [2]*4,\n",
    "                \"Start\": [4,4,10,80],\n",
    "                \"End\": [90,6,20,90],\n",
    "                \"Strand\": [\"-\"]*4,\n",
    "                \"Feature\": [\"transcript\"] + [\"exon\"]*3,\n",
    "                \"gene_id\": [\"nov_gene_2\"]*4,\n",
    "                \"transcript_id\": [\"novel_tx_any_not_first_intron_minus\"]*4,\n",
    "                \"exon_number\": [None,3,2,1],\n",
    "                                \"gene_type\": [\"protein_coding\"] * 4\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_any_not_fi_minus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "9a4e0682",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript</td>\n",
       "      <td>novel_gene_1</td>\n",
       "      <td>reass_ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>novel_gene_1</td>\n",
       "      <td>reass_ref_tr_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>novel_gene_1</td>\n",
       "      <td>reass_ref_tr_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>novel_gene_1</td>\n",
       "      <td>reass_ref_tr_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>novel_gene_1</td>\n",
       "      <td>reass_ref_tr_1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+--------------+-----------------+---------------+----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id      | transcript_id   |   exon_number | gene_type      |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)     | (object)        |     (float64) | (object)       |\n",
       "|--------------+-----------+-----------+--------------+------------+--------------+-----------------+---------------+----------------|\n",
       "|            1 |        10 |       340 | +            | transcript | novel_gene_1 | reass_ref_tr_1  |           nan | protein_coding |\n",
       "|            1 |        10 |        30 | +            | exon       | novel_gene_1 | reass_ref_tr_1  |             1 | protein_coding |\n",
       "|            1 |       100 |       120 | +            | exon       | novel_gene_1 | reass_ref_tr_1  |             2 | protein_coding |\n",
       "|            1 |       200 |       220 | +            | exon       | novel_gene_1 | reass_ref_tr_1  |             3 | protein_coding |\n",
       "|            1 |       300 |       340 | +            | exon       | novel_gene_1 | reass_ref_tr_1  |             4 | protein_coding |\n",
       "+--------------+-----------+-----------+--------------+------------+--------------+-----------------+---------------+----------------+\n",
       "Stranded PyRanges object has 5 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_reass_ref_tr = {\"Chromosome\": [1]*5,\n",
    "                \"Start\": [10,10,100,200,300],\n",
    "                \"End\": [340,30,120,220,340],\n",
    "                \"Strand\": [\"+\"]*5,\n",
    "                \"Feature\": [\"transcript\"] +[\"exon\"]*4,\n",
    "                \"gene_id\": [\"novel_gene_1\"]*5,\n",
    "                \"transcript_id\": [\"reass_ref_tr_1\"] * 5,\n",
    "                \"exon_number\": [None,1,2,3,4],\n",
    "                \"gene_type\": [\"protein_coding\"] * 5}\n",
    "\n",
    "pr.from_dict(test_reass_ref_tr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "4dc1c09e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     transcript\n",
       "1           exon\n",
       "2           exon\n",
       "3           exon\n",
       "4           exon\n",
       "         ...    \n",
       "13          exon\n",
       "14    transcript\n",
       "15          exon\n",
       "16          exon\n",
       "17          exon\n",
       "Name: Feature, Length: 76, dtype: object"
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_novel_gr = pr.concat([pr.from_dict(event) for event in [test_novel_3ui,\n",
    "                                                            test_novel_bl,\n",
    "                                                            test_novel_exc_dist,\n",
    "                                                            test_novel_fi,\n",
    "                                                            test_novel_le_dist,\n",
    "                                                            test_novel_mult,\n",
    "                                                            test_novel_si,\n",
    "                                                            test_novel_le_dist_minus,\n",
    "                                                             test_novel_3utr_ext_plus,\n",
    "                                                             test_novel_3utr_ext_minus,\n",
    "                                                             test_novel_bl_minus,\n",
    "                                                             test_novel_fi_minus,\n",
    "                                                             test_novel_any_not_fi_plus,\n",
    "                                                             test_novel_any_not_fi_minus,\n",
    "                                                             test_reass_ref_tr\n",
    "                                                            ]\n",
    "                          ]\n",
    "                         )\n",
    "\n",
    "test_novel_gr.Feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "12c9aece",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_novel_gr.to_gtf(\"../tests/novel.test_intron_chain_matching.transcripts.gtf\")\n",
    "test_ref.to_gtf(\"../tests/reference.test_intron_chain_matching.transcripts.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6e2ae531",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>220</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>310</td>\n",
       "      <td>330</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>70</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>220</td>\n",
       "      <td>360</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>35</td>\n",
       "      <td>50</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>50</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>220</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>310</td>\n",
       "      <td>360</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>130</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>150</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>70</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>140</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>140</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>2</td>\n",
       "      <td>50</td>\n",
       "      <td>80</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>2</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+------------+-----------------+\n",
       "| Chromosome   | Start     | End       | Strand     | Feature    | gene_id    | transcript_id   |\n",
       "| (object)     | (int64)   | (int64)   | (object)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+------------+------------+------------+-----------------|\n",
       "| 1            | 30        | 100       | +          | intron     | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| 1            | 120       | 200       | +          | intron     | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| 1            | 220       | 300       | +          | intron     | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| 1            | 310       | 330       | +          | intron     | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...        | ...             |\n",
       "| 1            | 70        | 100       | +          | intron     | nov_gene_1 | nov_tx_si_f     |\n",
       "| 1            | 120       | 140       | +          | intron     | nov_gene_1 | nov_tx_si_f     |\n",
       "| 1            | 30        | 100       | +          | intron     | nov_gene_1 | nov_tx_si_p     |\n",
       "| 1            | 120       | 140       | +          | intron     | nov_gene_1 | nov_tx_si_p     |\n",
       "+--------------+-----------+-----------+------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 26 rows and 7 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def introns_from_df(df):\n",
    "    '''\n",
    "    '''\n",
    "    \n",
    "    n_exons = len(df)\n",
    "    \n",
    "    if n_exons < 2:\n",
    "        return None\n",
    "        #print(df)\n",
    "        #raise Exception(\"at least two exons are required for transcript to have an intron\")\n",
    "    # n exons = n-1 introns\n",
    "    \n",
    "    strand = df[\"Strand\"].drop_duplicates().tolist()[0]\n",
    "#     print(strand)\n",
    "    chrom = df[\"Chromosome\"].drop_duplicates().tolist()[0]\n",
    "    gene_id = df[\"gene_id\"].drop_duplicates().tolist()[0]\n",
    "    tx_id = df[\"transcript_id\"].drop_duplicates().tolist()[0]\n",
    "    feature = \"intron\"\n",
    "    introns = {}\n",
    "    for i in range(0, n_exons - 1):\n",
    "        if strand == \"+\":\n",
    "            intron_start = df.iloc[i, lambda x: x.columns.get_loc(\"End\")]\n",
    "            intron_end = df.iloc[i+1, lambda x: x.columns.get_loc(\"Start\")]\n",
    "            introns[str(i)] = {\"Chromosome\": chrom,\n",
    "                               \"Start\": intron_start,\n",
    "                               \"End\": intron_end,\n",
    "                               \"Strand\": strand,\n",
    "                               \"Feature\": feature,\n",
    "                               \"gene_id\": gene_id,\n",
    "                               \"transcript_id\": tx_id}\n",
    "        elif strand == \"-\":\n",
    "            intron_start = df.iloc[i, lambda x: x.columns.get_loc(\"End\")]\n",
    "            intron_end = df.iloc[i+1, lambda x: x.columns.get_loc(\"Start\")]\n",
    "            introns[str(i)] = {\"Chromosome\": chrom,\n",
    "                               \"Start\": intron_start,\n",
    "                               \"End\": intron_end,\n",
    "                               \"Strand\": strand,\n",
    "                               \"Feature\": feature,\n",
    "                               \"gene_id\": gene_id,\n",
    "                               \"transcript_id\": tx_id}\n",
    "    return pd.DataFrame.from_dict(introns, orient = \"index\")\n",
    "        \n",
    "\n",
    "\n",
    "def introns_by_tx(gr, by=\"transcript_id\", nb_cpu=1):\n",
    "    '''\n",
    "    '''\n",
    "    # Sort by position (for safety)\n",
    "    gr = gr.sort()\n",
    "    \n",
    "    return gr.apply(lambda df: df.groupby(by).apply(introns_from_df), nb_cpu=nb_cpu)\n",
    "    \n",
    "    \n",
    "test_novel_introns = introns_by_tx(test_novel_gr)\n",
    "test_ref_introns = introns_by_tx(test_ref)\n",
    "    \n",
    "    \n",
    "test_novel_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2b610fd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7c243a35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>220</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "      <td>80</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>ref_tr_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>ref_tr_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand     | Feature    | gene_id    | transcript_id   |\n",
       "|     (object) |   (int64) |   (int64) | (object)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+------------+------------+------------+-----------------|\n",
       "|            1 |        30 |       100 | +          | intron     | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       120 |       200 | +          | intron     | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       220 |       300 | +          | intron     | ref_gene_1 | ref_tr_1        |\n",
       "|            2 |        20 |        80 | -          | intron     | ref_gene_2 | ref_tr_2        |\n",
       "|            2 |        90 |       100 | -          | intron     | ref_gene_2 | ref_tr_2        |\n",
       "+--------------+-----------+-----------+------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 5 rows and 7 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_ref_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d382d27d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rle(inarray):\n",
    "        \"\"\"\n",
    "        run length encoding. Partial credit to R rle function. \n",
    "        Multi datatype arrays catered for including non Numpy\n",
    "        returns: tuple (runlengths, startpositions, values)\n",
    "        https://stackoverflow.com/questions/1066758/find-length-of-sequences-of-identical-values-in-a-numpy-array-run-length-encodi\n",
    "        Thomas Browne\n",
    "        \"\"\"\n",
    "        ia = np.asarray(inarray)                # force numpy\n",
    "        n = len(ia)\n",
    "        if n == 0: \n",
    "            return (None, None, None)\n",
    "        else:\n",
    "            y = ia[1:] != ia[:-1]               # pairwise unequal (string safe)\n",
    "            i = np.append(np.where(y), n - 1)   # must include last element posi\n",
    "            z = np.diff(np.append(-1, i))       # run lengths\n",
    "            p = np.cumsum(np.append(0, z))[:-1] # positions\n",
    "            return(z, p, ia[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32cbab88",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Dict of {novel_tx_id: {matches: [ref_id], chain_match: [n_matching], terminal_non_match: [n_not_matching]}}\n",
    "novel_info_dict = {}\n",
    "\n",
    "test_novel_gr.apply(lambda df: df.groupby(\"transcript_id\").filter(lambda grp: len(grp) > 2))\n",
    "\n",
    "for key, dfs in pr.itergrs([test_novel_introns.apply(lambda df: df.groupby(\"transcript_id\").filter(lambda grp: len(grp) > 1)), test_ref_introns], strand=True, keys=True):\n",
    "    print(\"----processing chrom and strand pair {0} & {1}\".format(key[0], key[1]))\n",
    "  \n",
    "    by_tx = tuple([df.groupby(\"transcript_id\") for df in dfs])\n",
    "    novel_txipts = by_tx[0]\n",
    "    ref_txipts = by_tx[1]\n",
    "    \n",
    "    strand = key[1]\n",
    "#     print(strand)\n",
    "#    #1. for each novel transcript, test it against all ref transcripts:\n",
    "    \n",
    "    \n",
    "    for novel_id, novel_introns in novel_txipts:\n",
    "#         print(novel_id)\n",
    "#         print(novel_introns)\n",
    "#         print(\"class of novel_introns is {}\".format(type(novel_introns)))\n",
    "        #print(exons)\n",
    "        \n",
    "        if strand == \"-\":\n",
    "            #First in order = last intron - reverse df now\n",
    "            novel_introns = novel_introns[::-1].reset_index(drop=True)\n",
    "#             ref_introns = ref_introns[::-1].reset_index(drop=True)\n",
    "            #print(exons)\n",
    "        #by_tx[1]\n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "        \n",
    "        if strand == \"+\":\n",
    "            first_matches = [np.array_equal(novel_introns.head(1)[[\"Start\",\"End\"]],\n",
    "                                            ref_introns.head(1)[[\"Start\",\"End\"]]) for ref_id,ref_introns in ref_txipts]\n",
    "        else:\n",
    "            first_matches = [np.array_equal(novel_introns.head(1)[[\"Start\",\"End\"]],\n",
    "                                            ref_introns[::-1].reset_index(drop=True) # rev order\n",
    "                                            .head(1)[[\"Start\",\"End\"]]\n",
    "                                           )\n",
    "                             for ref_id,ref_introns in ref_txipts]\n",
    "\n",
    "        #print(novel_id)\n",
    "#         print(first_matches)\n",
    "        if not sum(first_matches) > 0:\n",
    "            print(\"{0} does not match any reference transcripts in its first intron. Skipping\".format(novel_id))\n",
    "            continue\n",
    "        \n",
    "        # Now compare novel transcript against each ref transcript with match in first intron.\n",
    "        # \n",
    "        for ref_tr, first_match in zip(ref_txipts, first_matches):\n",
    "            \n",
    "            if not first_match:\n",
    "                continue\n",
    "            #intron_chain_match = np.equal(np.asarray(exons[[\"Start\",\"End\"]]), np.asarray(ref_tr[1][[\"Start\",\"End\"]]))\n",
    "            ref_introns = ref_tr[1]\n",
    "            \n",
    "            if strand == \"-\":\n",
    "                ref_introns = ref_introns[::-1].reset_index(drop=True) # rev order\n",
    "            \n",
    "            n_novel_introns = len(novel_introns)\n",
    "            n_ref_introns = len(ref_introns)\n",
    "            \n",
    "            # To avoid a slicing error for txipts shorter than novel\n",
    "            max_chain = min(n_novel_introns, n_ref_introns)\n",
    "#             print(max_chain)\n",
    "#             print(ref_tr[1].iloc[0,])\n",
    "#             print(n_novel_introns)\n",
    "#             print(len(ref_tr[1]))\n",
    "            novel_chain_match = pd.DataFrame([(novel_introns.iloc[i,:][[\"Start\",\"End\"]]\n",
    "                                               .eq(ref_introns.iloc[i,:][[\"Start\",\"End\"]]\n",
    "                                                  )\n",
    "                                              )\n",
    "                                              for i in range(max_chain)])\n",
    "    \n",
    "            # Collapse to single True/False per row - does intron match?\n",
    "            novel_chain_match = novel_chain_match.apply(np.all, axis=1, raw=True)\n",
    "            \n",
    "#             print(novel_chain_match)\n",
    "            \n",
    "            runs, starts, vals = rle(novel_chain_match)\n",
    "            #print(runs[0])\n",
    "            #print(starts)\n",
    "            #print(\"\\n {0}\".format(np.where(vals == False)))\n",
    "            \n",
    "            if np.all(vals) or np.array_equal(vals, [True,False]):\n",
    "                # Possible genuine match, should update dict with info\n",
    "                \n",
    "                # match (from start of ref Txipt)\n",
    "                \n",
    "                if vals.size == 1:\n",
    "                    #i.e. all true/introns match (e.g. bleedthrough/ where )\n",
    "                    terminal_non_match = 0\n",
    "                else:\n",
    "                    terminal_non_match = runs[1]\n",
    "                \n",
    "#                 print(terminal_non_match)\n",
    "                \n",
    "                if novel_id not in novel_info_dict:\n",
    "                    \n",
    "                    novel_info_dict[novel_id] = {\"matches\": [ref_tr[0]],\n",
    "                                                 \"chain_match\": [runs[0]], #Always starts with true, so take length of true\n",
    "                                                 \"terminal_non_match\": [terminal_non_match],\n",
    "                                                }\n",
    "                \n",
    "                else:\n",
    "                    # Append to dict\n",
    "                    novel_info_dict[novel_id][\"matches\"].append(ref_tr[0])\n",
    "                    novel_info_dict[novel_id][\"chain_match\"].append(ref_tr[0])\n",
    "                    novel_info_dict[novel_id][\"terminal_non_match\"].append(ref_tr[0])\n",
    "                    \n",
    "                \n",
    "            else:\n",
    "                continue\n",
    "            \n",
    "            \n",
    "            \n",
    "print(novel_info_dict)\n",
    "#             intron_chain_match = exons[[\"Start\",\"End\"]].eq(ref_tr[1][[\"Start\",\"End\"]])\n",
    "#             print(\"\\n\")\n",
    "#             print(intron_chain_match)\n",
    "#             print(intron_chain_match.apply(np.all, axis = 1))\n",
    "            \n",
    "            \n",
    "                #loc[np.where(lambda x: (x[[\"Start\", \"End\"]].all()))])\n",
    "                                                  #apply(lambda x: all(x), axis = \"index\", raw = True))\n",
    "#             print(intron_chain_match)\n",
    "            \n",
    "            \n",
    "#         first_intron = exons.head(1)[[\"Start\",\"End\"]] if strand == \"+\" else exons.tail(1)[[\"Start\",\"End\"]]\n",
    "        \n",
    "#         grp_filter_matches = by_tx[1].filter(lambda x: np.array_equal(first_intron, x.head(1)[[\"Start\", \"End\"]]))\n",
    "#         print(grp_filter_matches)\n",
    "#         print(type(grp_filter_matches))\n",
    "     #for novel, ref_tr in zip(by_tx[0],by_tx[1]:\n",
    "#         print(nov_tr[1])\n",
    "#         print(ref_tr)\n",
    "             \n",
    "pd.concat({key: pd.DataFrame.from_dict(d, orient = \"columns\") for key, d in novel_info_dict.items()}, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54913d65",
   "metadata": {},
   "outputs": [],
   "source": [
    "def match_intron_chains(novel_gr, ref_gr, id_col = \"transcript_id\", nb_cpu = 1):\n",
    "    '''\n",
    "    grs should contain introns\n",
    "    '''\n",
    "    # {novel_id: {matches: [ref_id], chain_match: [n], terminal_non_match: [n]}}\n",
    "    match_info_dict = {}\n",
    "    \n",
    "\n",
    "    \n",
    "    for key, dfs in pr.itergrs([novel_gr, ref_gr], strand=True, keys=True):\n",
    "        print(\"----processing chrom and strand pair {0} & {1}\".format(key[0], key[1]))\n",
    "        # dfs = novel_gr & ref_gr matched by key (chromosome & strand)\n",
    "        # pandas group by objects of (transcript_id, df)\n",
    "        by_tx = tuple([df.groupby(\"transcript_id\") for df in dfs])\n",
    "        \n",
    "        novel_txipts = by_tx[0]\n",
    "        ref_txipts = by_tx[1]\n",
    "        \n",
    "        # Pyranges keys are tuples of (chr,strand)\n",
    "        strand = key[1]\n",
    "        \n",
    "        \n",
    "        #Comparing each novel transcript against all ref transcripts\n",
    "        for novel_id, novel_introns in novel_txipts:\n",
    "        \n",
    "            if strand == \"-\":\n",
    "                # Standard PyRanges sort - First in df = last intron (smallest values i.e. leftmost)\n",
    "                # Reverse so first row in df is always the first intron\n",
    "                # is reset_index necessary?\n",
    "                novel_introns = novel_introns[::-1].reset_index(drop=True)\n",
    "            \n",
    "            else:\n",
    "                pass\n",
    "        \n",
    "            # As a first pass, check for matches in intron chain of first introns between novel and reference Txs\n",
    "            if strand == \"+\":\n",
    "                first_matches = [np.array_equal(novel_introns.head(1)[[\"Start\",\"End\"]],\n",
    "                                                ref_introns.head(1)[[\"Start\",\"End\"]]) \n",
    "                                 for ref_id,ref_introns in ref_txipts]\n",
    "            else:\n",
    "                first_matches = [np.array_equal(novel_introns.head(1)[[\"Start\",\"End\"]],\n",
    "                                                ref_introns[::-1].reset_index(drop=True) # rev order so 1st row = 1st intron\n",
    "                                                .head(1)[[\"Start\",\"End\"]]\n",
    "                                               )\n",
    "                                 for ref_id,ref_introns in ref_txipts]\n",
    "\n",
    "\n",
    "            if not sum(first_matches) > 0:\n",
    "                print(\"{0} does not match any reference transcripts in its first intron. Skipping\".format(novel_id))\n",
    "                continue\n",
    "        \n",
    "            # Compare full introns chains of novel transcript against each ref transcript with match in first intron.\n",
    "        \n",
    "            for ref_tr, first_match in zip(ref_txipts, first_matches):\n",
    "            \n",
    "                if not first_match:\n",
    "                    continue\n",
    "                \n",
    "            \n",
    "                ref_id = ref_tr[0]\n",
    "#                 print(\"ref_id {}\".format(ref_id))\n",
    "                ref_introns = ref_tr[1]\n",
    "            \n",
    "                if strand == \"-\":\n",
    "                    ref_introns = ref_introns[::-1].reset_index(drop=True) # reverse so first row = first intron\n",
    "                        \n",
    "                # To avoid a slicing error for ref txipts shorter than novel\n",
    "                n_novel_introns = len(novel_introns)\n",
    "                n_ref_introns = len(ref_introns)\n",
    "                max_chain = min(n_novel_introns, n_ref_introns)\n",
    "#             print(max_chain)\n",
    "#             print(ref_tr[1].iloc[0,])\n",
    "#             print(n_novel_introns)\n",
    "#             print(len(ref_tr[1]))\n",
    "\n",
    "                # Row-wise, check whether match with corresponding intron of reference transcript\n",
    "                novel_chain_match = pd.DataFrame([(novel_introns.iloc[i,:][[\"Start\",\"End\"]]\n",
    "                                                   .eq(ref_introns.iloc[i,:][[\"Start\",\"End\"]]\n",
    "                                                      )\n",
    "                                                  )\n",
    "                                                  for i in range(max_chain)])\n",
    "    \n",
    "                # Collapse to single True/False per row - does intron completely match?\n",
    "                novel_chain_match = novel_chain_match.apply(np.all, axis=1, raw=True)\n",
    "            \n",
    "#             print(novel_chain_match)\n",
    "            \n",
    "                runs, starts, vals = rle(novel_chain_match)\n",
    "            #print(runs[0])\n",
    "            #print(starts)\n",
    "            #print(\"\\n {0}\".format(np.where(vals == False)))\n",
    "            \n",
    "                # Considered a valid match if a intron chain completely identical \n",
    "                # or matches at beginning of txipt but differs at the end\n",
    "                if np.all(vals) or np.array_equal(vals, [True,False]):\n",
    "                \n",
    "                    # Don't want to throw away all valid matches (yet)\n",
    "                    # Possible genuine match, should update dict with info\n",
    "                    # match (from start of ref Txipt)\n",
    "                \n",
    "                    if vals.size == 1:\n",
    "                        #i.e. all true/introns match (e.g. bleedthrough)\n",
    "                        terminal_non_match = 0\n",
    "                    else:\n",
    "                        # All valid = consective match & non-match, so non-match = 2nd in array \n",
    "                        terminal_non_match = runs[1]\n",
    "                \n",
    "#                 print(terminal_non_match)\n",
    "                \n",
    "                    if novel_id not in match_info_dict:\n",
    "                    \n",
    "                        match_info_dict[novel_id] = {\"matches\": [ref_id],\n",
    "                                                     \"chain_match\": [runs[0]], #Always starts with true, so take length of true\n",
    "                                                     \"terminal_non_match\": [terminal_non_match],\n",
    "                                                    }\n",
    "                \n",
    "                    else:\n",
    "                        # Append to dict\n",
    "                        match_info_dict[novel_id][\"matches\"].append(ref_id)\n",
    "                        match_info_dict[novel_id][\"chain_match\"].append(runs[0])\n",
    "                        match_info_dict[novel_id][\"terminal_non_match\"].append(terminal_non_match)\n",
    "                    \n",
    "                \n",
    "                else:\n",
    "                    continue\n",
    "    \n",
    "    # Output df for easier parsing\n",
    "#     match_df = pd.concat({key: pd.DataFrame.from_dict(d, orient = \"columns\") for key, d in match_info_dict.items()}, axis=0)\n",
    "    \n",
    "    return match_info_dict\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca48baec",
   "metadata": {},
   "outputs": [],
   "source": [
    "match_intron_chains(test_novel_introns, test_ref_introns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf33eff0",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.array_equal(np.array([True, False]), [True, False])\n",
    "np.all([True,True])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb031093",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test_introns = pr.from_dict(test_novel_3ui)#.subset(lambda df: df[\"transcript_id\"] == \"nov_tx_3ui_p\")\n",
    "\n",
    "# n_exons = len(test_introns)\n",
    "# # n exons = n -1 introns\n",
    "\n",
    "# # plus strand txipt - End of first exon = start of intron, Start of 3'exon = end coord\n",
    "\n",
    "# test_introns.as_df().iloc[0,lambda x: x.columns.get_loc(\"End\")]\n",
    "\n",
    "# def introns_from_df(df):\n",
    "#     '''\n",
    "#     '''\n",
    "    \n",
    "#     n_exons = len(df)\n",
    "    \n",
    "#     if n_exons < 2:\n",
    "#         raise Exception(\"at least two exons are required for transcript to have an intron\")\n",
    "#     # n exons = n-1 introns\n",
    "    \n",
    "#     strand = df[\"Strand\"].drop_duplicates().tolist()[0]\n",
    "# #     print(strand)\n",
    "#     chrom = df[\"Chromosome\"].drop_duplicates().tolist()[0]\n",
    "#     gene_id = df[\"gene_id\"].drop_duplicates().tolist()[0]\n",
    "#     tx_id = df[\"transcript_id\"].drop_duplicates().tolist()[0]\n",
    "#     feature = \"intron\"\n",
    "#     introns = {}\n",
    "#     for i in range(0, n_exons - 1):\n",
    "#         if strand == \"+\":\n",
    "#             intron_start = df.iloc[i, lambda x: x.columns.get_loc(\"End\")]\n",
    "#             intron_end = df.iloc[i+1, lambda x: x.columns.get_loc(\"Start\")]\n",
    "#             introns[str(i)] = {\"Chromosome\": chrom,\n",
    "#                                \"Start\": intron_start,\n",
    "#                                \"End\": intron_end,\n",
    "#                                \"Strand\": strand,\n",
    "#                                \"Feature\": feature,\n",
    "#                                \"gene_id\": gene_id,\n",
    "#                                \"transcript_id\": tx_id}\n",
    "#         elif strand == \"-\":\n",
    "#             intron_end = df.iloc[i, lambda x: x.columns.get_loc(\"Start\")]\n",
    "#             intron_start = df.iloc[i+1, lambda x: x.columns.get_loc(\"End\")]\n",
    "#             introns[str(i)] = {\"Chromosome\": chrom,\n",
    "#                                \"Start\": intron_start,\n",
    "#                                \"End\": intron_end,\n",
    "#                                \"Strand\": strand,\n",
    "#                                \"Feature\": feature,\n",
    "#                                \"gene_id\": gene_id,\n",
    "#                                \"transcript_id\": tx_id}\n",
    "#     return pd.DataFrame.from_dict(introns, orient = \"index\")\n",
    "        \n",
    "\n",
    "\n",
    "# def introns_by_tx(gr, by=\"transcript_id\", nb_cpu=1):\n",
    "#     '''\n",
    "#     '''\n",
    "    \n",
    "#     return gr.apply(lambda df: df.groupby(by).apply(introns_from_df), nb_cpu=1)\n",
    "    \n",
    "    \n",
    "# test_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "758737f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "introns_by_tx(test_introns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf44cbed",
   "metadata": {},
   "outputs": [],
   "source": [
    "introns_by_tx(pr.data.ensembl_gtf().subset(lambda df: (df.gene_id == \"ENSG00000205231\") & (df.Feature == \"exon\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5b71b28",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "pr.data.ensembl_gtf().subset(lambda df: df.transcript_id == \"ENST00000450305\")[[\"transcript_id\",\"Feature\"]].features.introns(by=\"transcript\").dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7e4891a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test_ref.features.introns(by=\"transcript\")\n",
    "\n",
    "(pr.data.ensembl_gtf().subset(lambda df: df.transcript_id == \"ENST00000450305\")\n",
    "#  [[\"gene_id\",\"transcript_id\"]]\n",
    " .features.introns(by=\"transcript\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c86eba9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pr.data.ensembl_gtf().subset(lambda df: (df.gene_id == \"ENSG00000205231\") & (df.Feature == \"exon\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fa672ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "pr.data.ensembl_gtf().subset(lambda df: (df.gene_id == \"ENSG00000205231\")).features.introns(by=\"transcript\")\n",
    "#Start of first feature, end of next along (if properly sorted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "8578c48f",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_stie_gtf_chr1_nov = \"../two_sample_example_output/stringtie/chr1.no_ref_id.TDP43-F_S6.assembled.gtf\"\n",
    "path_ref_gtf = \"../data/annotation/gencode.v34.annotation.gtf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "8a2eb6cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>cov</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>TPM</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>827750</td>\n",
       "      <td>859332</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>98.466469</td>\n",
       "      <td>11.679049</td>\n",
       "      <td>30.310225</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>827750</td>\n",
       "      <td>827775</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>96.734062</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>829002</td>\n",
       "      <td>829104</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>110.622597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>847653</td>\n",
       "      <td>847806</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>118.018204</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>851926</td>\n",
       "      <td>852110</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>115.267273</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26008</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248855557</td>\n",
       "      <td>248855635</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.161189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26009</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248855724</td>\n",
       "      <td>248855943</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>4.278336</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26010</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248856287</td>\n",
       "      <td>248856422</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.487612</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26011</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248856513</td>\n",
       "      <td>248856562</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.115693</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26012</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248857233</td>\n",
       "      <td>248858865</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>13.537362</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26013 rows  14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | Score     | +8    |\n",
       "| (category)   | (object)   | (category)   | (int32)   | (int32)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-----------+-------|\n",
       "| chr1         | StringTie  | transcript   | 827750    | 859332    | 1000      | ...   |\n",
       "| chr1         | StringTie  | exon         | 827750    | 827775    | 1000      | ...   |\n",
       "| chr1         | StringTie  | exon         | 829002    | 829104    | 1000      | ...   |\n",
       "| chr1         | StringTie  | exon         | 847653    | 847806    | 1000      | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...       | ...   |\n",
       "| chr1         | StringTie  | exon         | 248855724 | 248855943 | 1000      | ...   |\n",
       "| chr1         | StringTie  | exon         | 248856287 | 248856422 | 1000      | ...   |\n",
       "| chr1         | StringTie  | exon         | 248856513 | 248856562 | 1000      | ...   |\n",
       "| chr1         | StringTie  | exon         | 248857233 | 248858865 | 1000      | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 26,013 rows and 14 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "8 hidden columns: Strand, Frame, gene_id, transcript_id, cov, FPKM, TPM, exon_number"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stie_chr1 = pr.read_gtf(path_stie_gtf_chr1_nov)\n",
    "stie_chr1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c443d1a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2912491</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214349</td>\n",
       "      <td>57214397</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002072208.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2912492</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213879</td>\n",
       "      <td>57213964</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002046926.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2912493</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213525</td>\n",
       "      <td>57213602</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002021169.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2912494</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213203</td>\n",
       "      <td>57213357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00002036959.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2912495</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57213125</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00002023900.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2912496 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------------------------+------------------------------------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | Score      | Strand       | Frame      | gene_id                 | gene_type                          | +15   |\n",
       "| (category)   | (object)   | (object)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)                | (object)                           | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------------------------+------------------------------------+-------|\n",
       "| chr1         | HAVANA     | gene       | 11868     | 14409     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| chr1         | HAVANA     | transcript | 11868     | 14409     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| chr1         | HAVANA     | exon       | 11868     | 12227     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| chr1         | HAVANA     | exon       | 12612     | 12721     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...        | ...          | ...        | ...                     | ...                                | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57213879  | 57213964  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57213525  | 57213602  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57213203  | 57213357  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57212183  | 57213125  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------------------------+------------------------------------+-------+\n",
       "Stranded PyRanges object has 2,912,496 rows and 25 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "15 hidden columns: gene_name, level, hgnc_id, havana_gene, transcript_id, transcript_type, transcript_name, transcript_support_level, tag, havana_transcript, exon_number, exon_id, ont, ... (+ 2 more.)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref = pr.read_gtf(path_ref_gtf)\n",
    "ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4f516593",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>cov</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>TPM</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>827750</td>\n",
       "      <td>827775</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>96.734062</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>829002</td>\n",
       "      <td>829104</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>110.622597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>847653</td>\n",
       "      <td>847806</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>118.018204</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>851926</td>\n",
       "      <td>852110</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>115.267273</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>852670</td>\n",
       "      <td>852766</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>135.545364</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23563</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248855557</td>\n",
       "      <td>248855635</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.161189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23564</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248855724</td>\n",
       "      <td>248855943</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>4.278336</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23565</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248856287</td>\n",
       "      <td>248856422</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.487612</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23566</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248856513</td>\n",
       "      <td>248856562</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.115693</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23567</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248857233</td>\n",
       "      <td>248858865</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>13.537362</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23568 rows  14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | Score     | Strand       | Frame      | gene_id    | transcript_id   | cov        | FPKM       | TPM        | +1    |\n",
       "| (category)   | (object)   | (category)   | (int32)   | (int32)   | (int64)   | (category)   | (object)   | (object)   | (object)        | (object)   | (object)   | (object)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+-------|\n",
       "| chr1         | StringTie  | exon         | 827750    | 827775    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 96.734062  | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 829002    | 829104    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 110.622597 | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 847653    | 847806    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 118.018204 | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 851926    | 852110    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 115.267273 | nan        | nan        | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...       | ...          | ...        | ...        | ...             | ...        | ...        | ...        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248855724 | 248855943 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 4.278336   | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248856287 | 248856422 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 5.487612   | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248856513 | 248856562 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 5.115693   | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248857233 | 248858865 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 13.537362  | nan        | nan        | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+-------+\n",
       "Stranded PyRanges object has 23,568 rows and 14 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: exon_number"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stie_chr1_exons = stie_chr1.subset(lambda df: df[\"Feature\"] == \"exon\")\n",
    "stie_chr1_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "23078b88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000002844.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000002844.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378883</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214349</td>\n",
       "      <td>57214397</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002072208.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378884</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213879</td>\n",
       "      <td>57213964</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002046926.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378885</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213525</td>\n",
       "      <td>57213602</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002021169.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378886</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213203</td>\n",
       "      <td>57213357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00002036959.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378887</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57213125</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00002023900.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1378888 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------------------------+------------------------------------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | Score      | Strand       | Frame      | gene_id                 | gene_type                          | +15   |\n",
       "| (category)   | (object)   | (object)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)                | (object)                           | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------------------------+------------------------------------+-------|\n",
       "| chr1         | HAVANA     | exon       | 11868     | 12227     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| chr1         | HAVANA     | exon       | 12612     | 12721     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| chr1         | HAVANA     | exon       | 13220     | 14409     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| chr1         | HAVANA     | exon       | 12009     | 12057     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...        | ...          | ...        | ...                     | ...                                | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57213879  | 57213964  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57213525  | 57213602  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57213203  | 57213357  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57212183  | 57213125  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------------------------+------------------------------------+-------+\n",
       "Stranded PyRanges object has 1,378,888 rows and 25 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "15 hidden columns: gene_name, level, hgnc_id, havana_gene, transcript_id, transcript_type, transcript_name, transcript_support_level, tag, havana_transcript, exon_number, exon_id, ont, ... (+ 2 more.)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_exons = ref.subset(lambda df: df[\"Feature\"] == \"exon\")\n",
    "ref_chr1_exons = ref_exons.subset(lambda df: df[\"Chromosome\"] == \"chr1\")\n",
    "ref_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "44b602ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>29553</td>\n",
       "      <td>30039</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000002840.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001947070.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>30563</td>\n",
       "      <td>30667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000002840.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001922571.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>30975</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000002840.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001827679.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>30266</td>\n",
       "      <td>30667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-201</td>\n",
       "      <td>5</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000002841.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001841699.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>30975</td>\n",
       "      <td>31109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-201</td>\n",
       "      <td>5</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000002841.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001890064.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121907</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248857668</td>\n",
       "      <td>248858017</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000171163.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>ZNF692-218</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000097304.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001904444.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121908</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248859014</td>\n",
       "      <td>248859144</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000171163.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>ZNF692-229</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000382605.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002142255.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121909</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248857864</td>\n",
       "      <td>248858321</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000171163.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>ZNF692-229</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000382605.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002187693.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121910</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248858917</td>\n",
       "      <td>248859033</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000171163.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>ZNF692-228</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000382606.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002188413.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121911</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248857974</td>\n",
       "      <td>248858385</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000171163.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>ZNF692-228</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000382606.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002170424.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121912 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | Score      | Strand       | Frame      | gene_id            | gene_type      | gene_name   | level      | +13   |\n",
       "| (category)   | (object)   | (object)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)           | (object)       | (object)    | (object)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+-------|\n",
       "| chr1         | HAVANA     | exon       | 29553     | 30039     | .          | +            | .          | ENSG00000243485.5  | lncRNA         | MIR1302-2HG | 2          | ...   |\n",
       "| chr1         | HAVANA     | exon       | 30563     | 30667     | .          | +            | .          | ENSG00000243485.5  | lncRNA         | MIR1302-2HG | 2          | ...   |\n",
       "| chr1         | HAVANA     | exon       | 30975     | 31097     | .          | +            | .          | ENSG00000243485.5  | lncRNA         | MIR1302-2HG | 2          | ...   |\n",
       "| chr1         | HAVANA     | exon       | 30266     | 30667     | .          | +            | .          | ENSG00000243485.5  | lncRNA         | MIR1302-2HG | 2          | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...        | ...          | ...        | ...                | ...            | ...         | ...        | ...   |\n",
       "| chr1         | HAVANA     | exon       | 248859014 | 248859144 | .          | -            | .          | ENSG00000171163.16 | protein_coding | ZNF692      | 2          | ...   |\n",
       "| chr1         | HAVANA     | exon       | 248857864 | 248858321 | .          | -            | .          | ENSG00000171163.16 | protein_coding | ZNF692      | 2          | ...   |\n",
       "| chr1         | HAVANA     | exon       | 248858917 | 248859033 | .          | -            | .          | ENSG00000171163.16 | protein_coding | ZNF692      | 2          | ...   |\n",
       "| chr1         | HAVANA     | exon       | 248857974 | 248858385 | .          | -            | .          | ENSG00000171163.16 | protein_coding | ZNF692      | 2          | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+-------+\n",
       "Stranded PyRanges object has 121,912 rows and 25 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "13 hidden columns: hgnc_id, havana_gene, transcript_id, transcript_type, transcript_name, transcript_support_level, tag, havana_transcript, exon_number, exon_id, ont, protein_id, ccdsid"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#prot coding / lncRNA\n",
    "# print(ref_chr1_exons.gene_type.value_counts())\n",
    "ref_chr1_pc_exons = ref_chr1_exons.subset(lambda df: df.gene_type.isin([\"protein_coding\", \"lncRNA\"]))\n",
    "ref_chr1_pc_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46be9ffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# stie_chr1_introns = introns_by_tx(stie_chr1_exons,nb_cpu=1)\n",
    "# stie_chr1_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f8edcbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ref_all_introns = introns_by_tx(ref_exons, nb_cpu=4)\n",
    "# ref_all_introns\n",
    "# ref_chr1_introns = introns_by_tx(ref_chr1_pc_exons, nb_cpu=2)\n",
    "# ref_chr1_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29d16f4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr1_chain_matching = match_intron_chains(stie_chr1_introns, ref_chr1_introns)\n",
    "# chr1_chain_matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c6c96ea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9d6e3ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ref_chr1_introns.subset(lambda df: df.transcript_id == \"ENST00000674495.1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a19a84a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ref_chr1_exons.subset(lambda df: df.transcript_id == \"ENST00000674495.1\").sort()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07b55de0",
   "metadata": {},
   "source": [
    "### Alternative (hopefully more scalable approach)\n",
    "\n",
    "Try and use pyranges internals as much as possible, and avoid manual looping and comparisons\n",
    "\n",
    "Essentially, Find overlapping introns, then filter for those that are identical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "02c03397",
   "metadata": {},
   "outputs": [],
   "source": [
    "def intron_id(gr):\n",
    "\n",
    "    return gr.assign(\"intron_id\",\n",
    "                                         lambda df: pd.Series([\":\".join([tx_id, str(start), str(end)])\n",
    "                                                     for tx_id, start, end in zip(df[\"transcript_id\"],\n",
    "                                                                                  df[\"Start\"],\n",
    "                                                                                  df[\"End\"])\n",
    "                                                    ])\n",
    "                                        )\n",
    "\n",
    "def sort_introns_by_strand(df):\n",
    "    '''\n",
    "    '''\n",
    "    # first reset_index call removes the original index of the group (e.g. row 4005 in df)\n",
    "    # second reset_index call adds the sorted index as a column to the dataframe (the order along exon in each transcript)\n",
    "    if (df.Strand == '+').all():\n",
    "        return df.sort_values(by=['End']).reset_index(drop=True).reset_index()\n",
    "    elif (df.Strand == '-').all():\n",
    "        return df.sort_values(by=['Start'], ascending=False).reset_index(drop=True).reset_index()\n",
    "\n",
    "\n",
    "    \n",
    "def validate_matching_chain(df, max_terminal_non_match=2):\n",
    "    '''\n",
    "    apply to grouped df\n",
    "    '''\n",
    "    \n",
    "    runs, starts, vals = rle(df[\"match\"])\n",
    "    \n",
    "    # Valid matches are:\n",
    "    # All introns match (e.g. bleedthrough event)\n",
    "    # All but last x introns match (usually one) (i.e. runs = 1,0)\n",
    "    \n",
    "    if np.all(vals):\n",
    "        # All introns match (e.g. bleedthrough event)\n",
    "        return True\n",
    "    \n",
    "    elif np.array_equal(vals, [1,0]) and runs[-1] <= max_terminal_non_match:\n",
    "        # all but last x introns match (x = max_terminal_non_match) (i.e. runs = 1,0)\n",
    "        return True\n",
    "    \n",
    "    else:\n",
    "        return False\n",
    "\n",
    "    \n",
    "\n",
    "def stie_groupby_last_exon(df, exon_n_col, which=\"last\"):\n",
    "    '''\n",
    "    '''\n",
    "    \n",
    "    if (df[\"Strand\"] == \"+\").all():\n",
    "        if which == \"last\":\n",
    "            return df[exon_n_col].idxmax()\n",
    "        elif which == \"first\":\n",
    "            return df[exon_n_col].idxmin()\n",
    "    \n",
    "    elif (df[\"Strand\"] == \"-\").all():\n",
    "        if which == \"last\":\n",
    "            return df[exon_n_col].idxmin()\n",
    "        elif which == \"first\":\n",
    "            return df[exon_n_col].idxmax()\n",
    "\n",
    "        \n",
    "def filter_multi_exon(df, exon_n_col):\n",
    "    '''\n",
    "    Want transcripts with > 1 exon\n",
    "    '''\n",
    "    if df[exon_n_col].nunique() > 1:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "    \n",
    "def get_terminal_exons(gr,\n",
    "                   feature_col = \"Feature\",\n",
    "                   id_col = \"transcript_id\",\n",
    "                   exon_number_col = \"exon_number\",\n",
    "                   source = None,\n",
    "                   which_exon=\"last\",\n",
    "                   filter_single = False,\n",
    "                   nb_cpu = 1):\n",
    "    '''\n",
    "    Return gr of last exons for each transcript_id\n",
    "    In process, exon_number_col will be converted to type 'int'\n",
    "    StringTie merged GTFs (or whatever tool single_steps/stringtie_longreads.smk is using) \n",
    "    reports exon_number that DOES NOT RESPECT STRAND (from browsing in IGV)\n",
    "    i.e. for minus-strand - largest exon_number for transcript corresponds to FIRST EXON, not last\n",
    "    Annotated (i.e. Ensembl) reported exon_numbers DO RESPECT STRAND (i.e. max always = last exon)\n",
    "    \n",
    "    if Do respect strand, put source = None (default)\n",
    "    if Don't respect strand, put source = \"stringtie\" (i.e. plus strand = max, minus strand = min)\n",
    "    '''\n",
    "    # Pull out exons, convert exon_number to int\n",
    "    mod_gr = (gr.assign(exon_number_col,\n",
    "                      lambda df: df[exon_number_col].astype(int),\n",
    "                      nb_cpu = nb_cpu)\n",
    "             )\n",
    "    \n",
    "    \n",
    "    # Filter out single-exon transcripts\n",
    "    if filter_single:\n",
    "        print(\"Filtering for multi-exon transcripts...\")\n",
    "        print(\"Before: {}\".format(len(set(mod_gr.as_df()[id_col].tolist()))))\n",
    "        \n",
    "        mod_gr = (mod_gr.apply(lambda df: (df.groupby(id_col)\n",
    "                                       .filter(lambda x: filter_multi_exon(df, exon_number_col))\n",
    "                                      )\n",
    "                           ,\n",
    "                           nb_cpu= nb_cpu\n",
    "                          )\n",
    "                 )\n",
    "        print(\"After: {}\".format(len(set(mod_gr.as_df()[id_col].tolist()))))\n",
    "    \n",
    "    \n",
    "    \n",
    "    # Pick last exon entry by max exon number for each transcript (id_col)\n",
    "    if source is None:\n",
    "        \n",
    "        if which_exon == \"last\":\n",
    "            out_gr = mod_gr.apply(lambda df: df.iloc[df.groupby(id_col)[exon_number_col].idxmax(),], nb_cpu = nb_cpu)\n",
    "            \n",
    "        elif which_exon == \"first\":\n",
    "            out_gr = mod_gr.apply(lambda df: df.iloc[df.groupby(id_col)[exon_number_col].idxmin(),], nb_cpu = nb_cpu)\n",
    "    \n",
    "    elif source == \"stringtie\":\n",
    "        # Doesn't respect strand - pick min if Minus strand, max if plus strand\n",
    "        out_gr = (mod_gr.apply(lambda df: df.iloc[(df.groupby(id_col)\n",
    "                                                  .apply(lambda df: stie_groupby_last_exon(df, exon_number_col, which_exon)\n",
    "                                                        )\n",
    "                                                 ),],\n",
    "                               nb_cpu = nb_cpu\n",
    "                              ) \n",
    "                 )\n",
    "        \n",
    "\n",
    "        \n",
    "    else:\n",
    "        pass\n",
    "    \n",
    "    return out_gr\n",
    "\n",
    "    \n",
    "def filter_transcripts_by_chain(novel_exons,ref_exons, match_type = \"transcript\", max_terminal_non_match=2, nb_cpu = 1):\n",
    "    '''\n",
    "    '''\n",
    "\n",
    "    novel_cols_to_keep = [\"Feature\",\"transcript_id\"]\n",
    "    ref_cols_to_keep = [\"Feature\", \"transcript_id\", \"gene_id\", \"gene_name\"]\n",
    "\n",
    "    assert match_type in [\"transcript\", \"any\"], \"match_type must be one of 'transcript' or 'any'. value passed - {}\".format(str(match_type))\n",
    "\n",
    "    #1. Find introns by transcript & give each intron a unique ID\n",
    "    print(\"finding introns...\")\n",
    "    t1 = timer()\n",
    "    \n",
    "    novel_introns = introns_by_tx(novel_exons, nb_cpu=nb_cpu).sort()\n",
    "    ref_introns = introns_by_tx(ref_exons, nb_cpu=nb_cpu).sort()\n",
    "\n",
    "    t2 = timer()\n",
    "    \n",
    "    print(\"took {} (s)\".format(t2 - t1))\n",
    "    \n",
    "    print(\"adding intron_id column...\")\n",
    "    \n",
    "    t3 = timer()\n",
    "    novel_introns = intron_id(novel_introns)\n",
    "    ref_introns = intron_id(ref_introns)\n",
    "    t4 = timer()\n",
    "    \n",
    "    print(\"took {} s\".format(t4 - t3))\n",
    "\n",
    "    #2. Track number of introns in each novel transcript\n",
    "    novel_tx_intron_counts = (novel_introns.as_df()\n",
    "                              .groupby(\"transcript_id\").size())\n",
    "\n",
    "\n",
    "    # novel_introns, ref_introns\n",
    "\n",
    "    # 3. Store intron_ids for each transcript, sorted by intron_number (where 1 = first intron regardless of strand) in a df/Series\n",
    "    print(\"generating df of novel txipts sorted by intron number...\")\n",
    "    \n",
    "    t5 = timer()\n",
    "    \n",
    "    novel_intron_ids_ordered = (novel_introns.as_df()\n",
    "                                .groupby(\"transcript_id\")\n",
    "                                .apply(sort_introns_by_strand)\n",
    "                                .reset_index(drop=True)\n",
    "                                .rename({\"index\": \"intron_number\"}, axis=\"columns\")\n",
    "                               )\n",
    "    novel_intron_ids_ordered[\"intron_number\"] = novel_intron_ids_ordered[\"intron_number\"].add(1)\n",
    "    \n",
    "\n",
    "    # df of txipt_id | intron_id | intron_number\n",
    "    novel_intron_ids_ordered = novel_intron_ids_ordered.loc[:,[\"transcript_id\",\"intron_id\",\"intron_number\"]]\n",
    "    \n",
    "    t6 = timer()\n",
    "    print(\"took {} s\".format(t6 - t5))\n",
    "#     print(novel_intron_ids_ordered.dtypes)\n",
    "\n",
    "    ref_introns = (ref_introns.as_df()\n",
    "                                .groupby(\"transcript_id\")\n",
    "                                .apply(sort_introns_by_strand)\n",
    "                                .reset_index(drop=True)\n",
    "                                .rename({\"index\": \"intron_number\"}, axis=\"columns\")\n",
    "                               )\n",
    "    ref_introns[\"intron_number\"] = ref_introns[\"intron_number\"].add(1)\n",
    "    \n",
    "    ref_introns = pr.PyRanges(ref_introns, int64=True)\n",
    "\n",
    "\n",
    "    #4. Find novel introns with any overlap with reference introns\n",
    "    # Inner join to add ref_rows to novel gr\n",
    "    print(\"finding overlaps between novel and reference introns...\")\n",
    "    \n",
    "    t7 = timer()\n",
    "    joined = pr.PyRanges(novel_introns.as_df(), int64=True).join(ref_introns, strandedness=\"same\", suffix =\"_ref\",nb_cpu=nb_cpu)\n",
    "    t8 = timer()\n",
    "    \n",
    "    print(\"took {} s\".format(t8 - t7))\n",
    "    \n",
    "    #5. Filter for overlaps that exactly match (or differ by given tolerance)\n",
    "    print(\"filtering overlaps for exact matches...\")\n",
    "    \n",
    "    t9 = timer()\n",
    "    joined = joined.subset(lambda df: abs(df.Start - df.Start_ref) + abs(df.End - df.End_ref) <= 0, nb_cpu=nb_cpu)\n",
    "    t10 = timer()\n",
    "    \n",
    "    print(\"took {} s\".format(t10 - t9))\n",
    "    \n",
    "    # Minimal info needed on matches between novel and reference introns\n",
    "    joined = joined.as_df()[[\"transcript_id\",\"intron_id\",\"transcript_id_ref\",\"intron_id_ref\"]]\n",
    "\n",
    "#     print(joined.dtypes)\n",
    "\n",
    "    #6. Join ordered novel introns with match info\n",
    "    #7. Assign a simple tracker column 'match' of True (where intron is matched) and False (where intron is not matched)\n",
    "    \n",
    "    print(\"preparing for filtering intron matches...\")\n",
    "    t11 = timer()\n",
    "    \n",
    "    if match_type == \"any\":\n",
    "        # Looking for intron to match any annotated intron, regardless of reference transcript\n",
    "        novel_ref_match_info = novel_intron_ids_ordered.merge(joined,\n",
    "                                                              how=\"left\",\n",
    "                                                              on=\"intron_id\",\n",
    "                                                              suffixes=[\"_novel\",\"_match\"]\n",
    "                                                             )\n",
    "        \n",
    "        # Assign 'match' column for each intron.\n",
    "        # Since we don't really care which intron it matches, & no matches will mean NaN\n",
    "        novel_ref_match_info[\"match\"] = novel_ref_match_info[\"transcript_id_ref\"]\n",
    "        novel_ref_match_info[\"match\"] = novel_ref_match_info[\"match\"].fillna(0)\n",
    "        novel_ref_match_info[\"match\"] = novel_ref_match_info[\"match\"].replace(\"\\w*\", 1, regex=True)\n",
    "        \n",
    "        novel_ref_match_info = novel_ref_match_info.drop_duplicates(subset=[\"intron_id\"])\n",
    "        \n",
    "        # Minimal informative info is novel tx, novel intron_id & number, match column\n",
    "        novel_ref_match_info = novel_ref_match_info[[\"transcript_id_novel\",\"intron_id\",\"intron_number\",\"match\"]]\n",
    "        \n",
    "\n",
    "    elif match_type == \"transcript\":\n",
    "        # Looking for introns (except last) to match the same reference transcript\n",
    "        # merge_ordered can do a 'grouped merge' filling in empty rows (introns) for each transcript_id\n",
    "        # This is especially useful if want to do transcript-specific intron matching\n",
    "        # For each reference transcript, all novel introns will be filled with NaN if no overlap for given transcript_id\n",
    "        # (i.e. novel txipt matches all but last intron of reference transcript)\n",
    "\n",
    "        novel_ref_match_info = (pd.merge_ordered(novel_intron_ids_ordered,\n",
    "                                    joined,\n",
    "                                    how=\"left\",\n",
    "                                    on=\"intron_id\",\n",
    "                                    right_by=\"transcript_id_ref\", # group matches by ref tx & join tx by tx\n",
    "                                    suffixes=[\"_novel\",\"_match\"],\n",
    "                                    fill_method=None)\n",
    "                   .sort_values(by=[\"transcript_id_novel\",\"intron_number\"])\n",
    "                               )\n",
    "        \n",
    "        # merge_ordered fills rows for each intron for each ref tx in df, regardless of whether any overlap\n",
    "        # .dropna(axis=\"rows\", subset=[\"intron_id_ref\"]) \n",
    "        novel_ref_match_info = (novel_ref_match_info.groupby([\"transcript_id_novel\", \"transcript_id_ref\"])\n",
    "                                .filter(lambda df: (df[\"intron_id_ref\"].notna()).any()) # Retained if ref tx has >=1 matching introns\n",
    "                                .reset_index(drop=True))\n",
    "        \n",
    "        # Make a match column where 1 = match, 0 = no match for each ref id and novel intron\n",
    "        novel_ref_match_info[\"match\"] = novel_ref_match_info[\"intron_id_ref\"]\n",
    "        novel_ref_match_info[\"match\"] = novel_ref_match_info[\"match\"].fillna(0)\n",
    "        novel_ref_match_info[\"match\"] = novel_ref_match_info[\"match\"].replace(\"\\w*\", 1, regex=True)\n",
    "        \n",
    "    t12 = timer()\n",
    "    print(\"took {} s\".format(t12 - t11))\n",
    "    \n",
    "    \n",
    "    # 8. Filter down matching transcripts to those that all ref introns except penultimate or all introns...\n",
    "    print(\"filtering for valid intron chain matches...\")\n",
    "    t13 = timer()\n",
    "    if match_type == \"any\":\n",
    "        # Only need to check by novel transcript_id\n",
    "        filt_novel_ref_match_info = (novel_ref_match_info.groupby(\"transcript_id_novel\")\n",
    "                                     .filter(lambda x: validate_matching_chain(x, max_terminal_non_match)\n",
    "                                            )\n",
    "                                    )\n",
    "    \n",
    "    elif match_type == \"transcript\":\n",
    "        # Check novel tx vs each ref tx\n",
    "        filt_novel_ref_match_info = (novel_ref_match_info.groupby([\"transcript_id_novel\",\"transcript_id_ref\"])\n",
    "                                     .filter(lambda x: validate_matching_chain(x, max_terminal_non_match)\n",
    "                                            )\n",
    "                                    )\n",
    "    t14 = timer()\n",
    "    print(\"took {} s\".format(t14 - t13))\n",
    "        \n",
    "    \n",
    "    \n",
    "    # Return simplified df of novel transcript_id & matching transcript_ids if applicable\n",
    "    \n",
    "    if match_type == \"any\":\n",
    "        chain_match_novel_ids = filt_novel_ref_match_info[\"transcript_id_novel\"].drop_duplicates()\n",
    "    \n",
    "    elif match_type == \"transcript\":\n",
    "        chain_match_novel_ids = filt_novel_ref_match_info[[\"transcript_id_novel\",\"transcript_id_ref\"]].drop_duplicates()\n",
    "    \n",
    "    # Extract valid transcripts occurring within first intron of annotated transcript (where intron chain matching is impossible)\n",
    "    # For these to be valid, the 3'end of first exons should be identical (i.e. same outgoing junction)\n",
    "    \n",
    "    #2.1 - Pull out non-chain matched novel isoforms & their last exons\n",
    "    \n",
    "    if isinstance(chain_match_novel_ids, pd.Series):\n",
    "        # match type was any\n",
    "        novel_exons_nm = novel_exons.subset(lambda df: ~df[\"transcript_id\"].isin(set(chain_match_novel_ids.tolist())), nb_cpu=nb_cpu)\n",
    "        \n",
    "\n",
    "\n",
    "    elif isinstance(chain_match_novel_ids, pd.DataFrame):\n",
    "        # match_by/match_type was transcript\n",
    "        novel_exons_nm = novel_exons.subset(lambda df: ~df[\"transcript_id\"].isin(set(chain_match_novel_ids[\"transcript_id_novel\"].tolist())), nb_cpu=nb_cpu)\n",
    "\n",
    "    novel_nm_last_exons = get_terminal_exons(novel_exons_nm,\n",
    "                                              exon_number_col=\"exon_number\",\n",
    "                                              source=\"stringtie\",\n",
    "                                              filter_single=True,\n",
    "                                              which_exon=\"last\",\n",
    "                                              nb_cpu=nb_cpu)\n",
    "    \n",
    "    #2.2 - Extract first introns from ref transcripts\n",
    "    ref_first_introns = get_terminal_exons(ref_introns,\n",
    "                                           exon_number_col=\"intron_number\",\n",
    "                                           source=None,\n",
    "                                           filter_single=False,\n",
    "                                           which_exon=\"first\",\n",
    "                                           nb_cpu=nb_cpu\n",
    "                                          )\n",
    "    \n",
    "    \n",
    "    #2.3 - find last exons of non-matched txs completely contained within annotated first introns\n",
    "    novel_nm_fi = pr.PyRanges(novel_nm_last_exons.as_df(), int64=True).overlap(ref_first_introns,\n",
    "                                               how=\"containment\",\n",
    "                                               strandedness=\"same\",\n",
    "#                                                nb_cpu=nb_cpu\n",
    "                                             )\n",
    "    \n",
    "    #2.5 - Get 3'ends of first exons of transcripts with first-intron contained last exons\n",
    "    novel_nm_fi_fe_3p = (novel_exons_nm.subset(lambda df: \n",
    "                                            df[\"transcript_id\"].isin(set(novel_nm_fi.transcript_id.tolist())),\n",
    "                                            nb_cpu=nb_cpu\n",
    "                                           )\n",
    "                      .three_end()\n",
    "                     )\n",
    "    \n",
    "    #2.6 - Get 3'ends of first exons of reference transcripts\n",
    "    ref_first_exons_3p = (get_terminal_exons(ref_exons,\n",
    "                                             exon_number_col=\"exon_number\",\n",
    "                                             source=None,\n",
    "                                             filter_single=True,\n",
    "                                             which_exon=\"first\",\n",
    "                                             nb_cpu=nb_cpu\n",
    "                                            )\n",
    "                          .three_end()\n",
    "                         )\n",
    "    \n",
    "    # 2.7 - Find first-intron contained novel LE isoforms with the outgoing SJ of first exon matching a ref first exon\n",
    "    first_intron_contained_match = novel_nm_fi_fe_3p.join(ref_first_exons_3p,\n",
    "                                                          strandedness=\"same\",\n",
    "                                                          suffix=\"_ref\",\n",
    "                                                          nb_cpu=nb_cpu\n",
    "                                                         )\n",
    "    \n",
    "    first_intron_contained_match = first_intron_contained_match.as_df()[[\"transcript_id\",\"transcript_id_ref\"]].rename({\"transcript_id\": \"transcript_id_novel\"}, axis=1)\n",
    "    \n",
    "    \n",
    "    \n",
    "    return pd.concat([pd.DataFrame(chain_match_novel_ids), first_intron_contained_match]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2775920",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_transcripts_by_chain(test_novel_gr, test_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "f1495743",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "finding introns...\n",
      "took 0.0867136429878883 (s)\n",
      "adding intron_id column...\n",
      "took 0.017458201968111098 s\n",
      "generating df of novel txipts sorted by intron number...\n",
      "took 0.019409961998462677 s\n",
      "finding overlaps between novel and reference introns...\n",
      "took 0.02133792103268206 s\n",
      "filtering overlaps for exact matches...\n",
      "took 0.005796824989374727 s\n",
      "preparing for filtering intron matches...\n",
      "took 0.004225927987135947 s\n",
      "filtering for valid intron chain matches...\n",
      "took 0.0026123709976673126 s\n",
      "Filtering for multi-exon transcripts...\n",
      "Before: 4\n",
      "After: 4\n",
      "Filtering for multi-exon transcripts...\n",
      "Before: 2\n",
      "After: 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id_novel</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>nov_tx_si_p</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id_novel transcript_id_ref\n",
       "0            nov_tx_3ui_p               NaN\n",
       "1             nov_tx_bl_p               NaN\n",
       "2       nov_tx_exc_dist_p               NaN\n",
       "3        nov_tx_le_dist_p               NaN\n",
       "4  nov_tx_le_dist_p_minus               NaN\n",
       "5           nov_tx_mult_p               NaN\n",
       "6             nov_tx_si_p               NaN\n",
       "7             nov_tx_fi_p          ref_tr_1"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filter_transcripts_by_chain(test_novel_gr, test_ref,match_type=\"any\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a6f6c570",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "finding introns...\n",
      "took 79.48507030098699 (s)\n",
      "adding intron_id column...\n",
      "took 0.17554569896310568 s\n",
      "generating df of novel txipts sorted by intron number...\n",
      "took 4.423515614995267 s\n",
      "finding overlaps between novel and reference introns...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-26 14:19:48,128\tINFO services.py:1272 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "took 4.683659655973315 s\n",
      "filtering overlaps for exact matches...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-26 14:19:52,562\tINFO services.py:1272 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "took 3.6306387850199826 s\n",
      "preparing for filtering intron matches...\n",
      "took 0.19747438799822703 s\n",
      "filtering for valid intron chain matches...\n",
      "took 0.42044611898018047 s\n",
      "total time taken: 93.09141320799245\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "26        PAPA.100.1\n",
       "86       PAPA.1001.2\n",
       "139      PAPA.1002.1\n",
       "208      PAPA.1003.1\n",
       "210      PAPA.1004.1\n",
       "            ...     \n",
       "99544     PAPA.985.1\n",
       "99546     PAPA.985.6\n",
       "99612     PAPA.994.2\n",
       "99678     PAPA.995.2\n",
       "99690     PAPA.999.1\n",
       "Name: transcript_id_novel, Length: 1191, dtype: object"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pre_chr1 = timer()\n",
    "chr1_any_matched = filter_transcripts_by_chain(stie_chr1_exons, ref_chr1_pc_exons, match_type=\"any\", nb_cpu=2)\n",
    "print(\"total time taken: {}\".format(timer() - pre_chr1))\n",
    "chr1_any_matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "95cbc06c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "finding introns...\n",
      "took 84.49933640897507 (s)\n",
      "adding intron_id column...\n",
      "took 0.162128378986381 s\n",
      "generating df of novel txipts sorted by intron number...\n",
      "took 4.3569289009901695 s\n",
      "finding overlaps between novel and reference introns...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-26 14:21:25,989\tINFO services.py:1272 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "ename": "RayTaskError(KeyError)",
     "evalue": "\u001b[36mray::_write_both()\u001b[39m (pid=1275, ip=10.97.45.105)\n  File \"python/ray/_raylet.pyx\", line 501, in ray._raylet.execute_task\n  File \"/home/sam/miniconda3/envs/bioinfo/lib/python3.8/site-packages/ray/util/tracing/tracing_helper.py\", line 330, in _function_with_tracing\n    return function(*args, **kwargs)\n  File \"/home/sam/miniconda3/envs/bioinfo/lib/python3.8/site-packages/pyranges/methods/join.py\", line 111, in _write_both\n    how = kwargs[\"how\"]\nKeyError: 'how'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRayTaskError(KeyError)\u001b[0m                    Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-25-7f58aabf7342>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mpre_chr1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtimer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mchr1_tx_matched\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfilter_transcripts_by_chain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstie_chr1_exons\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mref_chr1_pc_exons\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmatch_type\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"transcript\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnb_cpu\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"time taken: {}\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mpre_chr1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mchr1_tx_matched\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-23-745311db9660>\u001b[0m in \u001b[0;36mfilter_transcripts_by_chain\u001b[0;34m(novel_exons, ref_exons, match_type, max_terminal_non_match, nb_cpu)\u001b[0m\n\u001b[1;32m    106\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    107\u001b[0m     \u001b[0mt7\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtimer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 108\u001b[0;31m     \u001b[0mjoined\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnovel_introns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mref_introns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstrandedness\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"same\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msuffix\u001b[0m \u001b[0;34m=\u001b[0m\u001b[0;34m\"_ref\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mnb_cpu\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnb_cpu\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    109\u001b[0m     \u001b[0mt8\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtimer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    110\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.8/site-packages/pyranges/pyranges.py\u001b[0m in \u001b[0;36mjoin\u001b[0;34m(self, other, strandedness, how, report_overlap, slack, suffix, nb_cpu)\u001b[0m\n\u001b[1;32m   2151\u001b[0m             \u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"example_header_self\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2152\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2153\u001b[0;31m         \u001b[0mdfs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpyrange_apply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_write_both\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2154\u001b[0m         \u001b[0mgr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPyRanges\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdfs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2155\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.8/site-packages/pyranges/multithreaded.py\u001b[0m in \u001b[0;36mpyrange_apply\u001b[0;34m(function, self, other, **kwargs)\u001b[0m\n\u001b[1;32m    306\u001b[0m                 \u001b[0mresults\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    307\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 308\u001b[0;31m     \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    309\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    310\u001b[0m     \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprocess_results\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeys\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.8/site-packages/ray/_private/client_mode_hook.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     60\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mclient_mode_should_convert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     61\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mray\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 62\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     63\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     64\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.8/site-packages/ray/worker.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(object_refs, timeout)\u001b[0m\n\u001b[1;32m   1492\u001b[0m                     \u001b[0mworker\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcore_worker\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdump_object_store_memory_usage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1493\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mRayTaskError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1494\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mas_instanceof_cause\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1495\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1496\u001b[0m                     \u001b[0;32mraise\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRayTaskError(KeyError)\u001b[0m: \u001b[36mray::_write_both()\u001b[39m (pid=1275, ip=10.97.45.105)\n  File \"python/ray/_raylet.pyx\", line 501, in ray._raylet.execute_task\n  File \"/home/sam/miniconda3/envs/bioinfo/lib/python3.8/site-packages/ray/util/tracing/tracing_helper.py\", line 330, in _function_with_tracing\n    return function(*args, **kwargs)\n  File \"/home/sam/miniconda3/envs/bioinfo/lib/python3.8/site-packages/pyranges/methods/join.py\", line 111, in _write_both\n    how = kwargs[\"how\"]\nKeyError: 'how'"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-26 14:21:32,663\tERROR worker.py:78 -- Unhandled error (suppress with RAY_IGNORE_UNHANDLED_ERRORS=1): \u001b[36mray::_write_both()\u001b[39m (pid=1276, ip=10.97.45.105)\n",
      "  File \"python/ray/_raylet.pyx\", line 501, in ray._raylet.execute_task\n",
      "  File \"/home/sam/miniconda3/envs/bioinfo/lib/python3.8/site-packages/ray/util/tracing/tracing_helper.py\", line 330, in _function_with_tracing\n",
      "    return function(*args, **kwargs)\n",
      "  File \"/home/sam/miniconda3/envs/bioinfo/lib/python3.8/site-packages/pyranges/methods/join.py\", line 111, in _write_both\n",
      "    how = kwargs[\"how\"]\n",
      "KeyError: 'how'\n"
     ]
    }
   ],
   "source": [
    "pre_chr1 = timer()\n",
    "chr1_tx_matched = filter_transcripts_by_chain(stie_chr1_exons, ref_chr1_pc_exons, match_type=\"transcript\", nb_cpu=2)\n",
    "print(\"time taken: {}\".format(timer() - pre_chr1))\n",
    "chr1_tx_matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "14c85a2b",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'chr1_any_matched' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-80-841d1f14edb7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mstie_chr1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"transcript_id\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchr1_any_matched\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.8/site-packages/pyranges/pyranges.py\u001b[0m in \u001b[0;36msubset\u001b[0;34m(self, f, strand, **kwargs)\u001b[0m\n\u001b[1;32m   4295\u001b[0m         \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m\"strand\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstrand\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4296\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4297\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpyrange_apply_single\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4298\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4299\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.8/site-packages/pyranges/multithreaded.py\u001b[0m in \u001b[0;36mpyrange_apply_single\u001b[0;34m(function, self, **kwargs)\u001b[0m\n\u001b[1;32m    345\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    346\u001b[0m             \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmake_unary_sparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 347\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcall_f_single\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunction\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnparams\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    348\u001b[0m             \u001b[0mresults\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    349\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.8/site-packages/pyranges/multithreaded.py\u001b[0m in \u001b[0;36mcall_f_single\u001b[0;34m(f, nparams, df, **kwargs)\u001b[0m\n\u001b[1;32m     30\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremote\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     31\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 32\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremote\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     33\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-80-841d1f14edb7>\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(df)\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mstie_chr1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"transcript_id\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchr1_any_matched\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'chr1_any_matched' is not defined"
     ]
    }
   ],
   "source": [
    "stie_chr1.subset(lambda df: df[\"transcript_id\"].isin(set(chr1_any_matched.tolist())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a3847553",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isinstance(chr1_any_matched,pd.Series)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "7dc10302",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>cov</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>TPM</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>2025054</td>\n",
       "      <td>2025333</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.100</td>\n",
       "      <td>PAPA.100.1</td>\n",
       "      <td>3.943609</td>\n",
       "      <td>0.467749</td>\n",
       "      <td>1.213933</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>2025401</td>\n",
       "      <td>2025517</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.100</td>\n",
       "      <td>PAPA.100.1</td>\n",
       "      <td>3.943609</td>\n",
       "      <td>0.467749</td>\n",
       "      <td>1.213933</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>2025738</td>\n",
       "      <td>2027576</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.100</td>\n",
       "      <td>PAPA.100.1</td>\n",
       "      <td>3.943609</td>\n",
       "      <td>0.467749</td>\n",
       "      <td>1.213933</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>2027659</td>\n",
       "      <td>2028154</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.100</td>\n",
       "      <td>PAPA.100.1</td>\n",
       "      <td>3.943609</td>\n",
       "      <td>0.467749</td>\n",
       "      <td>1.213933</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>2028292</td>\n",
       "      <td>2029110</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.100</td>\n",
       "      <td>PAPA.100.1</td>\n",
       "      <td>3.943609</td>\n",
       "      <td>0.467749</td>\n",
       "      <td>1.213933</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21118</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>71592968</td>\n",
       "      <td>71611025</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.995</td>\n",
       "      <td>PAPA.995.2</td>\n",
       "      <td>1.635092</td>\n",
       "      <td>0.193937</td>\n",
       "      <td>0.503319</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21119</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>71611146</td>\n",
       "      <td>71698007</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.995</td>\n",
       "      <td>PAPA.995.2</td>\n",
       "      <td>1.635092</td>\n",
       "      <td>0.193937</td>\n",
       "      <td>0.503319</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21120</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>71698139</td>\n",
       "      <td>71776171</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.995</td>\n",
       "      <td>PAPA.995.2</td>\n",
       "      <td>1.635092</td>\n",
       "      <td>0.193937</td>\n",
       "      <td>0.503319</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21121</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>71776297</td>\n",
       "      <td>71935078</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.995</td>\n",
       "      <td>PAPA.995.2</td>\n",
       "      <td>1.635092</td>\n",
       "      <td>0.193937</td>\n",
       "      <td>0.503319</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21122</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>71935311</td>\n",
       "      <td>72282318</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.995</td>\n",
       "      <td>PAPA.995.2</td>\n",
       "      <td>1.635092</td>\n",
       "      <td>0.193937</td>\n",
       "      <td>0.503319</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21123 rows  14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | Score     | +8    |\n",
       "| (object)     | (object)   | (object)   | (int32)   | (int32)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-----------+-------|\n",
       "| chr1         | StringTie  | intron     | 2025054   | 2025333   | 1000      | ...   |\n",
       "| chr1         | StringTie  | intron     | 2025401   | 2025517   | 1000      | ...   |\n",
       "| chr1         | StringTie  | intron     | 2025738   | 2027576   | 1000      | ...   |\n",
       "| chr1         | StringTie  | intron     | 2027659   | 2028154   | 1000      | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...       | ...   |\n",
       "| chr1         | StringTie  | intron     | 71611146  | 71698007  | 1000      | ...   |\n",
       "| chr1         | StringTie  | intron     | 71698139  | 71776171  | 1000      | ...   |\n",
       "| chr1         | StringTie  | intron     | 71776297  | 71935078  | 1000      | ...   |\n",
       "| chr1         | StringTie  | intron     | 71935311  | 72282318  | 1000      | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 21,123 rows and 14 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "8 hidden columns: Strand, Frame, gene_id, transcript_id, cov, FPKM, TPM, exon_number"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stie_chr1.features.introns(by=\"transcript\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "e2ed6164",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_intron_number(introns, id_col = \"transcript_id\", out_col=\"intron_number\", nb_cpu=1):\n",
    "    '''\n",
    "    '''\n",
    "\n",
    "    assert len(set(introns.as_df().Feature.tolist())) == 1, \"only one feature type (e.g. all exons, all introns) should be present in gr\"\n",
    "\n",
    "    introns_n = (introns.assign(out_col,\n",
    "                                lambda df: df.groupby(id_col)\n",
    "                                             .apply(sort_introns_by_strand)\n",
    "                                             .reset_index(drop=True)\n",
    "                                             .rename({\"index\": out_col}, axis=\"columns\")\n",
    "                                             .assign(**{out_col: lambda x: x[out_col] + 1})[out_col],\n",
    "                                             nb_cpu=nb_cpu\n",
    "                                ))\n",
    "\n",
    "    return introns_n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "7d45afdb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1118    ENST00000327044.6\n",
       "2308    ENST00000465727.5\n",
       "2931    ENST00000354700.9\n",
       "2553    ENST00000473215.5\n",
       "2120    ENST00000379289.5\n",
       "Name: transcript_id, dtype: object"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "random_tx = pr.data.gencode_gtf().transcript_id.sample(n=5, random_state=123)\n",
    "random_tx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "334fb16a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1173905</td>\n",
       "      <td>1197933</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000162571.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TTLL10-202</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000368591.1</td>\n",
       "      <td>CCDS44036.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>1173905</td>\n",
       "      <td>1173926</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000162571.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TTLL10-202</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001480388.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000368591.1</td>\n",
       "      <td>CCDS44036.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>1174284</td>\n",
       "      <td>1174321</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000162571.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TTLL10-202</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001480387.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000368591.1</td>\n",
       "      <td>CCDS44036.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>1174423</td>\n",
       "      <td>1174489</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000162571.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TTLL10-202</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001480465.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000368591.1</td>\n",
       "      <td>CCDS44036.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>1179188</td>\n",
       "      <td>1179333</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000162571.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TTLL10-202</td>\n",
       "      <td>2</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00002321912.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000368591.1</td>\n",
       "      <td>CCDS44036.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>1292375</td>\n",
       "      <td>1293708</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000131584.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ACAP3-202</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000006366.2</td>\n",
       "      <td>24</td>\n",
       "      <td>ENSE00001889322.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000346733.5</td>\n",
       "      <td>CCDS19.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1293566</td>\n",
       "      <td>1293708</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000131584.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ACAP3-202</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000006366.2</td>\n",
       "      <td>24</td>\n",
       "      <td>ENSE00001889322.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000346733.5</td>\n",
       "      <td>CCDS19.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>1293563</td>\n",
       "      <td>1293566</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000131584.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ACAP3-202</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000006366.2</td>\n",
       "      <td>24</td>\n",
       "      <td>ENSE00001889322.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000346733.5</td>\n",
       "      <td>CCDS19.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>1307815</td>\n",
       "      <td>1308018</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000131584.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ACAP3-202</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000006366.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001927436.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000346733.5</td>\n",
       "      <td>CCDS19.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>1292375</td>\n",
       "      <td>1293566</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000131584.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ACAP3-202</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMT00000006366.2</td>\n",
       "      <td>24</td>\n",
       "      <td>ENSE00001889322.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000346733.5</td>\n",
       "      <td>CCDS19.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>170 rows  24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+----------------------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | Score      | Strand       | Frame      | gene_id            | gene_type      | gene_name   | level      | havana_gene          | +11   |\n",
       "| (category)   | (object)   | (category)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)           | (object)       | (object)    | (object)   | (object)             | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+----------------------+-------|\n",
       "| chr1         | ENSEMBL    | transcript   | 1173905   | 1197933   | .          | +            | .          | ENSG00000162571.13 | protein_coding | TTLL10      | 3          | OTTHUMG00000000851.4 | ...   |\n",
       "| chr1         | ENSEMBL    | exon         | 1173905   | 1173926   | .          | +            | .          | ENSG00000162571.13 | protein_coding | TTLL10      | 3          | OTTHUMG00000000851.4 | ...   |\n",
       "| chr1         | ENSEMBL    | exon         | 1174284   | 1174321   | .          | +            | .          | ENSG00000162571.13 | protein_coding | TTLL10      | 3          | OTTHUMG00000000851.4 | ...   |\n",
       "| chr1         | ENSEMBL    | exon         | 1174423   | 1174489   | .          | +            | .          | ENSG00000162571.13 | protein_coding | TTLL10      | 3          | OTTHUMG00000000851.4 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...        | ...          | ...        | ...                | ...            | ...         | ...        | ...                  | ...   |\n",
       "| chr1         | HAVANA     | CDS          | 1293566   | 1293708   | .          | -            | 1          | ENSG00000131584.18 | protein_coding | ACAP3       | 2          | OTTHUMG00000002235.4 | ...   |\n",
       "| chr1         | HAVANA     | stop_codon   | 1293563   | 1293566   | .          | -            | 0          | ENSG00000131584.18 | protein_coding | ACAP3       | 2          | OTTHUMG00000002235.4 | ...   |\n",
       "| chr1         | HAVANA     | UTR          | 1307815   | 1308018   | .          | -            | .          | ENSG00000131584.18 | protein_coding | ACAP3       | 2          | OTTHUMG00000002235.4 | ...   |\n",
       "| chr1         | HAVANA     | UTR          | 1292375   | 1293566   | .          | -            | .          | ENSG00000131584.18 | protein_coding | ACAP3       | 2          | OTTHUMG00000002235.4 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+----------------------+-------+\n",
       "Stranded PyRanges object has 170 rows and 24 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "11 hidden columns: transcript_id, transcript_type, transcript_name, transcript_support_level, tag, havana_transcript, exon_number, exon_id, ont, protein_id, ccdsid"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testing_few_tx = pr.data.gencode_gtf().subset(lambda df: df.transcript_id.isin(random_tx.tolist()))\n",
    "testing_few_tx\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "a34adcb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1173905</td>\n",
       "      <td>1173926</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000379289.5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1174284</td>\n",
       "      <td>1174321</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000379289.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1174423</td>\n",
       "      <td>1174489</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000379289.5</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1179188</td>\n",
       "      <td>1179333</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000379289.5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1179656</td>\n",
       "      <td>1179737</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000379289.5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1294717</td>\n",
       "      <td>1294816</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000354700.9</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1294401</td>\n",
       "      <td>1294628</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000354700.9</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1294089</td>\n",
       "      <td>1294199</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000354700.9</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1293822</td>\n",
       "      <td>1293933</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000354700.9</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1292375</td>\n",
       "      <td>1293708</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000354700.9</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------------------+---------------+\n",
       "| Chromosome   | Start     | End       | Strand       | transcript_id     | exon_number   |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | (object)          | (object)      |\n",
       "|--------------+-----------+-----------+--------------+-------------------+---------------|\n",
       "| chr1         | 1173905   | 1173926   | +            | ENST00000379289.5 | 1             |\n",
       "| chr1         | 1174284   | 1174321   | +            | ENST00000379289.5 | 2             |\n",
       "| chr1         | 1174423   | 1174489   | +            | ENST00000379289.5 | 3             |\n",
       "| chr1         | 1179188   | 1179333   | +            | ENST00000379289.5 | 4             |\n",
       "| ...          | ...       | ...       | ...          | ...               | ...           |\n",
       "| chr1         | 1294401   | 1294628   | -            | ENST00000354700.9 | 21            |\n",
       "| chr1         | 1294089   | 1294199   | -            | ENST00000354700.9 | 22            |\n",
       "| chr1         | 1293822   | 1293933   | -            | ENST00000354700.9 | 23            |\n",
       "| chr1         | 1292375   | 1293708   | -            | ENST00000354700.9 | 24            |\n",
       "+--------------+-----------+-----------+--------------+-------------------+---------------+\n",
       "Stranded PyRanges object has 73 rows and 6 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testing_few_tx.subset(lambda df: df.Feature == \"exon\")[[\"transcript_id\", \"exon_number\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "4ff803f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2470\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>my_exon_number</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2465</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1430549</td>\n",
       "      <td>1430954</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000434150.1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2466</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1434177</td>\n",
       "      <td>1434488</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000454562.1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2467</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1430549</td>\n",
       "      <td>1430662</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000454562.1</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2468</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1434177</td>\n",
       "      <td>1434520</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000417917.1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2469</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1430663</td>\n",
       "      <td>1430954</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000417917.1</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2470 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------------------+------------------+---------------+\n",
       "| Chromosome   | Start     | End       | Strand       | transcript_id     | my_exon_number   | exon_number   |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | (object)          | (int64)          | (object)      |\n",
       "|--------------+-----------+-----------+--------------+-------------------+------------------+---------------|\n",
       "| chr1         | 11868     | 12227     | +            | ENST00000456328.2 | 1                | 1             |\n",
       "| chr1         | 12612     | 12721     | +            | ENST00000456328.2 | 2                | 2             |\n",
       "| chr1         | 13220     | 14409     | +            | ENST00000456328.2 | 3                | 3             |\n",
       "| chr1         | 12009     | 12057     | +            | ENST00000450305.2 | 4                | 1             |\n",
       "| ...          | ...       | ...       | ...          | ...               | ...              | ...           |\n",
       "| chr1         | 1434177   | 1434488   | -            | ENST00000454562.1 | 5                | 1             |\n",
       "| chr1         | 1430549   | 1430662   | -            | ENST00000454562.1 | 6                | 2             |\n",
       "| chr1         | 1434177   | 1434520   | -            | ENST00000417917.1 | 7                | 1             |\n",
       "| chr1         | 1430663   | 1430954   | -            | ENST00000417917.1 | 8                | 2             |\n",
       "+--------------+-----------+-----------+--------------+-------------------+------------------+---------------+\n",
       "Stranded PyRanges object has 2,470 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(pr.data.gencode_gtf().subset(lambda df: df.Feature == \"exon\").as_df().index))\n",
    "\n",
    "(add_intron_number(pr.data.gencode_gtf().subset(lambda df: df.Feature == \"exon\"),\n",
    "                   out_col=\"my_exon_number\")#.subset(lambda df: df.my_exon_number == df.exon_number.astype(int))\n",
    " [[\"transcript_id\", \"my_exon_number\", \"exon_number\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "32bb4490",
   "metadata": {},
   "outputs": [],
   "source": [
    "def _sort_introns_by_strand(df, out_col):\n",
    "    '''\n",
    "    '''\n",
    "    \n",
    "    n_exons = len(df.index)\n",
    "    \n",
    "    if (df[\"Strand\"] == \"+\").all():\n",
    "        # first in order by txipt = left-most start position (i.e. most 5')\n",
    "        return df.assign(**{out_col: list(range(1, n_exons + 1))})\n",
    "    # pd.Series(list(range(1, n_exons + 1)))\n",
    "    \n",
    "    elif (df[\"Strand\"] == \"-\").all():\n",
    "        # firs in order by strart os = most 3'\n",
    "        return df.assign(**{out_col: list(range(1, n_exons +1))[::-1]})\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "def _add_intron_number(introns, id_col = \"transcript_id\", out_col=\"intron_number\", nb_cpu=1):\n",
    "    '''\n",
    "    '''\n",
    "\n",
    "    assert len(set(introns.as_df().Feature.tolist())) == 1, \"only one feature type (e.g. all exons, all introns) should be present in gr\"\n",
    "\n",
    "    # Sort by position\n",
    "    introns = introns.sort()\n",
    "                       \n",
    "    introns_out = introns.apply(lambda df: df.groupby(id_col).apply(lambda x: _sort_introns_by_strand(x, out_col)))\n",
    "                       \n",
    "#     introns_n = (introns.assign(out_col,\n",
    "#                                 lambda df: df.groupby(id_col)\n",
    "#                                              .apply(\n",
    "#                                                  _sort_introns_by_strand)\n",
    "#                                              .reset_index(drop=True)\n",
    "#                                              .rename({\"index\": out_col}, axis=\"columns\")\n",
    "#                                              .assign(**{out_col: lambda x: x[out_col] + 1})[out_col],\n",
    "#                                              nb_cpu=nb_cpu\n",
    "#                                 ))\n",
    "\n",
    "    return introns_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "c0a09226",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>my_exon_number</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12612</td>\n",
       "      <td>12697</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2465</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1430663</td>\n",
       "      <td>1430954</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000417917.1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2466</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1431804</td>\n",
       "      <td>1431843</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000434150.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2467</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1434177</td>\n",
       "      <td>1434488</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000454562.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2468</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1434177</td>\n",
       "      <td>1434520</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000417917.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2469</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1434187</td>\n",
       "      <td>1434573</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000430109.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2470 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------------------+------------------+---------------+\n",
       "| Chromosome   | Start     | End       | Strand       | transcript_id     | my_exon_number   | exon_number   |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | (object)          | (int64)          | (object)      |\n",
       "|--------------+-----------+-----------+--------------+-------------------+------------------+---------------|\n",
       "| chr1         | 11868     | 12227     | +            | ENST00000456328.2 | 1                | 1             |\n",
       "| chr1         | 12009     | 12057     | +            | ENST00000450305.2 | 1                | 1             |\n",
       "| chr1         | 12178     | 12227     | +            | ENST00000450305.2 | 2                | 2             |\n",
       "| chr1         | 12612     | 12697     | +            | ENST00000450305.2 | 3                | 3             |\n",
       "| ...          | ...       | ...       | ...          | ...               | ...              | ...           |\n",
       "| chr1         | 1431804   | 1431843   | -            | ENST00000434150.1 | 1                | 1             |\n",
       "| chr1         | 1434177   | 1434488   | -            | ENST00000454562.1 | 1                | 1             |\n",
       "| chr1         | 1434177   | 1434520   | -            | ENST00000417917.1 | 1                | 1             |\n",
       "| chr1         | 1434187   | 1434573   | -            | ENST00000430109.1 | 1                | 1             |\n",
       "+--------------+-----------+-----------+--------------+-------------------+------------------+---------------+\n",
       "Stranded PyRanges object has 2,470 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(_add_intron_number(pr.data.gencode_gtf().subset(lambda df: df.Feature == \"exon\"),\n",
    "                   out_col=\"my_exon_number\").subset(lambda df: df.my_exon_number == df.exon_number.astype(int))\n",
    " [[\"transcript_id\", \"my_exon_number\", \"exon_number\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "80a50884",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>79611116</td>\n",
       "      <td>79666158</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>79611116</td>\n",
       "      <td>79666158</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79611116</td>\n",
       "      <td>79611214</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002093744.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79611195</td>\n",
       "      <td>79611214</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002093744.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>79611195</td>\n",
       "      <td>79611198</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002093744.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79636801</td>\n",
       "      <td>79636897</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003660857.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79636801</td>\n",
       "      <td>79636897</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003660857.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79641377</td>\n",
       "      <td>79641550</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00000909925.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79641377</td>\n",
       "      <td>79641550</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00000909925.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79654870</td>\n",
       "      <td>79655062</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001010210.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79654870</td>\n",
       "      <td>79655062</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001010210.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79664814</td>\n",
       "      <td>79666158</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00001323097.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79664814</td>\n",
       "      <td>79664871</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00001323097.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>79664871</td>\n",
       "      <td>79664874</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00001323097.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>79611116</td>\n",
       "      <td>79611195</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002093744.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>79664871</td>\n",
       "      <td>79666158</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00001323097.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>79611151</td>\n",
       "      <td>79665007</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79611151</td>\n",
       "      <td>79611214</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001538470.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79611195</td>\n",
       "      <td>79611214</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001538470.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>79611195</td>\n",
       "      <td>79611198</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001538470.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79636801</td>\n",
       "      <td>79636897</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003660857.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79636801</td>\n",
       "      <td>79636897</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003660857.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79641377</td>\n",
       "      <td>79641550</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00000909925.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79641377</td>\n",
       "      <td>79641550</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00000909925.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79654870</td>\n",
       "      <td>79655062</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001010210.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79654870</td>\n",
       "      <td>79655062</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001010210.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79663579</td>\n",
       "      <td>79663661</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00002108352.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79663579</td>\n",
       "      <td>79663660</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00002108352.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79664814</td>\n",
       "      <td>79665007</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00002102517.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>79663660</td>\n",
       "      <td>79663661</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00002102517.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>79664814</td>\n",
       "      <td>79664816</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00002102517.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>79611151</td>\n",
       "      <td>79611195</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001538470.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>79663660</td>\n",
       "      <td>79663661</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00002108352.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>79664814</td>\n",
       "      <td>79665007</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00002102517.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>79611734</td>\n",
       "      <td>79664881</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79611734</td>\n",
       "      <td>79611791</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002097140.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79636801</td>\n",
       "      <td>79636897</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003574037.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79636815</td>\n",
       "      <td>79636897</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003574037.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>79636815</td>\n",
       "      <td>79636818</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003574037.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79641377</td>\n",
       "      <td>79641550</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00000909925.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79641377</td>\n",
       "      <td>79641550</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00000909925.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79654870</td>\n",
       "      <td>79655062</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001010210.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79654870</td>\n",
       "      <td>79655062</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001010210.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79664814</td>\n",
       "      <td>79664881</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00002106021.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79664814</td>\n",
       "      <td>79664871</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00002106021.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>79664871</td>\n",
       "      <td>79664874</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00002106021.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>79611734</td>\n",
       "      <td>79611791</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002097140.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>79636801</td>\n",
       "      <td>79636815</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003574037.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>79664871</td>\n",
       "      <td>79664881</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00002106021.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>49 rows  24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+------------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | Score      | Strand       | Frame      | gene_id            | gene_type      | gene_name   | level      | hgnc_id    | +11   |\n",
       "| (category)   | (object)   | (category)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)           | (object)       | (object)    | (object)   | (object)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+------------+-------|\n",
       "| chr8         | HAVANA     | gene         | 79611116  | 79666158  | .          | +            | .          | ENSG00000104435.14 | protein_coding | STMN2       | 1          | HGNC:10577 | ...   |\n",
       "| chr8         | HAVANA     | transcript   | 79611116  | 79666158  | .          | +            | .          | ENSG00000104435.14 | protein_coding | STMN2       | 2          | HGNC:10577 | ...   |\n",
       "| chr8         | HAVANA     | exon         | 79611116  | 79611214  | .          | +            | .          | ENSG00000104435.14 | protein_coding | STMN2       | 2          | HGNC:10577 | ...   |\n",
       "| chr8         | HAVANA     | CDS          | 79611195  | 79611214  | .          | +            | 0          | ENSG00000104435.14 | protein_coding | STMN2       | 2          | HGNC:10577 | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...        | ...          | ...        | ...                | ...            | ...         | ...        | ...        | ...   |\n",
       "| chr8         | HAVANA     | stop_codon   | 79664871  | 79664874  | .          | +            | 0          | ENSG00000104435.14 | protein_coding | STMN2       | 2          | HGNC:10577 | ...   |\n",
       "| chr8         | HAVANA     | UTR          | 79611734  | 79611791  | .          | +            | .          | ENSG00000104435.14 | protein_coding | STMN2       | 2          | HGNC:10577 | ...   |\n",
       "| chr8         | HAVANA     | UTR          | 79636801  | 79636815  | .          | +            | .          | ENSG00000104435.14 | protein_coding | STMN2       | 2          | HGNC:10577 | ...   |\n",
       "| chr8         | HAVANA     | UTR          | 79664871  | 79664881  | .          | +            | .          | ENSG00000104435.14 | protein_coding | STMN2       | 2          | HGNC:10577 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+------------+-------+\n",
       "Stranded PyRanges object has 49 rows and 24 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "11 hidden columns: havana_gene, transcript_id, transcript_type, transcript_name, protein_id, transcript_support_level, tag, ccdsid, havana_transcript, exon_number, exon_id"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stmn2_ref = pr.read_gtf(\"../data/annotation/STMN2.gencode.v34.annotation.gtf\")\n",
    "stmn2_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "f7cf6b00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79611116</td>\n",
       "      <td>79611214</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002093744.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79611151</td>\n",
       "      <td>79611214</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001538470.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>79611734</td>\n",
       "      <td>79611791</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000518491.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002097140.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows  24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+------------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | Score      | Strand       | Frame      | gene_id            | gene_type      | gene_name   |      level | hgnc_id    | +11   |\n",
       "| (category)   | (object)   | (category)   |   (int32) |   (int32) | (object)   | (category)   | (object)   | (object)           | (object)       | (object)    |   (object) | (object)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+------------+-------|\n",
       "| chr8         | HAVANA     | exon         |  79611116 |  79611214 | .          | +            | .          | ENSG00000104435.14 | protein_coding | STMN2       |          2 | HGNC:10577 | ...   |\n",
       "| chr8         | HAVANA     | exon         |  79611151 |  79611214 | .          | +            | .          | ENSG00000104435.14 | protein_coding | STMN2       |          1 | HGNC:10577 | ...   |\n",
       "| chr8         | HAVANA     | exon         |  79611734 |  79611791 | .          | +            | .          | ENSG00000104435.14 | protein_coding | STMN2       |          2 | HGNC:10577 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+------------+-------+\n",
       "Stranded PyRanges object has 3 rows and 24 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "11 hidden columns: havana_gene, transcript_id, transcript_type, transcript_name, protein_id, transcript_support_level, tag, ccdsid, havana_transcript, exon_number, exon_id"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_terminal_exons(stmn2_ref.subset(lambda df: df.Feature == \"exon\"),which_exon=\"first\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "0a452d8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>cov</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>TPM</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>79611151</td>\n",
       "      <td>79617044</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.11426</td>\n",
       "      <td>PAPA.11426.2</td>\n",
       "      <td>3192.117432</td>\n",
       "      <td>378.615173</td>\n",
       "      <td>982.606506</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>79611151</td>\n",
       "      <td>79611214</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.11426</td>\n",
       "      <td>PAPA.11426.2</td>\n",
       "      <td>547.977051</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617044</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.11426</td>\n",
       "      <td>PAPA.11426.2</td>\n",
       "      <td>3939.116699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+---------------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End |     Score | Strand       | Frame      | gene_id    | transcript_id   |        cov |       FPKM |        TPM |   exon_number |\n",
       "| (category)   | (object)   | (category)   |   (int32) |   (int32) |   (int64) | (category)   | (object)   | (object)   | (object)        |   (object) |   (object) |   (object) |      (object) |\n",
       "|--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+---------------|\n",
       "| chr8         | StringTie  | transcript   |  79611151 |  79617044 |      1000 | +            | .          | PAPA.11426 | PAPA.11426.2    |   3192.12  |    378.615 |    982.607 |           nan |\n",
       "| chr8         | StringTie  | exon         |  79611151 |  79611214 |      1000 | +            | .          | PAPA.11426 | PAPA.11426.2    |    547.977 |    nan     |    nan     |             1 |\n",
       "| chr8         | StringTie  | exon         |  79616821 |  79617044 |      1000 | +            | .          | PAPA.11426 | PAPA.11426.2    |   3939.12  |    nan     |    nan     |             2 |\n",
       "+--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+---------------+\n",
       "Stranded PyRanges object has 3 rows and 14 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stie_chr8 = pr.read_gtf(\"../two_sample_example_output/stringtie/chr8.no_ref_id.TDP43-F_S6.assembled.gtf\")\n",
    "\n",
    "stie_stmn2 = stie_chr8.subset(lambda df: df.transcript_id == \"PAPA.11426.2\")\n",
    "\n",
    "stie_stmn2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "d0e001b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_terminal_regions(gr,\n",
    "                   feature_col = \"Feature\",\n",
    "                   feature_key = \"exon\",\n",
    "                   id_col = \"transcript_id\",\n",
    "                   region_number_col = \"exon_number\",\n",
    "                   source = None,\n",
    "                   which_region=\"last\",\n",
    "                   filter_single = False,\n",
    "                   nb_cpu = 1):\n",
    "    '''\n",
    "    Return gr of last exons for each transcript_id\n",
    "    In process, region_number_col will be converted to type 'int'\n",
    "    StringTie merged GTFs (or whatever tool single_steps/stringtie_longreads.smk is using)\n",
    "    reports exon_number that DOES NOT RESPECT STRAND (from browsing in IGV)\n",
    "    i.e. for minus-strand - largest exon_number for transcript corresponds to FIRST EXON, not last\n",
    "    Annotated (i.e. Ensembl) reported exon_numbers DO RESPECT STRAND (i.e. max always = last exon)\n",
    "\n",
    "    if Do respect strand, put source = None (default)\n",
    "    if Don't respect strand, put source = \"stringtie\" (i.e. plus strand = max, minus strand = min)\n",
    "    '''\n",
    "\n",
    "    assert source in [None, \"stringtie\"]\n",
    "    assert which_region in [\"first\", \"last\"]\n",
    "    assert region_number_col in gr.columns.tolist()\n",
    "    assert feature_col in gr.columns.tolist()\n",
    "    assert id_col in gr.columns.tolist()\n",
    "\n",
    "    # Make sure only 'exon' features are in the gr\n",
    "    assert gr.as_df()[feature_col].drop_duplicates().tolist() == [feature_key], \"only {} entries should be present in gr\".format(feature_key)\n",
    "\n",
    "    # Make sure region_number_col is int\n",
    "    mod_gr = (gr.assign(region_number_col,\n",
    "                      lambda df: df[region_number_col].astype(int),\n",
    "                      nb_cpu = nb_cpu)\n",
    "             )\n",
    "\n",
    "\n",
    "    # Filter out single-exon transcripts\n",
    "    if filter_single:\n",
    "        print(\"Filtering for multi-exon transcripts...\")\n",
    "        print(\"Before: {}\".format(len(set(mod_gr.as_df()[id_col].tolist()))))\n",
    "\n",
    "        mod_gr = (mod_gr.apply(lambda df: (df.groupby(id_col)\n",
    "                                       .filter(lambda x: filter_multi_exon(df, region_number_col))\n",
    "                                      )\n",
    "                           ,\n",
    "                           nb_cpu=nb_cpu\n",
    "                          )\n",
    "                 )\n",
    "        print(\"After: {}\".format(len(set(mod_gr.as_df()[id_col].tolist()))))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    if source is None:\n",
    "        # source = None means that 1 = first region of group regardless of strand\n",
    "        # Pick last region entry by max region number for each transcript (id_col)\n",
    "        # Pick first region entry by min region number for each transcript (id_col)\n",
    "\n",
    "        if which_region == \"last\":\n",
    "            out_gr = mod_gr.apply(lambda df: df.iloc[df.groupby(id_col)[region_number_col].idxmax(),], nb_cpu = nb_cpu)\n",
    "\n",
    "        elif which_region == \"first\":\n",
    "            out_gr = mod_gr.apply(lambda df: df.iloc[df.groupby(id_col)[region_number_col].idxmin(),], nb_cpu = nb_cpu)\n",
    "\n",
    "    elif source == \"stringtie\":\n",
    "        # Numbering Doesn't respect strand - pick min if Minus strand, max if plus strand\n",
    "        out_gr = (mod_gr.apply(lambda df: df.iloc[(df.groupby(id_col)\n",
    "                                                  .apply(lambda df: stie_groupby_last_exon(df, region_number_col, which_region)\n",
    "                                                        )\n",
    "                                                 ),],\n",
    "                               nb_cpu = nb_cpu\n",
    "                              )\n",
    "                 )\n",
    "\n",
    "\n",
    "    return out_gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "79dd923c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------+-----------+-----------+--------------+-------------+--------------------+-----------------+\n",
      "| Chromosome   | Start     | End       | Strand       | gene_name   | transcript_id      | intron_number   |\n",
      "| (object)     | (int32)   | (int32)   | (category)   | (object)    | (object)           | (int64)         |\n",
      "|--------------+-----------+-----------+--------------+-------------+--------------------+-----------------|\n",
      "| chr8         | 79611214  | 79636801  | +            | STMN2       | ENST00000220876.12 | 1               |\n",
      "| chr8         | 79611214  | 79636801  | +            | STMN2       | ENST00000518111.5  | 1               |\n",
      "| chr8         | 79611791  | 79636801  | +            | STMN2       | ENST00000518491.1  | 1               |\n",
      "| chr8         | 79636897  | 79641377  | +            | STMN2       | ENST00000220876.12 | 2               |\n",
      "| ...          | ...       | ...       | ...          | ...         | ...                | ...             |\n",
      "| chr8         | 79655062  | 79663579  | +            | STMN2       | ENST00000518111.5  | 4               |\n",
      "| chr8         | 79655062  | 79664814  | +            | STMN2       | ENST00000220876.12 | 4               |\n",
      "| chr8         | 79655062  | 79664814  | +            | STMN2       | ENST00000518491.1  | 4               |\n",
      "| chr8         | 79663661  | 79664814  | +            | STMN2       | ENST00000518111.5  | 5               |\n",
      "+--------------+-----------+-----------+--------------+-------------+--------------------+-----------------+\n",
      "Stranded PyRanges object has 13 rows and 7 columns from 1 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>intron_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>intron</td>\n",
       "      <td>79611214</td>\n",
       "      <td>79636801</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-201</td>\n",
       "      <td>ENSP00000220876.7</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS43748.1</td>\n",
       "      <td>OTTHUMT00000379261.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>intron</td>\n",
       "      <td>79611214</td>\n",
       "      <td>79636801</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-202</td>\n",
       "      <td>ENSP00000429243.1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS56542.1</td>\n",
       "      <td>OTTHUMT00000379266.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>intron</td>\n",
       "      <td>79611791</td>\n",
       "      <td>79636801</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000104435.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>STMN2-203</td>\n",
       "      <td>ENSP00000430102.1</td>\n",
       "      <td>2</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000379267.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+------------+-------+\n",
       "| Chromosome   | Source     | Feature    |     Start |       End | Score      | Strand       | Frame      | gene_id            | gene_type      | gene_name   |      level | hgnc_id    | +12   |\n",
       "| (object)     | (object)   | (object)   |   (int32) |   (int32) | (object)   | (category)   | (object)   | (object)           | (object)       | (object)    |   (object) | (object)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+------------+-------|\n",
       "| chr8         | HAVANA     | intron     |  79611214 |  79636801 | .          | +            | .          | ENSG00000104435.14 | protein_coding | STMN2       |          2 | HGNC:10577 | ...   |\n",
       "| chr8         | HAVANA     | intron     |  79611214 |  79636801 | .          | +            | .          | ENSG00000104435.14 | protein_coding | STMN2       |          1 | HGNC:10577 | ...   |\n",
       "| chr8         | HAVANA     | intron     |  79611791 |  79636801 | .          | +            | .          | ENSG00000104435.14 | protein_coding | STMN2       |          2 | HGNC:10577 | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+------------+-------+\n",
       "Stranded PyRanges object has 3 rows and 25 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "12 hidden columns: havana_gene, transcript_id, transcript_type, transcript_name, protein_id, transcript_support_level, tag, ccdsid, havana_transcript, exon_number, exon_id, intron_number"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Is STMN2 transcript completely contained within reference first intron (it should be...)?\n",
    "ref_stmn2_introns = stmn2_ref.features.introns(by=\"transcript\")\n",
    "ref_stmn2_introns = _add_intron_number(ref_stmn2_introns)\n",
    "print(ref_stmn2_introns[[\"gene_name\", \"transcript_id\", \"intron_number\"]])\n",
    "\n",
    "ref_stmn2_fi = get_terminal_regions(ref_stmn2_introns, feature_key=\"intron\", region_number_col=\"intron_number\", which_region=\"first\")\n",
    "ref_stmn2_fi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "2449c4fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>cov</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>TPM</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617044</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.11426</td>\n",
       "      <td>PAPA.11426.2</td>\n",
       "      <td>3939.116699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+---------------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End |     Score | Strand       | Frame      | gene_id    | transcript_id   |        cov |       FPKM |        TPM |   exon_number |\n",
       "| (category)   | (object)   | (category)   |   (int32) |   (int32) |   (int64) | (category)   | (object)   | (object)   | (object)        |   (object) |   (object) |   (object) |      (object) |\n",
       "|--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+---------------|\n",
       "| chr8         | StringTie  | exon         |  79616821 |  79617044 |      1000 | +            | .          | PAPA.11426 | PAPA.11426.2    |    3939.12 |        nan |        nan |             2 |\n",
       "+--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+---------------+\n",
       "Stranded PyRanges object has 1 rows and 14 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stie_stmn2.subset(lambda df: df.exon_number == \"2\").overlap(ref_stmn2_fi, how=\"containment\").drop_duplicate_positions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "89dbc62c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>cov</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>TPM</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>79616821</td>\n",
       "      <td>79617044</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.11426</td>\n",
       "      <td>PAPA.11426.2</td>\n",
       "      <td>3939.116699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+---------------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End |     Score | Strand       | Frame      | gene_id    | transcript_id   |        cov |       FPKM |        TPM |   exon_number |\n",
       "| (category)   | (object)   | (category)   |   (int32) |   (int32) |   (int64) | (category)   | (object)   | (object)   | (object)        |   (object) |   (object) |   (object) |       (int64) |\n",
       "|--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+---------------|\n",
       "| chr8         | StringTie  | exon         |  79616821 |  79617044 |      1000 | +            | .          | PAPA.11426 | PAPA.11426.2    |    3939.12 |        nan |        nan |             2 |\n",
       "+--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+---------------+\n",
       "Stranded PyRanges object has 1 rows and 14 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_terminal_regions(stie_stmn2.subset(lambda df: df.Feature == \"exon\"),which_region=\"last\",source=\"stringtie\").overlap(ref_stmn2_fi, how=\"containment\").drop_duplicate_positions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "d659a7c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start_ref</th>\n",
       "      <th>End_ref</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79611214</td>\n",
       "      <td>79611215</td>\n",
       "      <td>+</td>\n",
       "      <td>79611214</td>\n",
       "      <td>79611215</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79611214</td>\n",
       "      <td>79611215</td>\n",
       "      <td>+</td>\n",
       "      <td>79611214</td>\n",
       "      <td>79611215</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------------+-----------+---------------------+\n",
       "| Chromosome   |     Start |       End | Strand       |   Start_ref |   End_ref | transcript_id_ref   |\n",
       "| (category)   |   (int32) |   (int32) | (category)   |     (int32) |   (int32) | (object)            |\n",
       "|--------------+-----------+-----------+--------------+-------------+-----------+---------------------|\n",
       "| chr8         |  79611214 |  79611215 | +            |    79611214 |  79611215 | ENST00000220876.12  |\n",
       "| chr8         |  79611214 |  79611215 | +            |    79611214 |  79611215 | ENST00000518111.5   |\n",
       "+--------------+-----------+-----------+--------------+-------------+-----------+---------------------+\n",
       "Stranded PyRanges object has 2 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Not lost to first intron filter...\n",
    "# Is is lost to first exon 3'end matching (it shouldn't be... by eye ends are identical)\n",
    "stie_stmn2_fe = get_terminal_regions(stie_stmn2.subset(lambda df: df.Feature == \"exon\"), which_region=\"first\", source=\"stringtie\")\n",
    "ref_stmn2_fe = get_terminal_regions(stmn2_ref.subset(lambda df: df.Feature == \"exon\"), which_region=\"first\")\n",
    "\n",
    "stie_stmn2_fe.three_end().join(ref_stmn2_fe.three_end(), suffix=\"_ref\")[[\"Start_ref\", \"End_ref\", \"transcript_id_ref\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "2d63b208",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>transcript_id_b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79611214</td>\n",
       "      <td>79611215</td>\n",
       "      <td>+</td>\n",
       "      <td>79611214</td>\n",
       "      <td>79611215</td>\n",
       "      <td>ENST00000220876.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>79611214</td>\n",
       "      <td>79611215</td>\n",
       "      <td>+</td>\n",
       "      <td>79611214</td>\n",
       "      <td>79611215</td>\n",
       "      <td>ENST00000518111.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-----------+-----------+--------------------+\n",
       "| Chromosome   |     Start |       End | Strand       |   Start_b |     End_b | transcript_id_b    |\n",
       "| (category)   |   (int32) |   (int32) | (category)   |   (int32) |   (int32) | (object)           |\n",
       "|--------------+-----------+-----------+--------------+-----------+-----------+--------------------|\n",
       "| chr8         |  79611214 |  79611215 | +            |  79611214 |  79611215 | ENST00000220876.12 |\n",
       "| chr8         |  79611214 |  79611215 | +            |  79611214 |  79611215 | ENST00000518111.5  |\n",
       "+--------------+-----------+-----------+--------------+-----------+-----------+--------------------+\n",
       "Stranded PyRanges object has 2 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stie_stmn2.subset(lambda df: df.Feature == \"exon\").three_end().join(ref_stmn2_fe.three_end())[[\"Start_b\",\"End_b\",\"transcript_id_b\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "bb226e80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_testing = stie_stmn2_fe.three_end().as_df()\n",
    "any(df_testing.Start == df_testing.End)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:bioinfo] *",
   "language": "python",
   "name": "conda-env-bioinfo-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
