{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b5170bd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "import time\n",
    "from timeit import default_timer as timer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fb15ee4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "|            1 |        10 |        30 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       100 |       120 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       200 |       220 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       300 |       340 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 4 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A plus strand transcript\n",
    "test_ref_tr1 = {\"Chromosome\": [1]*4,\n",
    "                \"Start\": [10,100,200,300],\n",
    "                \"End\": [30,120,220,340],\n",
    "                \"Strand\": [\"+\"]*4,\n",
    "                \"Feature\": [\"exon\"]*4,\n",
    "                \"gene_id\": [\"ref_gene_1\"]*4,\n",
    "                \"transcript_id\": [\"ref_tr_1\"] * 4}\n",
    "\n",
    "\n",
    "pr.from_dict(test_ref_tr1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1f982b51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Chromosome': [],\n",
       " 'Start': [],\n",
       " 'End': [],\n",
       " 'Strand': [],\n",
       " 'Feature': [],\n",
       " 'gene_id': [],\n",
       " 'transcript_id': []}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{\"Chromosome\": [],\n",
    "                \"Start\": [],\n",
    "                \"End\": [],\n",
    "                \"Strand\": [],\n",
    "                \"Feature\": [],\n",
    "                \"gene_id\": [],\n",
    "                \"transcript_id\": []\n",
    "               }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "37648c16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>ref_tr_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>80</td>\n",
       "      <td>90</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>ref_tr_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>ref_tr_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "|            1 |        10 |        30 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       100 |       120 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       200 |       220 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       300 |       340 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "|            2 |        10 |        20 | -            | exon       | ref_gene_2 | ref_tr_2        |\n",
       "|            2 |        80 |        90 | -            | exon       | ref_gene_2 | ref_tr_2        |\n",
       "|            2 |       100 |       120 | -            | exon       | ref_gene_2 | ref_tr_2        |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 7 rows and 7 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A minus strand transcript\n",
    "test_ref_tr2 = {\"Chromosome\": [2]*3,\n",
    "                \"Start\": [10,80,100],\n",
    "                \"End\": [20,90,120],\n",
    "                \"Strand\": [\"-\"]*3,\n",
    "                \"Feature\": [\"exon\"]*3,\n",
    "                \"gene_id\": [\"ref_gene_2\"]*3,\n",
    "                \"transcript_id\": [\"ref_tr_2\"]*3\n",
    "               }\n",
    "\n",
    "\n",
    "test_ref = pr.concat([pr.from_dict(test_ref_tr1), pr.from_dict(test_ref_tr2)])\n",
    "test_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fbba9f36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>35</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_fi_p     |\n",
       "|            1 |        50 |        70 | +            | exon       | nov_gene_1 | nov_tx_fi_p     |\n",
       "|            1 |        10 |        35 | +            | exon       | nov_gene_1 | nov_tx_fi_f     |\n",
       "|            1 |        50 |        70 | +            | exon       | nov_gene_1 | nov_tx_fi_f     |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 4 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now make test novel transcripts to cover my test cases\n",
    "\n",
    "# 1. Novel last exons in first intron of annotated transcript (e.g. STMN2)\n",
    "# For these to pass, they should share an identical 3'end with a first exon of a known transcript\n",
    "\n",
    "test_novel_fi = {\"Chromosome\": [1]*4,\n",
    "                \"Start\": [10,50]*2,\n",
    "                \"End\": [30,70,35,70],\n",
    "                \"Strand\": [\"+\"]*4,\n",
    "                \"Feature\": [\"exon\"]*4,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*4,\n",
    "                \"transcript_id\": [\"nov_tx_fi_p\"]*2 + [\"nov_tx_fi_f\"]*2,\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_fi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "213b2de9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_si_p     |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_si_p     |\n",
       "|            1 |       140 |       160 | +            | exon       | nov_gene_1 | nov_tx_si_p     |\n",
       "|            1 |        50 |        70 | +            | exon       | nov_gene_1 | nov_tx_si_f     |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_si_f     |\n",
       "|            1 |       140 |       160 | +            | exon       | nov_gene_1 | nov_tx_si_f     |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 6 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2. Internal intron, spliced in last exon (fully contained within last exon) (e.g. ONECUT1)\n",
    "# For these to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "\n",
    "test_novel_si = {\"Chromosome\": [1]*6,\n",
    "                \"Start\": [10,100,140] + [50,100,140],\n",
    "                \"End\": [30,120,160] + [70,120,160],\n",
    "                \"Strand\": [\"+\"]*6,\n",
    "                \"Feature\": [\"exon\"]*6,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*6,\n",
    "                \"transcript_id\": [\"nov_tx_si_p\"]*3 + [\"nov_tx_si_f\"]*3\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_si)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2f34a0e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>240</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>240</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_bl_p     |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_bl_p     |\n",
       "|            1 |       200 |       240 | +            | exon       | nov_gene_1 | nov_tx_bl_p     |\n",
       "|            1 |        50 |        70 | +            | exon       | nov_gene_1 | nov_tx_bl_f     |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_bl_f     |\n",
       "|            1 |       200 |       240 | +            | exon       | nov_gene_1 | nov_tx_bl_f     |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 6 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3. Internal intron bleedthrough (e.g. SIN3B)\n",
    "# For these events to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "\n",
    "test_novel_bl = {\"Chromosome\": [1]*6,\n",
    "                \"Start\": [10,100,200] + [50,100,200],\n",
    "                \"End\": [30,120,240] + [70,120,240],\n",
    "                \"Strand\": [\"+\"]*6,\n",
    "                \"Feature\": [\"exon\"]*6,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*6,\n",
    "                \"transcript_id\": [\"nov_tx_bl_p\"]*3 + [\"nov_tx_bl_f\"]*3\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_bl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f32f14c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>130</td>\n",
       "      <td>140</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>150</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_mult_p   |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_mult_p   |\n",
       "|            1 |       130 |       140 | +            | exon       | nov_gene_1 | nov_tx_mult_p   |\n",
       "|            1 |       150 |       160 | +            | exon       | nov_gene_1 | nov_tx_mult_p   |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 4 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 4. Internal intron with novel internal and terminal exon (e.g.)\n",
    "# For these events to pass, they should match the intron chain of a known transcript,\n",
    "# but have a continuous chain of length n of novel events at the 3'end of the transcript\n",
    "# (n can be varied)\n",
    "# Event know from NP is fully contained within annotated intron - also set this constraint?\n",
    "\n",
    "test_novel_mult = {\"Chromosome\": [1]*4,\n",
    "                   \"Start\": [10,100,130,150],\n",
    "                   \"End\": [30,120,140,160],\n",
    "                   \"Strand\": [\"+\"]*4,\n",
    "                   \"Feature\": [\"exon\"]*4,\n",
    "                   \"gene_id\": [\"nov_gene_1\"]*4,\n",
    "                   \"transcript_id\": [\"nov_tx_mult_p\"]*4\n",
    "                    }\n",
    "\n",
    "pr.from_dict(test_novel_mult)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c68fdf5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>310</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>330</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "|            1 |       200 |       220 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "|            1 |       300 |       310 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "|            1 |       330 |       340 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 5 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 5. 3'UTR intron fully contained within an annotated 3'UTR (e.g. TDP-43)\n",
    "# For this to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "# (Annotate as a 3'UTR intron (spliced out) after filtering for intron chain match)\n",
    "test_novel_3ui = {\"Chromosome\": [1]*5,\n",
    "                \"Start\": [10,100,200,300,330],\n",
    "                \"End\": [30,120,220,310,340],\n",
    "                \"Strand\": [\"+\"]*5,\n",
    "                \"Feature\": [\"exon\"]*5,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*5,\n",
    "                \"transcript_id\": [\"nov_tx_3ui_p\"]*5\n",
    "                 }\n",
    "\n",
    "pr.from_dict(test_novel_3ui)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "883cf45e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>360</td>\n",
       "      <td>380</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-------------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id     |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)          |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-------------------|\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_exc_dist_p |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_exc_dist_p |\n",
       "|            1 |       200 |       220 | +            | exon       | nov_gene_1 | nov_tx_exc_dist_p |\n",
       "|            1 |       360 |       380 | +            | exon       | nov_gene_1 | nov_tx_exc_dist_p |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-------------------+\n",
       "Stranded PyRanges object has 4 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#6. Distal last exon spliced from penultimate exon (i.e. a mutually exclusive last exon) (e.g. SMC1A)\n",
    "# For this to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "# Will have to annotate more precisely later (i.e. differentiate from 7)\n",
    "\n",
    "test_novel_exc_dist = {\"Chromosome\": [1]*4,\n",
    "                \"Start\": [10,100,200,360],\n",
    "                \"End\": [30,120,220,380],\n",
    "                \"Strand\": [\"+\"]*4,\n",
    "                \"Feature\": [\"exon\"]*4,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*4,\n",
    "                \"transcript_id\": [\"nov_tx_exc_dist_p\"]*4\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_exc_dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "44d51aea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>310</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>360</td>\n",
       "      <td>380</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+------------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id    |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)         |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+------------------|\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |\n",
       "|            1 |       200 |       220 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |\n",
       "|            1 |       300 |       310 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |\n",
       "|            1 |       360 |       380 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+------------------+\n",
       "Stranded PyRanges object has 5 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#7. Distal last exon spliced from ann\n",
    "# For this to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "test_novel_le_dist = {\"Chromosome\": [1]*5,\n",
    "                      \"Start\": [10,100,200,300,360],\n",
    "                      \"End\": [30,120,220,310,380],\n",
    "                      \"Strand\": [\"+\"]*5,\n",
    "                      \"Feature\": [\"exon\"]*5,\n",
    "                      \"gene_id\": [\"nov_gene_1\"]*5,\n",
    "                      \"transcript_id\": [\"nov_tx_le_dist_p\"]*5\n",
    "                     }\n",
    "\n",
    "pr.from_dict(test_novel_le_dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8d51550f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>80</td>\n",
       "      <td>90</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+------------------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id          |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)               |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+------------------------|\n",
       "|            2 |        40 |        50 | -            | exon       | nov_gene_2 | nov_tx_le_dist_p_minus |\n",
       "|            2 |        80 |        90 | -            | exon       | nov_gene_2 | nov_tx_le_dist_p_minus |\n",
       "|            2 |       100 |       120 | -            | exon       | nov_gene_2 | nov_tx_le_dist_p_minus |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+------------------------+\n",
       "Stranded PyRanges object has 3 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#8. Distal last exon spliced from annoatrd (minus strand)\n",
    "# For this to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "test_novel_le_dist_minus = {\"Chromosome\": [2]*3,\n",
    "                      \"Start\": [40,80,100],\n",
    "                      \"End\": [50,90,120],\n",
    "                      \"Strand\": [\"-\"]*3,\n",
    "                      \"Feature\": [\"exon\"]*3,\n",
    "                      \"gene_id\": [\"nov_gene_2\"]*3,\n",
    "                      \"transcript_id\": [\"nov_tx_le_dist_p_minus\"]*3\n",
    "                     }\n",
    "\n",
    "pr.from_dict(test_novel_le_dist_minus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4dc1c09e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>310</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>330</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>240</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>240</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>360</td>\n",
       "      <td>380</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>35</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>310</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>360</td>\n",
       "      <td>380</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1</td>\n",
       "      <td>130</td>\n",
       "      <td>140</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1</td>\n",
       "      <td>150</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>2</td>\n",
       "      <td>80</td>\n",
       "      <td>90</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "| Chromosome   | Start     | End       | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "| 1            | 10        | 30        | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| 1            | 100       | 120       | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| 1            | 200       | 220       | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| 1            | 300       | 310       | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| ...          | ...       | ...       | ...          | ...        | ...        | ...             |\n",
       "| 1            | 140       | 160       | +            | exon       | nov_gene_1 | nov_tx_si_p     |\n",
       "| 1            | 50        | 70        | +            | exon       | nov_gene_1 | nov_tx_si_f     |\n",
       "| 1            | 100       | 120       | +            | exon       | nov_gene_1 | nov_tx_si_f     |\n",
       "| 1            | 140       | 160       | +            | exon       | nov_gene_1 | nov_tx_si_f     |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 37 rows and 7 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_novel_gr = pr.concat([pr.from_dict(event) for event in [test_novel_3ui,\n",
    "                                                            test_novel_bl,\n",
    "                                                            test_novel_exc_dist,\n",
    "                                                            test_novel_fi,\n",
    "                                                            test_novel_le_dist,\n",
    "                                                            test_novel_mult,\n",
    "                                                            test_novel_si,\n",
    "                                                            test_novel_le_dist_minus]\n",
    "                          ]\n",
    "                         )\n",
    "\n",
    "test_novel_gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6e2ae531",
   "metadata": {},
   "outputs": [],
   "source": [
    "def introns_from_df(df):\n",
    "    '''\n",
    "    '''\n",
    "    \n",
    "    n_exons = len(df)\n",
    "    \n",
    "    if n_exons < 2:\n",
    "        return None\n",
    "        #print(df)\n",
    "        #raise Exception(\"at least two exons are required for transcript to have an intron\")\n",
    "    # n exons = n-1 introns\n",
    "    \n",
    "    strand = df[\"Strand\"].drop_duplicates().tolist()[0]\n",
    "#     print(strand)\n",
    "    chrom = df[\"Chromosome\"].drop_duplicates().tolist()[0]\n",
    "    gene_id = df[\"gene_id\"].drop_duplicates().tolist()[0]\n",
    "    tx_id = df[\"transcript_id\"].drop_duplicates().tolist()[0]\n",
    "    feature = \"intron\"\n",
    "    introns = {}\n",
    "    for i in range(0, n_exons - 1):\n",
    "        if strand == \"+\":\n",
    "            intron_start = df.iloc[i, lambda x: x.columns.get_loc(\"End\")]\n",
    "            intron_end = df.iloc[i+1, lambda x: x.columns.get_loc(\"Start\")]\n",
    "            introns[str(i)] = {\"Chromosome\": chrom,\n",
    "                               \"Start\": intron_start,\n",
    "                               \"End\": intron_end,\n",
    "                               \"Strand\": strand,\n",
    "                               \"Feature\": feature,\n",
    "                               \"gene_id\": gene_id,\n",
    "                               \"transcript_id\": tx_id}\n",
    "        elif strand == \"-\":\n",
    "            intron_start = df.iloc[i, lambda x: x.columns.get_loc(\"End\")]\n",
    "            intron_end = df.iloc[i+1, lambda x: x.columns.get_loc(\"Start\")]\n",
    "            introns[str(i)] = {\"Chromosome\": chrom,\n",
    "                               \"Start\": intron_start,\n",
    "                               \"End\": intron_end,\n",
    "                               \"Strand\": strand,\n",
    "                               \"Feature\": feature,\n",
    "                               \"gene_id\": gene_id,\n",
    "                               \"transcript_id\": tx_id}\n",
    "    return pd.DataFrame.from_dict(introns, orient = \"index\")\n",
    "        \n",
    "\n",
    "\n",
    "def introns_by_tx(gr, by=\"transcript_id\", nb_cpu=1):\n",
    "    '''\n",
    "    '''\n",
    "    # Sort by position (for safety)\n",
    "    gr = gr.sort()\n",
    "    \n",
    "    return gr.apply(lambda df: df.groupby(by).apply(introns_from_df), nb_cpu=nb_cpu)\n",
    "    \n",
    "    \n",
    "# test_novel_introns = introns_by_tx(test_novel_gr)\n",
    "# test_ref_introns = introns_by_tx(test_ref)\n",
    "    \n",
    "    \n",
    "# test_novel_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7c243a35",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'test_ref_introns' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-15-2be5f70afd7d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mtest_ref_introns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'test_ref_introns' is not defined"
     ]
    }
   ],
   "source": [
    "# test_ref_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d382d27d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rle(inarray):\n",
    "        \"\"\"\n",
    "        run length encoding. Partial credit to R rle function. \n",
    "        Multi datatype arrays catered for including non Numpy\n",
    "        returns: tuple (runlengths, startpositions, values)\n",
    "        https://stackoverflow.com/questions/1066758/find-length-of-sequences-of-identical-values-in-a-numpy-array-run-length-encodi\n",
    "        Thomas Browne\n",
    "        \"\"\"\n",
    "        ia = np.asarray(inarray)                # force numpy\n",
    "        n = len(ia)\n",
    "        if n == 0: \n",
    "            return (None, None, None)\n",
    "        else:\n",
    "            y = ia[1:] != ia[:-1]               # pairwise unequal (string safe)\n",
    "            i = np.append(np.where(y), n - 1)   # must include last element posi\n",
    "            z = np.diff(np.append(-1, i))       # run lengths\n",
    "            p = np.cumsum(np.append(0, z))[:-1] # positions\n",
    "            return(z, p, ia[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32cbab88",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Dict of {novel_tx_id: {matches: [ref_id], chain_match: [n_matching], terminal_non_match: [n_not_matching]}}\n",
    "novel_info_dict = {}\n",
    "\n",
    "test_novel_gr.apply(lambda df: df.groupby(\"transcript_id\").filter(lambda grp: len(grp) > 2))\n",
    "\n",
    "for key, dfs in pr.itergrs([test_novel_introns.apply(lambda df: df.groupby(\"transcript_id\").filter(lambda grp: len(grp) > 1)), test_ref_introns], strand=True, keys=True):\n",
    "    print(\"----processing chrom and strand pair {0} & {1}\".format(key[0], key[1]))\n",
    "  \n",
    "    by_tx = tuple([df.groupby(\"transcript_id\") for df in dfs])\n",
    "    novel_txipts = by_tx[0]\n",
    "    ref_txipts = by_tx[1]\n",
    "    \n",
    "    strand = key[1]\n",
    "#     print(strand)\n",
    "#    #1. for each novel transcript, test it against all ref transcripts:\n",
    "    \n",
    "    \n",
    "    for novel_id, novel_introns in novel_txipts:\n",
    "#         print(novel_id)\n",
    "#         print(novel_introns)\n",
    "#         print(\"class of novel_introns is {}\".format(type(novel_introns)))\n",
    "        #print(exons)\n",
    "        \n",
    "        if strand == \"-\":\n",
    "            #First in order = last intron - reverse df now\n",
    "            novel_introns = novel_introns[::-1].reset_index(drop=True)\n",
    "#             ref_introns = ref_introns[::-1].reset_index(drop=True)\n",
    "            #print(exons)\n",
    "        #by_tx[1]\n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "        \n",
    "        if strand == \"+\":\n",
    "            first_matches = [np.array_equal(novel_introns.head(1)[[\"Start\",\"End\"]],\n",
    "                                            ref_introns.head(1)[[\"Start\",\"End\"]]) for ref_id,ref_introns in ref_txipts]\n",
    "        else:\n",
    "            first_matches = [np.array_equal(novel_introns.head(1)[[\"Start\",\"End\"]],\n",
    "                                            ref_introns[::-1].reset_index(drop=True) # rev order\n",
    "                                            .head(1)[[\"Start\",\"End\"]]\n",
    "                                           )\n",
    "                             for ref_id,ref_introns in ref_txipts]\n",
    "\n",
    "        #print(novel_id)\n",
    "#         print(first_matches)\n",
    "        if not sum(first_matches) > 0:\n",
    "            print(\"{0} does not match any reference transcripts in its first intron. Skipping\".format(novel_id))\n",
    "            continue\n",
    "        \n",
    "        # Now compare novel transcript against each ref transcript with match in first intron.\n",
    "        # \n",
    "        for ref_tr, first_match in zip(ref_txipts, first_matches):\n",
    "            \n",
    "            if not first_match:\n",
    "                continue\n",
    "            #intron_chain_match = np.equal(np.asarray(exons[[\"Start\",\"End\"]]), np.asarray(ref_tr[1][[\"Start\",\"End\"]]))\n",
    "            ref_introns = ref_tr[1]\n",
    "            \n",
    "            if strand == \"-\":\n",
    "                ref_introns = ref_introns[::-1].reset_index(drop=True) # rev order\n",
    "            \n",
    "            n_novel_introns = len(novel_introns)\n",
    "            n_ref_introns = len(ref_introns)\n",
    "            \n",
    "            # To avoid a slicing error for txipts shorter than novel\n",
    "            max_chain = min(n_novel_introns, n_ref_introns)\n",
    "#             print(max_chain)\n",
    "#             print(ref_tr[1].iloc[0,])\n",
    "#             print(n_novel_introns)\n",
    "#             print(len(ref_tr[1]))\n",
    "            novel_chain_match = pd.DataFrame([(novel_introns.iloc[i,:][[\"Start\",\"End\"]]\n",
    "                                               .eq(ref_introns.iloc[i,:][[\"Start\",\"End\"]]\n",
    "                                                  )\n",
    "                                              )\n",
    "                                              for i in range(max_chain)])\n",
    "    \n",
    "            # Collapse to single True/False per row - does intron match?\n",
    "            novel_chain_match = novel_chain_match.apply(np.all, axis=1, raw=True)\n",
    "            \n",
    "#             print(novel_chain_match)\n",
    "            \n",
    "            runs, starts, vals = rle(novel_chain_match)\n",
    "            #print(runs[0])\n",
    "            #print(starts)\n",
    "            #print(\"\\n {0}\".format(np.where(vals == False)))\n",
    "            \n",
    "            if np.all(vals) or np.array_equal(vals, [True,False]):\n",
    "                # Possible genuine match, should update dict with info\n",
    "                \n",
    "                # match (from start of ref Txipt)\n",
    "                \n",
    "                if vals.size == 1:\n",
    "                    #i.e. all true/introns match (e.g. bleedthrough/ where )\n",
    "                    terminal_non_match = 0\n",
    "                else:\n",
    "                    terminal_non_match = runs[1]\n",
    "                \n",
    "#                 print(terminal_non_match)\n",
    "                \n",
    "                if novel_id not in novel_info_dict:\n",
    "                    \n",
    "                    novel_info_dict[novel_id] = {\"matches\": [ref_tr[0]],\n",
    "                                                 \"chain_match\": [runs[0]], #Always starts with true, so take length of true\n",
    "                                                 \"terminal_non_match\": [terminal_non_match],\n",
    "                                                }\n",
    "                \n",
    "                else:\n",
    "                    # Append to dict\n",
    "                    novel_info_dict[novel_id][\"matches\"].append(ref_tr[0])\n",
    "                    novel_info_dict[novel_id][\"chain_match\"].append(ref_tr[0])\n",
    "                    novel_info_dict[novel_id][\"terminal_non_match\"].append(ref_tr[0])\n",
    "                    \n",
    "                \n",
    "            else:\n",
    "                continue\n",
    "            \n",
    "            \n",
    "            \n",
    "print(novel_info_dict)\n",
    "#             intron_chain_match = exons[[\"Start\",\"End\"]].eq(ref_tr[1][[\"Start\",\"End\"]])\n",
    "#             print(\"\\n\")\n",
    "#             print(intron_chain_match)\n",
    "#             print(intron_chain_match.apply(np.all, axis = 1))\n",
    "            \n",
    "            \n",
    "                #loc[np.where(lambda x: (x[[\"Start\", \"End\"]].all()))])\n",
    "                                                  #apply(lambda x: all(x), axis = \"index\", raw = True))\n",
    "#             print(intron_chain_match)\n",
    "            \n",
    "            \n",
    "#         first_intron = exons.head(1)[[\"Start\",\"End\"]] if strand == \"+\" else exons.tail(1)[[\"Start\",\"End\"]]\n",
    "        \n",
    "#         grp_filter_matches = by_tx[1].filter(lambda x: np.array_equal(first_intron, x.head(1)[[\"Start\", \"End\"]]))\n",
    "#         print(grp_filter_matches)\n",
    "#         print(type(grp_filter_matches))\n",
    "     #for novel, ref_tr in zip(by_tx[0],by_tx[1]:\n",
    "#         print(nov_tr[1])\n",
    "#         print(ref_tr)\n",
    "             \n",
    "pd.concat({key: pd.DataFrame.from_dict(d, orient = \"columns\") for key, d in novel_info_dict.items()}, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54913d65",
   "metadata": {},
   "outputs": [],
   "source": [
    "def match_intron_chains(novel_gr, ref_gr, id_col = \"transcript_id\", nb_cpu = 1):\n",
    "    '''\n",
    "    grs should contain introns\n",
    "    '''\n",
    "    # {novel_id: {matches: [ref_id], chain_match: [n], terminal_non_match: [n]}}\n",
    "    match_info_dict = {}\n",
    "    \n",
    "\n",
    "    \n",
    "    for key, dfs in pr.itergrs([novel_gr, ref_gr], strand=True, keys=True):\n",
    "        print(\"----processing chrom and strand pair {0} & {1}\".format(key[0], key[1]))\n",
    "        # dfs = novel_gr & ref_gr matched by key (chromosome & strand)\n",
    "        # pandas group by objects of (transcript_id, df)\n",
    "        by_tx = tuple([df.groupby(\"transcript_id\") for df in dfs])\n",
    "        \n",
    "        novel_txipts = by_tx[0]\n",
    "        ref_txipts = by_tx[1]\n",
    "        \n",
    "        # Pyranges keys are tuples of (chr,strand)\n",
    "        strand = key[1]\n",
    "        \n",
    "        \n",
    "        #Comparing each novel transcript against all ref transcripts\n",
    "        for novel_id, novel_introns in novel_txipts:\n",
    "        \n",
    "            if strand == \"-\":\n",
    "                # Standard PyRanges sort - First in df = last intron (smallest values i.e. leftmost)\n",
    "                # Reverse so first row in df is always the first intron\n",
    "                # is reset_index necessary?\n",
    "                novel_introns = novel_introns[::-1].reset_index(drop=True)\n",
    "            \n",
    "            else:\n",
    "                pass\n",
    "        \n",
    "            # As a first pass, check for matches in intron chain of first introns between novel and reference Txs\n",
    "            if strand == \"+\":\n",
    "                first_matches = [np.array_equal(novel_introns.head(1)[[\"Start\",\"End\"]],\n",
    "                                                ref_introns.head(1)[[\"Start\",\"End\"]]) \n",
    "                                 for ref_id,ref_introns in ref_txipts]\n",
    "            else:\n",
    "                first_matches = [np.array_equal(novel_introns.head(1)[[\"Start\",\"End\"]],\n",
    "                                                ref_introns[::-1].reset_index(drop=True) # rev order so 1st row = 1st intron\n",
    "                                                .head(1)[[\"Start\",\"End\"]]\n",
    "                                               )\n",
    "                                 for ref_id,ref_introns in ref_txipts]\n",
    "\n",
    "\n",
    "            if not sum(first_matches) > 0:\n",
    "                print(\"{0} does not match any reference transcripts in its first intron. Skipping\".format(novel_id))\n",
    "                continue\n",
    "        \n",
    "            # Compare full introns chains of novel transcript against each ref transcript with match in first intron.\n",
    "        \n",
    "            for ref_tr, first_match in zip(ref_txipts, first_matches):\n",
    "            \n",
    "                if not first_match:\n",
    "                    continue\n",
    "                \n",
    "            \n",
    "                ref_id = ref_tr[0]\n",
    "#                 print(\"ref_id {}\".format(ref_id))\n",
    "                ref_introns = ref_tr[1]\n",
    "            \n",
    "                if strand == \"-\":\n",
    "                    ref_introns = ref_introns[::-1].reset_index(drop=True) # reverse so first row = first intron\n",
    "                        \n",
    "                # To avoid a slicing error for ref txipts shorter than novel\n",
    "                n_novel_introns = len(novel_introns)\n",
    "                n_ref_introns = len(ref_introns)\n",
    "                max_chain = min(n_novel_introns, n_ref_introns)\n",
    "#             print(max_chain)\n",
    "#             print(ref_tr[1].iloc[0,])\n",
    "#             print(n_novel_introns)\n",
    "#             print(len(ref_tr[1]))\n",
    "\n",
    "                # Row-wise, check whether match with corresponding intron of reference transcript\n",
    "                novel_chain_match = pd.DataFrame([(novel_introns.iloc[i,:][[\"Start\",\"End\"]]\n",
    "                                                   .eq(ref_introns.iloc[i,:][[\"Start\",\"End\"]]\n",
    "                                                      )\n",
    "                                                  )\n",
    "                                                  for i in range(max_chain)])\n",
    "    \n",
    "                # Collapse to single True/False per row - does intron completely match?\n",
    "                novel_chain_match = novel_chain_match.apply(np.all, axis=1, raw=True)\n",
    "            \n",
    "#             print(novel_chain_match)\n",
    "            \n",
    "                runs, starts, vals = rle(novel_chain_match)\n",
    "            #print(runs[0])\n",
    "            #print(starts)\n",
    "            #print(\"\\n {0}\".format(np.where(vals == False)))\n",
    "            \n",
    "                # Considered a valid match if a intron chain completely identical \n",
    "                # or matches at beginning of txipt but differs at the end\n",
    "                if np.all(vals) or np.array_equal(vals, [True,False]):\n",
    "                \n",
    "                    # Don't want to throw away all valid matches (yet)\n",
    "                    # Possible genuine match, should update dict with info\n",
    "                    # match (from start of ref Txipt)\n",
    "                \n",
    "                    if vals.size == 1:\n",
    "                        #i.e. all true/introns match (e.g. bleedthrough)\n",
    "                        terminal_non_match = 0\n",
    "                    else:\n",
    "                        # All valid = consective match & non-match, so non-match = 2nd in array \n",
    "                        terminal_non_match = runs[1]\n",
    "                \n",
    "#                 print(terminal_non_match)\n",
    "                \n",
    "                    if novel_id not in match_info_dict:\n",
    "                    \n",
    "                        match_info_dict[novel_id] = {\"matches\": [ref_id],\n",
    "                                                     \"chain_match\": [runs[0]], #Always starts with true, so take length of true\n",
    "                                                     \"terminal_non_match\": [terminal_non_match],\n",
    "                                                    }\n",
    "                \n",
    "                    else:\n",
    "                        # Append to dict\n",
    "                        match_info_dict[novel_id][\"matches\"].append(ref_id)\n",
    "                        match_info_dict[novel_id][\"chain_match\"].append(runs[0])\n",
    "                        match_info_dict[novel_id][\"terminal_non_match\"].append(terminal_non_match)\n",
    "                    \n",
    "                \n",
    "                else:\n",
    "                    continue\n",
    "    \n",
    "    # Output df for easier parsing\n",
    "#     match_df = pd.concat({key: pd.DataFrame.from_dict(d, orient = \"columns\") for key, d in match_info_dict.items()}, axis=0)\n",
    "    \n",
    "    return match_info_dict\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca48baec",
   "metadata": {},
   "outputs": [],
   "source": [
    "match_intron_chains(test_novel_introns, test_ref_introns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf33eff0",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.array_equal(np.array([True, False]), [True, False])\n",
    "np.all([True,True])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb031093",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test_introns = pr.from_dict(test_novel_3ui)#.subset(lambda df: df[\"transcript_id\"] == \"nov_tx_3ui_p\")\n",
    "\n",
    "# n_exons = len(test_introns)\n",
    "# # n exons = n -1 introns\n",
    "\n",
    "# # plus strand txipt - End of first exon = start of intron, Start of 3'exon = end coord\n",
    "\n",
    "# test_introns.as_df().iloc[0,lambda x: x.columns.get_loc(\"End\")]\n",
    "\n",
    "# def introns_from_df(df):\n",
    "#     '''\n",
    "#     '''\n",
    "    \n",
    "#     n_exons = len(df)\n",
    "    \n",
    "#     if n_exons < 2:\n",
    "#         raise Exception(\"at least two exons are required for transcript to have an intron\")\n",
    "#     # n exons = n-1 introns\n",
    "    \n",
    "#     strand = df[\"Strand\"].drop_duplicates().tolist()[0]\n",
    "# #     print(strand)\n",
    "#     chrom = df[\"Chromosome\"].drop_duplicates().tolist()[0]\n",
    "#     gene_id = df[\"gene_id\"].drop_duplicates().tolist()[0]\n",
    "#     tx_id = df[\"transcript_id\"].drop_duplicates().tolist()[0]\n",
    "#     feature = \"intron\"\n",
    "#     introns = {}\n",
    "#     for i in range(0, n_exons - 1):\n",
    "#         if strand == \"+\":\n",
    "#             intron_start = df.iloc[i, lambda x: x.columns.get_loc(\"End\")]\n",
    "#             intron_end = df.iloc[i+1, lambda x: x.columns.get_loc(\"Start\")]\n",
    "#             introns[str(i)] = {\"Chromosome\": chrom,\n",
    "#                                \"Start\": intron_start,\n",
    "#                                \"End\": intron_end,\n",
    "#                                \"Strand\": strand,\n",
    "#                                \"Feature\": feature,\n",
    "#                                \"gene_id\": gene_id,\n",
    "#                                \"transcript_id\": tx_id}\n",
    "#         elif strand == \"-\":\n",
    "#             intron_end = df.iloc[i, lambda x: x.columns.get_loc(\"Start\")]\n",
    "#             intron_start = df.iloc[i+1, lambda x: x.columns.get_loc(\"End\")]\n",
    "#             introns[str(i)] = {\"Chromosome\": chrom,\n",
    "#                                \"Start\": intron_start,\n",
    "#                                \"End\": intron_end,\n",
    "#                                \"Strand\": strand,\n",
    "#                                \"Feature\": feature,\n",
    "#                                \"gene_id\": gene_id,\n",
    "#                                \"transcript_id\": tx_id}\n",
    "#     return pd.DataFrame.from_dict(introns, orient = \"index\")\n",
    "        \n",
    "\n",
    "\n",
    "# def introns_by_tx(gr, by=\"transcript_id\", nb_cpu=1):\n",
    "#     '''\n",
    "#     '''\n",
    "    \n",
    "#     return gr.apply(lambda df: df.groupby(by).apply(introns_from_df), nb_cpu=1)\n",
    "    \n",
    "    \n",
    "# test_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "758737f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "introns_by_tx(test_introns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf44cbed",
   "metadata": {},
   "outputs": [],
   "source": [
    "introns_by_tx(pr.data.ensembl_gtf().subset(lambda df: (df.gene_id == \"ENSG00000205231\") & (df.Feature == \"exon\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5b71b28",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "pr.data.ensembl_gtf().subset(lambda df: df.transcript_id == \"ENST00000450305\")[[\"transcript_id\",\"Feature\"]].features.introns(by=\"transcript\").dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7e4891a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test_ref.features.introns(by=\"transcript\")\n",
    "\n",
    "(pr.data.ensembl_gtf().subset(lambda df: df.transcript_id == \"ENST00000450305\")\n",
    "#  [[\"gene_id\",\"transcript_id\"]]\n",
    " .features.introns(by=\"transcript\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c86eba9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pr.data.ensembl_gtf().subset(lambda df: (df.gene_id == \"ENSG00000205231\") & (df.Feature == \"exon\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fa672ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "pr.data.ensembl_gtf().subset(lambda df: (df.gene_id == \"ENSG00000205231\")).features.introns(by=\"transcript\")\n",
    "#Start of first feature, end of next along (if properly sorted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8578c48f",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_stie_gtf_chr1_nov = \"../two_sample_example_output/stringtie/chr1.no_ref_id.TDP43-F_S6.assembled.gtf\"\n",
    "path_ref_gtf = \"../data/annotation/gencode.v34.annotation.gtf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8a2eb6cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>cov</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>TPM</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>827750</td>\n",
       "      <td>859332</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>98.466469</td>\n",
       "      <td>11.679049</td>\n",
       "      <td>30.310225</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>827750</td>\n",
       "      <td>827775</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>96.734062</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>829002</td>\n",
       "      <td>829104</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>110.622597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>847653</td>\n",
       "      <td>847806</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>118.018204</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>851926</td>\n",
       "      <td>852110</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>115.267273</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26008</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248855557</td>\n",
       "      <td>248855635</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.161189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26009</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248855724</td>\n",
       "      <td>248855943</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>4.278336</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26010</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248856287</td>\n",
       "      <td>248856422</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.487612</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26011</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248856513</td>\n",
       "      <td>248856562</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.115693</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26012</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248857233</td>\n",
       "      <td>248858865</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>13.537362</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26013 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | Score     | Strand       | Frame      | gene_id    | transcript_id   | cov        | FPKM       | TPM        | +1    |\n",
       "| (category)   | (object)   | (category)   | (int32)   | (int32)   | (int64)   | (category)   | (object)   | (object)   | (object)        | (object)   | (object)   | (object)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+-------|\n",
       "| chr1         | StringTie  | transcript   | 827750    | 859332    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 98.466469  | 11.679049  | 30.310225  | ...   |\n",
       "| chr1         | StringTie  | exon         | 827750    | 827775    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 96.734062  | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 829002    | 829104    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 110.622597 | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 847653    | 847806    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 118.018204 | nan        | nan        | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...       | ...          | ...        | ...        | ...             | ...        | ...        | ...        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248855724 | 248855943 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 4.278336   | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248856287 | 248856422 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 5.487612   | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248856513 | 248856562 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 5.115693   | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248857233 | 248858865 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 13.537362  | nan        | nan        | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+-------+\n",
       "Stranded PyRanges object has 26,013 rows and 14 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: exon_number"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stie_chr1 = pr.read_gtf(path_stie_gtf_chr1_nov)\n",
    "stie_chr1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c443d1a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2912491</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214349</td>\n",
       "      <td>57214397</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002072208.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2912492</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213879</td>\n",
       "      <td>57213964</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002046926.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2912493</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213525</td>\n",
       "      <td>57213602</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002021169.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2912494</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213203</td>\n",
       "      <td>57213357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00002036959.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2912495</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57213125</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00002023900.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2912496 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------------------------+------------------------------------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | Score      | Strand       | Frame      | gene_id                 | gene_type                          | +15   |\n",
       "| (category)   | (object)   | (object)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)                | (object)                           | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------------------------+------------------------------------+-------|\n",
       "| chr1         | HAVANA     | gene       | 11868     | 14409     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| chr1         | HAVANA     | transcript | 11868     | 14409     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| chr1         | HAVANA     | exon       | 11868     | 12227     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| chr1         | HAVANA     | exon       | 12612     | 12721     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...        | ...          | ...        | ...                     | ...                                | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57213879  | 57213964  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57213525  | 57213602  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57213203  | 57213357  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57212183  | 57213125  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------------------------+------------------------------------+-------+\n",
       "Stranded PyRanges object has 2,912,496 rows and 25 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "15 hidden columns: gene_name, level, hgnc_id, havana_gene, transcript_id, transcript_type, transcript_name, transcript_support_level, tag, havana_transcript, exon_number, exon_id, ont, ... (+ 2 more.)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref = pr.read_gtf(path_ref_gtf)\n",
    "ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4f516593",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>cov</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>TPM</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>827750</td>\n",
       "      <td>827775</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>96.734062</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>829002</td>\n",
       "      <td>829104</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>110.622597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>847653</td>\n",
       "      <td>847806</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>118.018204</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>851926</td>\n",
       "      <td>852110</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>115.267273</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>852670</td>\n",
       "      <td>852766</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>135.545364</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23563</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248855557</td>\n",
       "      <td>248855635</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.161189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23564</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248855724</td>\n",
       "      <td>248855943</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>4.278336</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23565</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248856287</td>\n",
       "      <td>248856422</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.487612</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23566</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248856513</td>\n",
       "      <td>248856562</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.115693</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23567</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248857233</td>\n",
       "      <td>248858865</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>13.537362</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23568 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | Score     | Strand       | Frame      | gene_id    | transcript_id   | cov        | FPKM       | TPM        | +1    |\n",
       "| (category)   | (object)   | (category)   | (int32)   | (int32)   | (int64)   | (category)   | (object)   | (object)   | (object)        | (object)   | (object)   | (object)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+-------|\n",
       "| chr1         | StringTie  | exon         | 827750    | 827775    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 96.734062  | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 829002    | 829104    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 110.622597 | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 847653    | 847806    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 118.018204 | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 851926    | 852110    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 115.267273 | nan        | nan        | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...       | ...          | ...        | ...        | ...             | ...        | ...        | ...        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248855724 | 248855943 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 4.278336   | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248856287 | 248856422 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 5.487612   | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248856513 | 248856562 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 5.115693   | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248857233 | 248858865 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 13.537362  | nan        | nan        | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+-------+\n",
       "Stranded PyRanges object has 23,568 rows and 14 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: exon_number"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stie_chr1_exons = stie_chr1.subset(lambda df: df[\"Feature\"] == \"exon\")\n",
    "stie_chr1_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "23078b88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000002844.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000002844.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378883</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214349</td>\n",
       "      <td>57214397</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002072208.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378884</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213879</td>\n",
       "      <td>57213964</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002046926.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378885</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213525</td>\n",
       "      <td>57213602</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002021169.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378886</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213203</td>\n",
       "      <td>57213357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00002036959.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378887</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57213125</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00002023900.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1378888 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------------------------+------------------------------------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | Score      | Strand       | Frame      | gene_id                 | gene_type                          | +15   |\n",
       "| (category)   | (object)   | (object)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)                | (object)                           | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------------------------+------------------------------------+-------|\n",
       "| chr1         | HAVANA     | exon       | 11868     | 12227     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| chr1         | HAVANA     | exon       | 12612     | 12721     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| chr1         | HAVANA     | exon       | 13220     | 14409     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| chr1         | HAVANA     | exon       | 12009     | 12057     | .          | +            | .          | ENSG00000223972.5       | transcribed_unprocessed_pseudogene | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...        | ...          | ...        | ...                     | ...                                | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57213879  | 57213964  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57213525  | 57213602  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57213203  | 57213357  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "| chrY         | HAVANA     | exon       | 57212183  | 57213125  | .          | -            | .          | ENSG00000227159.8_PAR_Y | unprocessed_pseudogene             | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------------------------+------------------------------------+-------+\n",
       "Stranded PyRanges object has 1,378,888 rows and 25 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "15 hidden columns: gene_name, level, hgnc_id, havana_gene, transcript_id, transcript_type, transcript_name, transcript_support_level, tag, havana_transcript, exon_number, exon_id, ont, ... (+ 2 more.)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_exons = ref.subset(lambda df: df[\"Feature\"] == \"exon\")\n",
    "ref_chr1_exons = ref_exons.subset(lambda df: df[\"Chromosome\"] == \"chr1\")\n",
    "ref_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "44b602ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>29553</td>\n",
       "      <td>30039</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000002840.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001947070.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>30563</td>\n",
       "      <td>30667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000002840.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001922571.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>30975</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000002840.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001827679.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>30266</td>\n",
       "      <td>30667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-201</td>\n",
       "      <td>5</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000002841.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001841699.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>30975</td>\n",
       "      <td>31109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-201</td>\n",
       "      <td>5</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000002841.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001890064.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121907</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248857668</td>\n",
       "      <td>248858017</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000171163.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>ZNF692-218</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000097304.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001904444.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121908</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248859014</td>\n",
       "      <td>248859144</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000171163.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>ZNF692-229</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000382605.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002142255.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121909</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248857864</td>\n",
       "      <td>248858321</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000171163.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>ZNF692-229</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000382605.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002187693.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121910</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248858917</td>\n",
       "      <td>248859033</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000171163.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>ZNF692-228</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000382606.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002188413.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121911</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248857974</td>\n",
       "      <td>248858385</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000171163.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>ZNF692-228</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMT00000382606.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002170424.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121912 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | Score      | Strand       | Frame      | gene_id            | gene_type      | gene_name   | level      | +13   |\n",
       "| (category)   | (object)   | (object)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)           | (object)       | (object)    | (object)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+-------|\n",
       "| chr1         | HAVANA     | exon       | 29553     | 30039     | .          | +            | .          | ENSG00000243485.5  | lncRNA         | MIR1302-2HG | 2          | ...   |\n",
       "| chr1         | HAVANA     | exon       | 30563     | 30667     | .          | +            | .          | ENSG00000243485.5  | lncRNA         | MIR1302-2HG | 2          | ...   |\n",
       "| chr1         | HAVANA     | exon       | 30975     | 31097     | .          | +            | .          | ENSG00000243485.5  | lncRNA         | MIR1302-2HG | 2          | ...   |\n",
       "| chr1         | HAVANA     | exon       | 30266     | 30667     | .          | +            | .          | ENSG00000243485.5  | lncRNA         | MIR1302-2HG | 2          | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...        | ...          | ...        | ...                | ...            | ...         | ...        | ...   |\n",
       "| chr1         | HAVANA     | exon       | 248859014 | 248859144 | .          | -            | .          | ENSG00000171163.16 | protein_coding | ZNF692      | 2          | ...   |\n",
       "| chr1         | HAVANA     | exon       | 248857864 | 248858321 | .          | -            | .          | ENSG00000171163.16 | protein_coding | ZNF692      | 2          | ...   |\n",
       "| chr1         | HAVANA     | exon       | 248858917 | 248859033 | .          | -            | .          | ENSG00000171163.16 | protein_coding | ZNF692      | 2          | ...   |\n",
       "| chr1         | HAVANA     | exon       | 248857974 | 248858385 | .          | -            | .          | ENSG00000171163.16 | protein_coding | ZNF692      | 2          | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+--------------------+----------------+-------------+------------+-------+\n",
       "Stranded PyRanges object has 121,912 rows and 25 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "13 hidden columns: hgnc_id, havana_gene, transcript_id, transcript_type, transcript_name, transcript_support_level, tag, havana_transcript, exon_number, exon_id, ont, protein_id, ccdsid"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#prot coding / lncRNA\n",
    "# print(ref_chr1_exons.gene_type.value_counts())\n",
    "ref_chr1_pc_exons = ref_chr1_exons.subset(lambda df: df.gene_type.isin([\"protein_coding\", \"lncRNA\"]))\n",
    "ref_chr1_pc_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46be9ffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# stie_chr1_introns = introns_by_tx(stie_chr1_exons,nb_cpu=1)\n",
    "# stie_chr1_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f8edcbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ref_all_introns = introns_by_tx(ref_exons, nb_cpu=4)\n",
    "# ref_all_introns\n",
    "# ref_chr1_introns = introns_by_tx(ref_chr1_pc_exons, nb_cpu=2)\n",
    "# ref_chr1_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29d16f4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr1_chain_matching = match_intron_chains(stie_chr1_introns, ref_chr1_introns)\n",
    "# chr1_chain_matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c6c96ea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9d6e3ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ref_chr1_introns.subset(lambda df: df.transcript_id == \"ENST00000674495.1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a19a84a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ref_chr1_exons.subset(lambda df: df.transcript_id == \"ENST00000674495.1\").sort()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12c9755b",
   "metadata": {},
   "source": [
    "### Alternative (hopefully more scalable approach)\n",
    "\n",
    "Try and use pyranges internals as much as possible, and avoid manual looping and comparisons\n",
    "\n",
    "Essentially, Find overlapping introns, then filter for those that are identical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a29094ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "def intron_id(gr):\n",
    "\n",
    "    return gr.assign(\"intron_id\",\n",
    "                                         lambda df: pd.Series([\":\".join([tx_id, str(start), str(end)])\n",
    "                                                     for tx_id, start, end in zip(df[\"transcript_id\"],\n",
    "                                                                                  df[\"Start\"],\n",
    "                                                                                  df[\"End\"])\n",
    "                                                    ])\n",
    "                                        )\n",
    "\n",
    "def sort_introns_by_strand(df):\n",
    "    '''\n",
    "    '''\n",
    "    # first reset_index call removes the original index of the group (e.g. row 4005 in df)\n",
    "    # second reset_index call adds the sorted index as a column to the dataframe (the order along exon in each transcript)\n",
    "    if (df.Strand == '+').all():\n",
    "        return df.sort_values(by=['End']).reset_index(drop=True).reset_index()\n",
    "    elif (df.Strand == '-').all():\n",
    "        return df.sort_values(by=['Start'], ascending=False).reset_index(drop=True).reset_index()\n",
    "\n",
    "\n",
    "    \n",
    "def validate_matching_chain(df, max_terminal_non_match=2):\n",
    "    '''\n",
    "    apply to grouped df\n",
    "    '''\n",
    "    \n",
    "    runs, starts, vals = rle(df[\"match\"])\n",
    "    \n",
    "    # Valid matches are:\n",
    "    # All introns match (e.g. bleedthrough event)\n",
    "    # All but last x introns match (usually one) (i.e. runs = 1,0)\n",
    "    \n",
    "    if np.all(vals):\n",
    "        # All introns match (e.g. bleedthrough event)\n",
    "        return True\n",
    "    \n",
    "    elif np.array_equal(vals, [1,0]) and runs[-1] <= max_terminal_non_match:\n",
    "        # all but last x introns match (x = max_terminal_non_match) (i.e. runs = 1,0)\n",
    "        return True\n",
    "    \n",
    "    else:\n",
    "        return False\n",
    "\n",
    "    \n",
    "def filter_transcripts_by_chain(novel_exons,ref_exons, match_type = \"transcript\", max_terminal_non_match=2, nb_cpu = 1):\n",
    "    '''\n",
    "    '''\n",
    "\n",
    "    novel_cols_to_keep = [\"Feature\",\"transcript_id\"]\n",
    "    ref_cols_to_keep = [\"Feature\", \"transcript_id\", \"gene_id\", \"gene_name\"]\n",
    "\n",
    "    assert match_type in [\"transcript\", \"any\"], \"match_type must be one of 'transcript' or 'any'. value passed - {}\".format(str(match_type))\n",
    "\n",
    "    #1. Find introns by transcript & give each intron a unique ID\n",
    "    print(\"finding introns...\")\n",
    "    t1 = timer()\n",
    "    \n",
    "    novel_introns = introns_by_tx(novel_exons, nb_cpu=nb_cpu).sort()\n",
    "    ref_introns = introns_by_tx(ref_exons, nb_cpu=nb_cpu).sort()\n",
    "\n",
    "    t2 = timer()\n",
    "    \n",
    "    print(\"took {} (s)\".format(t2 - t1))\n",
    "    \n",
    "    print(\"adding intron_id column...\")\n",
    "    \n",
    "    t3 = timer()\n",
    "    novel_introns = intron_id(novel_introns)\n",
    "    ref_introns = intron_id(ref_introns)\n",
    "    t4 = timer()\n",
    "    \n",
    "    print(\"took {} s\".format(t4 - t3))\n",
    "\n",
    "    #2. Track number of introns in each novel transcript\n",
    "    novel_tx_intron_counts = (novel_introns.as_df()\n",
    "                              .groupby(\"transcript_id\").size())\n",
    "\n",
    "\n",
    "    # novel_introns, ref_introns\n",
    "\n",
    "    # 3. Store intron_ids for each transcript, sorted by intron_number (where 1 = first intron regardless of strand) in a df/Series\n",
    "    print(\"generating df of novel txipts sorted by intron number...\")\n",
    "    \n",
    "    t5 = timer()\n",
    "    \n",
    "    novel_intron_ids_ordered = (novel_introns.as_df()\n",
    "                                .groupby(\"transcript_id\")\n",
    "                                .apply(sort_introns_by_strand)\n",
    "                                .reset_index(drop=True)\n",
    "                                .rename({\"index\": \"intron_number\"}, axis=\"columns\")\n",
    "                               )\n",
    "    novel_intron_ids_ordered[\"intron_number\"] = novel_intron_ids_ordered[\"intron_number\"].add(1)\n",
    "\n",
    "    # df of txipt_id | intron_id | intron_number\n",
    "    novel_intron_ids_ordered = novel_intron_ids_ordered.loc[:,[\"transcript_id\",\"intron_id\",\"intron_number\"]]\n",
    "    \n",
    "    t6 = timer()\n",
    "    print(\"took {} s\".format(t6 - t5))\n",
    "#     print(novel_intron_ids_ordered.dtypes)\n",
    "\n",
    "\n",
    "    #4. Find novel introns with any overlap with reference introns\n",
    "    # Inner join to add ref_rows to novel gr\n",
    "    print(\"finding overlaps between novel and reference introns...\")\n",
    "    \n",
    "    t7 = timer()\n",
    "    joined = novel_introns.join(ref_introns, strandedness=\"same\", suffix =\"_ref\",nb_cpu=nb_cpu)\n",
    "    t8 = timer()\n",
    "    \n",
    "    print(\"took {} s\".format(t8 - t7))\n",
    "    \n",
    "    #5. Filter for overlaps that exactly match (or differ by given tolerance)\n",
    "    print(\"filtering overlaps for exact matches...\")\n",
    "    \n",
    "    t9 = timer()\n",
    "    joined = joined.subset(lambda df: abs(df.Start - df.Start_ref) + abs(df.End - df.End_ref) <= 0, nb_cpu=nb_cpu)\n",
    "    t10 = timer()\n",
    "    \n",
    "    print(\"took {} s\".format(t10 - t9))\n",
    "    \n",
    "    # Minimal info needed on matches between novel and reference introns\n",
    "    joined = joined.as_df()[[\"transcript_id\",\"intron_id\",\"transcript_id_ref\",\"intron_id_ref\"]]\n",
    "\n",
    "#     print(joined.dtypes)\n",
    "\n",
    "    #6. Join ordered novel introns with match info\n",
    "    #7. Assign a simple tracker column 'match' of True (where intron is matched) and False (where intron is not matched)\n",
    "    \n",
    "    print(\"preparing for filtering intron matches...\")\n",
    "    t11 = timer()\n",
    "    \n",
    "    if match_type == \"any\":\n",
    "        # Looking for intron to match any annotated intron, regardless of reference transcript\n",
    "        novel_ref_match_info = novel_intron_ids_ordered.merge(joined,\n",
    "                                                              how=\"left\",\n",
    "                                                              on=\"intron_id\",\n",
    "                                                              suffixes=[\"_novel\",\"_match\"]\n",
    "                                                             )\n",
    "        \n",
    "        # Assign 'match' column for each intron.\n",
    "        # Since we don't really care which intron it matches, & no matches will mean NaN\n",
    "        novel_ref_match_info[\"match\"] = novel_ref_match_info[\"transcript_id_ref\"]\n",
    "        novel_ref_match_info[\"match\"] = novel_ref_match_info[\"match\"].fillna(0)\n",
    "        novel_ref_match_info[\"match\"] = novel_ref_match_info[\"match\"].replace(\"\\w*\", 1, regex=True)\n",
    "        \n",
    "        novel_ref_match_info = novel_ref_match_info.drop_duplicates(subset=[\"intron_id\"])\n",
    "        \n",
    "        # Minimal informative info is novel tx, novel intron_id & number, match column\n",
    "        novel_ref_match_info = novel_ref_match_info[[\"transcript_id_novel\",\"intron_id\",\"intron_number\",\"match\"]]\n",
    "        \n",
    "\n",
    "    elif match_type == \"transcript\":\n",
    "        # Looking for introns (except last) to match the same reference transcript\n",
    "        # merge_ordered can do a 'grouped merge' filling in empty rows (introns) for each transcript_id\n",
    "        # This is especially useful if want to do transcript-specific intron matching\n",
    "        # For each reference transcript, all novel introns will be filled with NaN if no overlap for given transcript_id\n",
    "        # (i.e. novel txipt matches all but last intron of reference transcript)\n",
    "\n",
    "        novel_ref_match_info = (pd.merge_ordered(novel_intron_ids_ordered,\n",
    "                                    joined,\n",
    "                                    how=\"left\",\n",
    "                                    on=\"intron_id\",\n",
    "                                    right_by=\"transcript_id_ref\", # group matches by ref tx & join tx by tx\n",
    "                                    suffixes=[\"_novel\",\"_match\"],\n",
    "                                    fill_method=None)\n",
    "                   .sort_values(by=[\"transcript_id_novel\",\"intron_number\"])\n",
    "                               )\n",
    "        \n",
    "        # merge_ordered fills rows for each intron for each ref tx in df, regardless of whether any overlap\n",
    "        # .dropna(axis=\"rows\", subset=[\"intron_id_ref\"]) \n",
    "        novel_ref_match_info = (novel_ref_match_info.groupby([\"transcript_id_novel\", \"transcript_id_ref\"])\n",
    "                                .filter(lambda df: (df[\"intron_id_ref\"].notna()).any()) # Retained if ref tx has >=1 matching introns\n",
    "                                .reset_index(drop=True))\n",
    "        \n",
    "        # Make a match column where 1 = match, 0 = no match for each ref id and novel intron\n",
    "        novel_ref_match_info[\"match\"] = novel_ref_match_info[\"intron_id_ref\"]\n",
    "        novel_ref_match_info[\"match\"] = novel_ref_match_info[\"match\"].fillna(0)\n",
    "        novel_ref_match_info[\"match\"] = novel_ref_match_info[\"match\"].replace(\"\\w*\", 1, regex=True)\n",
    "        \n",
    "    t12 = timer()\n",
    "    print(\"took {} s\".format(t12 - t11))\n",
    "    \n",
    "    \n",
    "    # 8. Filter down matching transcripts to those that all ref introns except penultimate or all introns...\n",
    "    print(\"filtering for valid intron chain matches...\")\n",
    "    t13 = timer()\n",
    "    if match_type == \"any\":\n",
    "        # Only need to check by novel transcript_id\n",
    "        filt_novel_ref_match_info = (novel_ref_match_info.groupby(\"transcript_id_novel\")\n",
    "                                     .filter(lambda x: validate_matching_chain(x, max_terminal_non_match)\n",
    "                                            )\n",
    "                                    )\n",
    "    \n",
    "    elif match_type == \"transcript\":\n",
    "        # Check novel tx vs each ref tx\n",
    "        filt_novel_ref_match_info = (novel_ref_match_info.groupby([\"transcript_id_novel\",\"transcript_id_ref\"])\n",
    "                                     .filter(lambda x: validate_matching_chain(x, max_terminal_non_match)\n",
    "                                            )\n",
    "                                    )\n",
    "    t14 = timer()\n",
    "    print(\"took {} s\".format(t14 - t13))\n",
    "        \n",
    "     \n",
    "    # Return simplified df of novel transcript_id & matching transcript_id if applicable\n",
    "    \n",
    "    if match_type == \"any\":\n",
    "        return filt_novel_ref_match_info[\"transcript_id_novel\"].drop_duplicates()\n",
    "    \n",
    "    elif match_type == \"transcript\":\n",
    "        return filt_novel_ref_match_info[[\"transcript_id_novel\",\"transcript_id_ref\"]].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c17cfdba",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_transcripts_by_chain(test_novel_gr, test_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8312aea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_transcripts_by_chain(test_novel_gr, test_ref,match_type=\"any\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "83b385c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "finding introns...\n",
      "took 79.48507030098699 (s)\n",
      "adding intron_id column...\n",
      "took 0.17554569896310568 s\n",
      "generating df of novel txipts sorted by intron number...\n",
      "took 4.423515614995267 s\n",
      "finding overlaps between novel and reference introns...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-26 14:19:48,128\tINFO services.py:1272 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "took 4.683659655973315 s\n",
      "filtering overlaps for exact matches...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-26 14:19:52,562\tINFO services.py:1272 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "took 3.6306387850199826 s\n",
      "preparing for filtering intron matches...\n",
      "took 0.19747438799822703 s\n",
      "filtering for valid intron chain matches...\n",
      "took 0.42044611898018047 s\n",
      "total time taken: 93.09141320799245\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "26        PAPA.100.1\n",
       "86       PAPA.1001.2\n",
       "139      PAPA.1002.1\n",
       "208      PAPA.1003.1\n",
       "210      PAPA.1004.1\n",
       "            ...     \n",
       "99544     PAPA.985.1\n",
       "99546     PAPA.985.6\n",
       "99612     PAPA.994.2\n",
       "99678     PAPA.995.2\n",
       "99690     PAPA.999.1\n",
       "Name: transcript_id_novel, Length: 1191, dtype: object"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pre_chr1 = timer()\n",
    "chr1_any_matched = filter_transcripts_by_chain(stie_chr1_exons, ref_chr1_pc_exons, match_type=\"any\", nb_cpu=2)\n",
    "print(\"total time taken: {}\".format(timer() - pre_chr1))\n",
    "chr1_any_matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e5ba1053",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "finding introns...\n",
      "took 84.49933640897507 (s)\n",
      "adding intron_id column...\n",
      "took 0.162128378986381 s\n",
      "generating df of novel txipts sorted by intron number...\n",
      "took 4.3569289009901695 s\n",
      "finding overlaps between novel and reference introns...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-26 14:21:25,989\tINFO services.py:1272 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "ename": "RayTaskError(KeyError)",
     "evalue": "\u001b[36mray::_write_both()\u001b[39m (pid=1275, ip=10.97.45.105)\n  File \"python/ray/_raylet.pyx\", line 501, in ray._raylet.execute_task\n  File \"/home/sam/miniconda3/envs/bioinfo/lib/python3.8/site-packages/ray/util/tracing/tracing_helper.py\", line 330, in _function_with_tracing\n    return function(*args, **kwargs)\n  File \"/home/sam/miniconda3/envs/bioinfo/lib/python3.8/site-packages/pyranges/methods/join.py\", line 111, in _write_both\n    how = kwargs[\"how\"]\nKeyError: 'how'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRayTaskError(KeyError)\u001b[0m                    Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-25-7f58aabf7342>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mpre_chr1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtimer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mchr1_tx_matched\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfilter_transcripts_by_chain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstie_chr1_exons\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mref_chr1_pc_exons\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmatch_type\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"transcript\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnb_cpu\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"time taken: {}\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mpre_chr1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mchr1_tx_matched\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-23-745311db9660>\u001b[0m in \u001b[0;36mfilter_transcripts_by_chain\u001b[0;34m(novel_exons, ref_exons, match_type, max_terminal_non_match, nb_cpu)\u001b[0m\n\u001b[1;32m    106\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    107\u001b[0m     \u001b[0mt7\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtimer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 108\u001b[0;31m     \u001b[0mjoined\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnovel_introns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mref_introns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstrandedness\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"same\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msuffix\u001b[0m \u001b[0;34m=\u001b[0m\u001b[0;34m\"_ref\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mnb_cpu\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnb_cpu\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    109\u001b[0m     \u001b[0mt8\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtimer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    110\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.8/site-packages/pyranges/pyranges.py\u001b[0m in \u001b[0;36mjoin\u001b[0;34m(self, other, strandedness, how, report_overlap, slack, suffix, nb_cpu)\u001b[0m\n\u001b[1;32m   2151\u001b[0m             \u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"example_header_self\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2152\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2153\u001b[0;31m         \u001b[0mdfs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpyrange_apply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_write_both\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2154\u001b[0m         \u001b[0mgr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPyRanges\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdfs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2155\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.8/site-packages/pyranges/multithreaded.py\u001b[0m in \u001b[0;36mpyrange_apply\u001b[0;34m(function, self, other, **kwargs)\u001b[0m\n\u001b[1;32m    306\u001b[0m                 \u001b[0mresults\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    307\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 308\u001b[0;31m     \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    309\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    310\u001b[0m     \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprocess_results\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeys\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.8/site-packages/ray/_private/client_mode_hook.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     60\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mclient_mode_should_convert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     61\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mray\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 62\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     63\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     64\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.8/site-packages/ray/worker.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(object_refs, timeout)\u001b[0m\n\u001b[1;32m   1492\u001b[0m                     \u001b[0mworker\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcore_worker\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdump_object_store_memory_usage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1493\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mRayTaskError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1494\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mas_instanceof_cause\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1495\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1496\u001b[0m                     \u001b[0;32mraise\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRayTaskError(KeyError)\u001b[0m: \u001b[36mray::_write_both()\u001b[39m (pid=1275, ip=10.97.45.105)\n  File \"python/ray/_raylet.pyx\", line 501, in ray._raylet.execute_task\n  File \"/home/sam/miniconda3/envs/bioinfo/lib/python3.8/site-packages/ray/util/tracing/tracing_helper.py\", line 330, in _function_with_tracing\n    return function(*args, **kwargs)\n  File \"/home/sam/miniconda3/envs/bioinfo/lib/python3.8/site-packages/pyranges/methods/join.py\", line 111, in _write_both\n    how = kwargs[\"how\"]\nKeyError: 'how'"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-26 14:21:32,663\tERROR worker.py:78 -- Unhandled error (suppress with RAY_IGNORE_UNHANDLED_ERRORS=1): \u001b[36mray::_write_both()\u001b[39m (pid=1276, ip=10.97.45.105)\n",
      "  File \"python/ray/_raylet.pyx\", line 501, in ray._raylet.execute_task\n",
      "  File \"/home/sam/miniconda3/envs/bioinfo/lib/python3.8/site-packages/ray/util/tracing/tracing_helper.py\", line 330, in _function_with_tracing\n",
      "    return function(*args, **kwargs)\n",
      "  File \"/home/sam/miniconda3/envs/bioinfo/lib/python3.8/site-packages/pyranges/methods/join.py\", line 111, in _write_both\n",
      "    how = kwargs[\"how\"]\n",
      "KeyError: 'how'\n"
     ]
    }
   ],
   "source": [
    "pre_chr1 = timer()\n",
    "chr1_tx_matched = filter_transcripts_by_chain(stie_chr1_exons, ref_chr1_pc_exons, match_type=\"transcript\", nb_cpu=2)\n",
    "print(\"time taken: {}\".format(timer() - pre_chr1))\n",
    "chr1_tx_matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "266d95d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>cov</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>TPM</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>827750</td>\n",
       "      <td>859332</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>98.466469</td>\n",
       "      <td>11.679049</td>\n",
       "      <td>30.310225</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>827750</td>\n",
       "      <td>827775</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>96.734062</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>829002</td>\n",
       "      <td>829104</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>110.622597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>847653</td>\n",
       "      <td>847806</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>118.018204</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>851926</td>\n",
       "      <td>852110</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.35</td>\n",
       "      <td>PAPA.35.14</td>\n",
       "      <td>115.267273</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12873</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248855557</td>\n",
       "      <td>248855635</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.161189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12874</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248855724</td>\n",
       "      <td>248855943</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>4.278336</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12875</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248856287</td>\n",
       "      <td>248856422</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.487612</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12876</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248856513</td>\n",
       "      <td>248856562</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>5.115693</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12877</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>248857233</td>\n",
       "      <td>248858865</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PAPA.2531</td>\n",
       "      <td>PAPA.2531.4</td>\n",
       "      <td>13.537362</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12878 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | Score     | Strand       | Frame      | gene_id    | transcript_id   | cov        | FPKM       | TPM        | +1    |\n",
       "| (category)   | (object)   | (category)   | (int32)   | (int32)   | (int64)   | (category)   | (object)   | (object)   | (object)        | (object)   | (object)   | (object)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+-------|\n",
       "| chr1         | StringTie  | transcript   | 827750    | 859332    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 98.466469  | 11.679049  | 30.310225  | ...   |\n",
       "| chr1         | StringTie  | exon         | 827750    | 827775    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 96.734062  | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 829002    | 829104    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 110.622597 | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 847653    | 847806    | 1000      | +            | .          | PAPA.35    | PAPA.35.14      | 118.018204 | nan        | nan        | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...       | ...          | ...        | ...        | ...             | ...        | ...        | ...        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248855724 | 248855943 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 4.278336   | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248856287 | 248856422 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 5.487612   | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248856513 | 248856562 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 5.115693   | nan        | nan        | ...   |\n",
       "| chr1         | StringTie  | exon         | 248857233 | 248858865 | 1000      | -            | .          | PAPA.2531  | PAPA.2531.4     | 13.537362  | nan        | nan        | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-----------+--------------+------------+------------+-----------------+------------+------------+------------+-------+\n",
       "Stranded PyRanges object has 12,878 rows and 14 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "1 hidden columns: exon_number"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stie_chr1.subset(lambda df: df[\"transcript_id\"].isin(set(chr1_any_matched.tolist())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ac2af658",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3157548",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:bioinfo] *",
   "language": "python",
   "name": "conda-env-bioinfo-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
