{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b5170bd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "import time "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fb15ee4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "|            1 |        10 |        30 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       100 |       120 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       200 |       220 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       300 |       340 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 4 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A plus strand transcript\n",
    "test_ref_tr1 = {\"Chromosome\": [1]*4,\n",
    "                \"Start\": [10,100,200,300],\n",
    "                \"End\": [30,120,220,340],\n",
    "                \"Strand\": [\"+\"]*4,\n",
    "                \"Feature\": [\"exon\"]*4,\n",
    "                \"gene_id\": [\"ref_gene_1\"]*4,\n",
    "                \"transcript_id\": [\"ref_tr_1\"] * 4}\n",
    "\n",
    "\n",
    "pr.from_dict(test_ref_tr1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1f982b51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Chromosome': [],\n",
       " 'Start': [],\n",
       " 'End': [],\n",
       " 'Strand': [],\n",
       " 'Feature': [],\n",
       " 'gene_id': [],\n",
       " 'transcript_id': []}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{\"Chromosome\": [],\n",
    "                \"Start\": [],\n",
    "                \"End\": [],\n",
    "                \"Strand\": [],\n",
    "                \"Feature\": [],\n",
    "                \"gene_id\": [],\n",
    "                \"transcript_id\": []\n",
    "               }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "37648c16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>ref_tr_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>80</td>\n",
       "      <td>90</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>ref_tr_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>ref_tr_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "|            1 |        10 |        30 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       100 |       120 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       200 |       220 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       300 |       340 | +            | exon       | ref_gene_1 | ref_tr_1        |\n",
       "|            2 |        10 |        20 | -            | exon       | ref_gene_2 | ref_tr_2        |\n",
       "|            2 |        80 |        90 | -            | exon       | ref_gene_2 | ref_tr_2        |\n",
       "|            2 |       100 |       120 | -            | exon       | ref_gene_2 | ref_tr_2        |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 7 rows and 7 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A minus strand transcript\n",
    "test_ref_tr2 = {\"Chromosome\": [2]*3,\n",
    "                \"Start\": [10,80,100],\n",
    "                \"End\": [20,90,120],\n",
    "                \"Strand\": [\"-\"]*3,\n",
    "                \"Feature\": [\"exon\"]*3,\n",
    "                \"gene_id\": [\"ref_gene_2\"]*3,\n",
    "                \"transcript_id\": [\"ref_tr_2\"]*3\n",
    "               }\n",
    "\n",
    "\n",
    "test_ref = pr.concat([pr.from_dict(test_ref_tr1), pr.from_dict(test_ref_tr2)])\n",
    "test_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fbba9f36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>35</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_fi_p     |\n",
       "|            1 |        50 |        70 | +            | exon       | nov_gene_1 | nov_tx_fi_p     |\n",
       "|            1 |        10 |        35 | +            | exon       | nov_gene_1 | nov_tx_fi_f     |\n",
       "|            1 |        50 |        70 | +            | exon       | nov_gene_1 | nov_tx_fi_f     |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 4 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now make test novel transcripts to cover my test cases\n",
    "\n",
    "# 1. Novel last exons in first intron of annotated transcript (e.g. STMN2)\n",
    "# For these to pass, they should share an identical 3'end with a first exon of a known transcript\n",
    "\n",
    "test_novel_fi = {\"Chromosome\": [1]*4,\n",
    "                \"Start\": [10,50]*2,\n",
    "                \"End\": [30,70,35,70],\n",
    "                \"Strand\": [\"+\"]*4,\n",
    "                \"Feature\": [\"exon\"]*4,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*4,\n",
    "                \"transcript_id\": [\"nov_tx_fi_p\"]*2 + [\"nov_tx_fi_f\"]*2,\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_fi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "213b2de9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_si_p     |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_si_p     |\n",
       "|            1 |       140 |       160 | +            | exon       | nov_gene_1 | nov_tx_si_p     |\n",
       "|            1 |        50 |        70 | +            | exon       | nov_gene_1 | nov_tx_si_f     |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_si_f     |\n",
       "|            1 |       140 |       160 | +            | exon       | nov_gene_1 | nov_tx_si_f     |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 6 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2. Internal intron, spliced in last exon (fully contained within last exon) (e.g. ONECUT1)\n",
    "# For these to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "\n",
    "test_novel_si = {\"Chromosome\": [1]*6,\n",
    "                \"Start\": [10,100,140] + [50,100,140],\n",
    "                \"End\": [30,120,160] + [70,120,160],\n",
    "                \"Strand\": [\"+\"]*6,\n",
    "                \"Feature\": [\"exon\"]*6,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*6,\n",
    "                \"transcript_id\": [\"nov_tx_si_p\"]*3 + [\"nov_tx_si_f\"]*3\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_si)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2f34a0e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>240</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>240</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_bl_p     |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_bl_p     |\n",
       "|            1 |       200 |       240 | +            | exon       | nov_gene_1 | nov_tx_bl_p     |\n",
       "|            1 |        50 |        70 | +            | exon       | nov_gene_1 | nov_tx_bl_f     |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_bl_f     |\n",
       "|            1 |       200 |       240 | +            | exon       | nov_gene_1 | nov_tx_bl_f     |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 6 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3. Internal intron bleedthrough (e.g. SIN3B)\n",
    "# For these events to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "\n",
    "test_novel_bl = {\"Chromosome\": [1]*6,\n",
    "                \"Start\": [10,100,200] + [50,100,200],\n",
    "                \"End\": [30,120,240] + [70,120,240],\n",
    "                \"Strand\": [\"+\"]*6,\n",
    "                \"Feature\": [\"exon\"]*6,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*6,\n",
    "                \"transcript_id\": [\"nov_tx_bl_p\"]*3 + [\"nov_tx_bl_f\"]*3\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_bl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f32f14c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>130</td>\n",
       "      <td>140</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>150</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_mult_p   |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_mult_p   |\n",
       "|            1 |       130 |       140 | +            | exon       | nov_gene_1 | nov_tx_mult_p   |\n",
       "|            1 |       150 |       160 | +            | exon       | nov_gene_1 | nov_tx_mult_p   |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 4 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 4. Internal intron with novel internal and terminal exon (e.g.)\n",
    "# For these events to pass, they should match the intron chain of a known transcript,\n",
    "# but have a continuous chain of length n of novel events at the 3'end of the transcript\n",
    "# (n can be varied)\n",
    "# Event know from NP is fully contained within annotated intron - also set this constraint?\n",
    "\n",
    "test_novel_mult = {\"Chromosome\": [1]*4,\n",
    "                   \"Start\": [10,100,130,150],\n",
    "                   \"End\": [30,120,140,160],\n",
    "                   \"Strand\": [\"+\"]*4,\n",
    "                   \"Feature\": [\"exon\"]*4,\n",
    "                   \"gene_id\": [\"nov_gene_1\"]*4,\n",
    "                   \"transcript_id\": [\"nov_tx_mult_p\"]*4\n",
    "                    }\n",
    "\n",
    "pr.from_dict(test_novel_mult)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c68fdf5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>310</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>330</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "|            1 |       200 |       220 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "|            1 |       300 |       310 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "|            1 |       330 |       340 | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 5 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 5. 3'UTR intron fully contained within an annotated 3'UTR (e.g. TDP-43)\n",
    "# For this to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "# (Annotate as a 3'UTR intron (spliced out) after filtering for intron chain match)\n",
    "test_novel_3ui = {\"Chromosome\": [1]*5,\n",
    "                \"Start\": [10,100,200,300,330],\n",
    "                \"End\": [30,120,220,310,340],\n",
    "                \"Strand\": [\"+\"]*5,\n",
    "                \"Feature\": [\"exon\"]*5,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*5,\n",
    "                \"transcript_id\": [\"nov_tx_3ui_p\"]*5\n",
    "                 }\n",
    "\n",
    "pr.from_dict(test_novel_3ui)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "883cf45e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>360</td>\n",
       "      <td>380</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-------------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id     |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)          |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-------------------|\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_exc_dist_p |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_exc_dist_p |\n",
       "|            1 |       200 |       220 | +            | exon       | nov_gene_1 | nov_tx_exc_dist_p |\n",
       "|            1 |       360 |       380 | +            | exon       | nov_gene_1 | nov_tx_exc_dist_p |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-------------------+\n",
       "Stranded PyRanges object has 4 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#6. Distal last exon spliced from penultimate exon (i.e. a mutually exclusive last exon) (e.g. SMC1A)\n",
    "# For this to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "# Will have to annotate more precisely later (i.e. differentiate from 7)\n",
    "\n",
    "test_novel_exc_dist = {\"Chromosome\": [1]*4,\n",
    "                \"Start\": [10,100,200,360],\n",
    "                \"End\": [30,120,220,380],\n",
    "                \"Strand\": [\"+\"]*4,\n",
    "                \"Feature\": [\"exon\"]*4,\n",
    "                \"gene_id\": [\"nov_gene_1\"]*4,\n",
    "                \"transcript_id\": [\"nov_tx_exc_dist_p\"]*4\n",
    "               }\n",
    "\n",
    "pr.from_dict(test_novel_exc_dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "44d51aea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>310</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>360</td>\n",
       "      <td>380</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+------------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id    |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)         |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+------------------|\n",
       "|            1 |        10 |        30 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |\n",
       "|            1 |       100 |       120 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |\n",
       "|            1 |       200 |       220 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |\n",
       "|            1 |       300 |       310 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |\n",
       "|            1 |       360 |       380 | +            | exon       | nov_gene_1 | nov_tx_le_dist_p |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+------------------+\n",
       "Stranded PyRanges object has 5 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#7. Distal last exon spliced from ann\n",
    "# For this to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "test_novel_le_dist = {\"Chromosome\": [1]*5,\n",
    "                      \"Start\": [10,100,200,300,360],\n",
    "                      \"End\": [30,120,220,310,380],\n",
    "                      \"Strand\": [\"+\"]*5,\n",
    "                      \"Feature\": [\"exon\"]*5,\n",
    "                      \"gene_id\": [\"nov_gene_1\"]*5,\n",
    "                      \"transcript_id\": [\"nov_tx_le_dist_p\"]*5\n",
    "                     }\n",
    "\n",
    "pr.from_dict(test_novel_le_dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8d51550f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>80</td>\n",
       "      <td>90</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+------------------------+\n",
       "|   Chromosome |     Start |       End | Strand       | Feature    | gene_id    | transcript_id          |\n",
       "|   (category) |   (int32) |   (int32) | (category)   | (object)   | (object)   | (object)               |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+------------------------|\n",
       "|            2 |        40 |        50 | -            | exon       | nov_gene_2 | nov_tx_le_dist_p_minus |\n",
       "|            2 |        80 |        90 | -            | exon       | nov_gene_2 | nov_tx_le_dist_p_minus |\n",
       "|            2 |       100 |       120 | -            | exon       | nov_gene_2 | nov_tx_le_dist_p_minus |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+------------------------+\n",
       "Stranded PyRanges object has 3 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#8. Distal last exon spliced from annoatrd (minus strand)\n",
    "# For this to pass, they should match the intron chain of a known transcript up until the penultimate exon\n",
    "test_novel_le_dist_minus = {\"Chromosome\": [2]*3,\n",
    "                      \"Start\": [40,80,100],\n",
    "                      \"End\": [50,90,120],\n",
    "                      \"Strand\": [\"-\"]*3,\n",
    "                      \"Feature\": [\"exon\"]*3,\n",
    "                      \"gene_id\": [\"nov_gene_2\"]*3,\n",
    "                      \"transcript_id\": [\"nov_tx_le_dist_p_minus\"]*3\n",
    "                     }\n",
    "\n",
    "pr.from_dict(test_novel_le_dist_minus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4dc1c09e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>310</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>330</td>\n",
       "      <td>340</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>240</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>240</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>360</td>\n",
       "      <td>380</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>35</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>310</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>360</td>\n",
       "      <td>380</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1</td>\n",
       "      <td>130</td>\n",
       "      <td>140</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1</td>\n",
       "      <td>150</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>70</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>2</td>\n",
       "      <td>80</td>\n",
       "      <td>90</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "| Chromosome   | Start     | End       | Strand       | Feature    | gene_id    | transcript_id   |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+--------------+------------+------------+-----------------|\n",
       "| 1            | 10        | 30        | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| 1            | 100       | 120       | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| 1            | 200       | 220       | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| 1            | 300       | 310       | +            | exon       | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| ...          | ...       | ...       | ...          | ...        | ...        | ...             |\n",
       "| 1            | 140       | 160       | +            | exon       | nov_gene_1 | nov_tx_si_p     |\n",
       "| 1            | 50        | 70        | +            | exon       | nov_gene_1 | nov_tx_si_f     |\n",
       "| 1            | 100       | 120       | +            | exon       | nov_gene_1 | nov_tx_si_f     |\n",
       "| 1            | 140       | 160       | +            | exon       | nov_gene_1 | nov_tx_si_f     |\n",
       "+--------------+-----------+-----------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 37 rows and 7 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_novel_gr = pr.concat([pr.from_dict(event) for event in [test_novel_3ui,\n",
    "                                                            test_novel_bl,\n",
    "                                                            test_novel_exc_dist,\n",
    "                                                            test_novel_fi,\n",
    "                                                            test_novel_le_dist,\n",
    "                                                            test_novel_mult,\n",
    "                                                            test_novel_si,\n",
    "                                                            test_novel_le_dist_minus]\n",
    "                          ]\n",
    "                         )\n",
    "\n",
    "test_novel_gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6e2ae531",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>220</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>310</td>\n",
       "      <td>330</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>70</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>220</td>\n",
       "      <td>360</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>35</td>\n",
       "      <td>50</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>50</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>220</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>310</td>\n",
       "      <td>360</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>130</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>150</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>70</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>140</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>140</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>2</td>\n",
       "      <td>50</td>\n",
       "      <td>80</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>2</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>nov_gene_2</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+------------+-----------------+\n",
       "| Chromosome   | Start     | End       | Strand     | Feature    | gene_id    | transcript_id   |\n",
       "| (object)     | (int64)   | (int64)   | (object)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+------------+------------+------------+-----------------|\n",
       "| 1            | 30        | 100       | +          | intron     | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| 1            | 120       | 200       | +          | intron     | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| 1            | 220       | 300       | +          | intron     | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| 1            | 310       | 330       | +          | intron     | nov_gene_1 | nov_tx_3ui_p    |\n",
       "| ...          | ...       | ...       | ...        | ...        | ...        | ...             |\n",
       "| 1            | 70        | 100       | +          | intron     | nov_gene_1 | nov_tx_si_f     |\n",
       "| 1            | 120       | 140       | +          | intron     | nov_gene_1 | nov_tx_si_f     |\n",
       "| 1            | 30        | 100       | +          | intron     | nov_gene_1 | nov_tx_si_p     |\n",
       "| 1            | 120       | 140       | +          | intron     | nov_gene_1 | nov_tx_si_p     |\n",
       "+--------------+-----------+-----------+------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 26 rows and 7 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def introns_from_df(df):\n",
    "    '''\n",
    "    '''\n",
    "    \n",
    "    n_exons = len(df)\n",
    "    \n",
    "    if n_exons < 2:\n",
    "        return None\n",
    "        #print(df)\n",
    "        #raise Exception(\"at least two exons are required for transcript to have an intron\")\n",
    "    # n exons = n-1 introns\n",
    "    \n",
    "    strand = df[\"Strand\"].drop_duplicates().tolist()[0]\n",
    "#     print(strand)\n",
    "    chrom = df[\"Chromosome\"].drop_duplicates().tolist()[0]\n",
    "    gene_id = df[\"gene_id\"].drop_duplicates().tolist()[0]\n",
    "    tx_id = df[\"transcript_id\"].drop_duplicates().tolist()[0]\n",
    "    feature = \"intron\"\n",
    "    introns = {}\n",
    "    for i in range(0, n_exons - 1):\n",
    "        if strand == \"+\":\n",
    "            intron_start = df.iloc[i, lambda x: x.columns.get_loc(\"End\")]\n",
    "            intron_end = df.iloc[i+1, lambda x: x.columns.get_loc(\"Start\")]\n",
    "            introns[str(i)] = {\"Chromosome\": chrom,\n",
    "                               \"Start\": intron_start,\n",
    "                               \"End\": intron_end,\n",
    "                               \"Strand\": strand,\n",
    "                               \"Feature\": feature,\n",
    "                               \"gene_id\": gene_id,\n",
    "                               \"transcript_id\": tx_id}\n",
    "        elif strand == \"-\":\n",
    "            intron_start = df.iloc[i, lambda x: x.columns.get_loc(\"End\")]\n",
    "            intron_end = df.iloc[i+1, lambda x: x.columns.get_loc(\"Start\")]\n",
    "            introns[str(i)] = {\"Chromosome\": chrom,\n",
    "                               \"Start\": intron_start,\n",
    "                               \"End\": intron_end,\n",
    "                               \"Strand\": strand,\n",
    "                               \"Feature\": feature,\n",
    "                               \"gene_id\": gene_id,\n",
    "                               \"transcript_id\": tx_id}\n",
    "    return pd.DataFrame.from_dict(introns, orient = \"index\")\n",
    "        \n",
    "\n",
    "\n",
    "def introns_by_tx(gr, by=\"transcript_id\", nb_cpu=1):\n",
    "    '''\n",
    "    '''\n",
    "    # Sort by position (for safety)\n",
    "    gr = gr.sort()\n",
    "    \n",
    "    return gr.apply(lambda df: df.groupby(by).apply(introns_from_df), nb_cpu=nb_cpu)\n",
    "    \n",
    "    \n",
    "test_novel_introns = introns_by_tx(test_novel_gr)\n",
    "test_ref_introns = introns_by_tx(test_ref)\n",
    "    \n",
    "    \n",
    "test_novel_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7c243a35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>220</td>\n",
       "      <td>300</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>ref_gene_1</td>\n",
       "      <td>ref_tr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "      <td>80</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>ref_tr_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>ref_gene_2</td>\n",
       "      <td>ref_tr_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+------------+------------+-----------------+\n",
       "|   Chromosome |     Start |       End | Strand     | Feature    | gene_id    | transcript_id   |\n",
       "|     (object) |   (int64) |   (int64) | (object)   | (object)   | (object)   | (object)        |\n",
       "|--------------+-----------+-----------+------------+------------+------------+-----------------|\n",
       "|            1 |        30 |       100 | +          | intron     | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       120 |       200 | +          | intron     | ref_gene_1 | ref_tr_1        |\n",
       "|            1 |       220 |       300 | +          | intron     | ref_gene_1 | ref_tr_1        |\n",
       "|            2 |        20 |        80 | -          | intron     | ref_gene_2 | ref_tr_2        |\n",
       "|            2 |        90 |       100 | -          | intron     | ref_gene_2 | ref_tr_2        |\n",
       "+--------------+-----------+-----------+------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 5 rows and 7 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_ref_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d382d27d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rle(inarray):\n",
    "        \"\"\"\n",
    "        run length encoding. Partial credit to R rle function. \n",
    "        Multi datatype arrays catered for including non Numpy\n",
    "        returns: tuple (runlengths, startpositions, values)\n",
    "        https://stackoverflow.com/questions/1066758/find-length-of-sequences-of-identical-values-in-a-numpy-array-run-length-encodi\n",
    "        Thomas Browne\n",
    "        \"\"\"\n",
    "        ia = np.asarray(inarray)                # force numpy\n",
    "        n = len(ia)\n",
    "        if n == 0: \n",
    "            return (None, None, None)\n",
    "        else:\n",
    "            y = ia[1:] != ia[:-1]               # pairwise unequal (string safe)\n",
    "            i = np.append(np.where(y), n - 1)   # must include last element posi\n",
    "            z = np.diff(np.append(-1, i))       # run lengths\n",
    "            p = np.cumsum(np.append(0, z))[:-1] # positions\n",
    "            return(z, p, ia[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "32cbab88",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----processing chrom and strand pair 1 & +\n",
      "nov_tx_bl_f does not match any reference transcripts in its first intron. Skipping\n",
      "nov_tx_si_f does not match any reference transcripts in its first intron. Skipping\n",
      "----processing chrom and strand pair 2 & -\n",
      "{'nov_tx_3ui_p': {'matches': ['ref_tr_1'], 'chain_match': [3], 'terminal_non_match': [0]}, 'nov_tx_bl_p': {'matches': ['ref_tr_1'], 'chain_match': [2], 'terminal_non_match': [0]}, 'nov_tx_exc_dist_p': {'matches': ['ref_tr_1'], 'chain_match': [2], 'terminal_non_match': [1]}, 'nov_tx_le_dist_p': {'matches': ['ref_tr_1'], 'chain_match': [3], 'terminal_non_match': [0]}, 'nov_tx_mult_p': {'matches': ['ref_tr_1'], 'chain_match': [1], 'terminal_non_match': [2]}, 'nov_tx_si_p': {'matches': ['ref_tr_1'], 'chain_match': [1], 'terminal_non_match': [1]}, 'nov_tx_le_dist_p_minus': {'matches': ['ref_tr_2'], 'chain_match': [1], 'terminal_non_match': [1]}}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>matches</th>\n",
       "      <th>chain_match</th>\n",
       "      <th>terminal_non_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>nov_tx_3ui_p</th>\n",
       "      <th>0</th>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nov_tx_bl_p</th>\n",
       "      <th>0</th>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nov_tx_exc_dist_p</th>\n",
       "      <th>0</th>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nov_tx_le_dist_p</th>\n",
       "      <th>0</th>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nov_tx_mult_p</th>\n",
       "      <th>0</th>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nov_tx_si_p</th>\n",
       "      <th>0</th>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nov_tx_le_dist_p_minus</th>\n",
       "      <th>0</th>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           matches  chain_match  terminal_non_match\n",
       "nov_tx_3ui_p           0  ref_tr_1            3                   0\n",
       "nov_tx_bl_p            0  ref_tr_1            2                   0\n",
       "nov_tx_exc_dist_p      0  ref_tr_1            2                   1\n",
       "nov_tx_le_dist_p       0  ref_tr_1            3                   0\n",
       "nov_tx_mult_p          0  ref_tr_1            1                   2\n",
       "nov_tx_si_p            0  ref_tr_1            1                   1\n",
       "nov_tx_le_dist_p_minus 0  ref_tr_2            1                   1"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dict of {novel_tx_id: {matches: [ref_id], chain_match: [n_matching], terminal_non_match: [n_not_matching]}}\n",
    "novel_info_dict = {}\n",
    "\n",
    "test_novel_gr.apply(lambda df: df.groupby(\"transcript_id\").filter(lambda grp: len(grp) > 2))\n",
    "\n",
    "for key, dfs in pr.itergrs([test_novel_introns.apply(lambda df: df.groupby(\"transcript_id\").filter(lambda grp: len(grp) > 1)), test_ref_introns], strand=True, keys=True):\n",
    "    print(\"----processing chrom and strand pair {0} & {1}\".format(key[0], key[1]))\n",
    "  \n",
    "    by_tx = tuple([df.groupby(\"transcript_id\") for df in dfs])\n",
    "    novel_txipts = by_tx[0]\n",
    "    ref_txipts = by_tx[1]\n",
    "    \n",
    "    strand = key[1]\n",
    "#     print(strand)\n",
    "#    #1. for each novel transcript, test it against all ref transcripts:\n",
    "    \n",
    "    \n",
    "    for novel_id, novel_introns in novel_txipts:\n",
    "#         print(novel_id)\n",
    "#         print(novel_introns)\n",
    "#         print(\"class of novel_introns is {}\".format(type(novel_introns)))\n",
    "        #print(exons)\n",
    "        \n",
    "        if strand == \"-\":\n",
    "            #First in order = last intron - reverse df now\n",
    "            novel_introns = novel_introns[::-1].reset_index(drop=True)\n",
    "#             ref_introns = ref_introns[::-1].reset_index(drop=True)\n",
    "            #print(exons)\n",
    "        #by_tx[1]\n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "        \n",
    "        if strand == \"+\":\n",
    "            first_matches = [np.array_equal(novel_introns.head(1)[[\"Start\",\"End\"]],\n",
    "                                            ref_introns.head(1)[[\"Start\",\"End\"]]) for ref_id,ref_introns in ref_txipts]\n",
    "        else:\n",
    "            first_matches = [np.array_equal(novel_introns.head(1)[[\"Start\",\"End\"]],\n",
    "                                            ref_introns[::-1].reset_index(drop=True) # rev order\n",
    "                                            .head(1)[[\"Start\",\"End\"]]\n",
    "                                           )\n",
    "                             for ref_id,ref_introns in ref_txipts]\n",
    "\n",
    "        #print(novel_id)\n",
    "#         print(first_matches)\n",
    "        if not sum(first_matches) > 0:\n",
    "            print(\"{0} does not match any reference transcripts in its first intron. Skipping\".format(novel_id))\n",
    "            continue\n",
    "        \n",
    "        # Now compare novel transcript against each ref transcript with match in first intron.\n",
    "        # \n",
    "        for ref_tr, first_match in zip(ref_txipts, first_matches):\n",
    "            \n",
    "            if not first_match:\n",
    "                continue\n",
    "            #intron_chain_match = np.equal(np.asarray(exons[[\"Start\",\"End\"]]), np.asarray(ref_tr[1][[\"Start\",\"End\"]]))\n",
    "            ref_introns = ref_tr[1]\n",
    "            \n",
    "            if strand == \"-\":\n",
    "                ref_introns = ref_introns[::-1].reset_index(drop=True) # rev order\n",
    "            \n",
    "            n_novel_introns = len(novel_introns)\n",
    "            n_ref_introns = len(ref_introns)\n",
    "            \n",
    "            # To avoid a slicing error for txipts shorter than novel\n",
    "            max_chain = min(n_novel_introns, n_ref_introns)\n",
    "#             print(max_chain)\n",
    "#             print(ref_tr[1].iloc[0,])\n",
    "#             print(n_novel_introns)\n",
    "#             print(len(ref_tr[1]))\n",
    "            novel_chain_match = pd.DataFrame([(novel_introns.iloc[i,:][[\"Start\",\"End\"]]\n",
    "                                               .eq(ref_introns.iloc[i,:][[\"Start\",\"End\"]]\n",
    "                                                  )\n",
    "                                              )\n",
    "                                              for i in range(max_chain)])\n",
    "    \n",
    "            # Collapse to single True/False per row - does intron match?\n",
    "            novel_chain_match = novel_chain_match.apply(np.all, axis=1, raw=True)\n",
    "            \n",
    "#             print(novel_chain_match)\n",
    "            \n",
    "            runs, starts, vals = rle(novel_chain_match)\n",
    "            #print(runs[0])\n",
    "            #print(starts)\n",
    "            #print(\"\\n {0}\".format(np.where(vals == False)))\n",
    "            \n",
    "            if np.all(vals) or np.array_equal(vals, [True,False]):\n",
    "                # Possible genuine match, should update dict with info\n",
    "                \n",
    "                # match (from start of ref Txipt)\n",
    "                \n",
    "                if vals.size == 1:\n",
    "                    #i.e. all true/introns match (e.g. bleedthrough/ where )\n",
    "                    terminal_non_match = 0\n",
    "                else:\n",
    "                    terminal_non_match = runs[1]\n",
    "                \n",
    "#                 print(terminal_non_match)\n",
    "                \n",
    "                if novel_id not in novel_info_dict:\n",
    "                    \n",
    "                    novel_info_dict[novel_id] = {\"matches\": [ref_tr[0]],\n",
    "                                                 \"chain_match\": [runs[0]], #Always starts with true, so take length of true\n",
    "                                                 \"terminal_non_match\": [terminal_non_match],\n",
    "                                                }\n",
    "                \n",
    "                else:\n",
    "                    # Append to dict\n",
    "                    novel_info_dict[novel_id][\"matches\"].append(ref_tr[0])\n",
    "                    novel_info_dict[novel_id][\"chain_match\"].append(ref_tr[0])\n",
    "                    novel_info_dict[novel_id][\"terminal_non_match\"].append(ref_tr[0])\n",
    "                    \n",
    "                \n",
    "            else:\n",
    "                continue\n",
    "            \n",
    "            \n",
    "            \n",
    "print(novel_info_dict)\n",
    "#             intron_chain_match = exons[[\"Start\",\"End\"]].eq(ref_tr[1][[\"Start\",\"End\"]])\n",
    "#             print(\"\\n\")\n",
    "#             print(intron_chain_match)\n",
    "#             print(intron_chain_match.apply(np.all, axis = 1))\n",
    "            \n",
    "            \n",
    "                #loc[np.where(lambda x: (x[[\"Start\", \"End\"]].all()))])\n",
    "                                                  #apply(lambda x: all(x), axis = \"index\", raw = True))\n",
    "#             print(intron_chain_match)\n",
    "            \n",
    "            \n",
    "#         first_intron = exons.head(1)[[\"Start\",\"End\"]] if strand == \"+\" else exons.tail(1)[[\"Start\",\"End\"]]\n",
    "        \n",
    "#         grp_filter_matches = by_tx[1].filter(lambda x: np.array_equal(first_intron, x.head(1)[[\"Start\", \"End\"]]))\n",
    "#         print(grp_filter_matches)\n",
    "#         print(type(grp_filter_matches))\n",
    "     #for novel, ref_tr in zip(by_tx[0],by_tx[1]:\n",
    "#         print(nov_tr[1])\n",
    "#         print(ref_tr)\n",
    "             \n",
    "pd.concat({key: pd.DataFrame.from_dict(d, orient = \"columns\") for key, d in novel_info_dict.items()}, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "54913d65",
   "metadata": {},
   "outputs": [],
   "source": [
    "def match_intron_chains(novel_gr, ref_gr, id_col = \"transcript_id\", nb_cpu = 1):\n",
    "    '''\n",
    "    grs should contain introns\n",
    "    '''\n",
    "    # {novel_id: {matches: [ref_id], chain_match: [n], terminal_non_match: [n]}}\n",
    "    match_info_dict = {}\n",
    "    \n",
    "\n",
    "    \n",
    "    for key, dfs in pr.itergrs([novel_gr, ref_gr], strand=True, keys=True):\n",
    "        print(\"----processing chrom and strand pair {0} & {1}\".format(key[0], key[1]))\n",
    "        # dfs = novel_gr & ref_gr matched by key (chromosome & strand)\n",
    "        # pandas group by objects of (transcript_id, df)\n",
    "        by_tx = tuple([df.groupby(\"transcript_id\") for df in dfs])\n",
    "        \n",
    "        novel_txipts = by_tx[0]\n",
    "        ref_txipts = by_tx[1]\n",
    "        \n",
    "        # Pyranges keys are tuples of (chr,strand)\n",
    "        strand = key[1]\n",
    "        \n",
    "        \n",
    "        #Comparing each novel transcript against all ref transcripts\n",
    "        for novel_id, novel_introns in novel_txipts:\n",
    "        \n",
    "            if strand == \"-\":\n",
    "                # Standard PyRanges sort - First in df = last intron (smallest values i.e. leftmost)\n",
    "                # Reverse so first row in df is always the first intron\n",
    "                # is reset_index necessary?\n",
    "                novel_introns = novel_introns[::-1].reset_index(drop=True)\n",
    "            \n",
    "            else:\n",
    "                pass\n",
    "        \n",
    "            # As a first pass, check for matches in intron chain of first introns between novel and reference Txs\n",
    "            if strand == \"+\":\n",
    "                first_matches = [np.array_equal(novel_introns.head(1)[[\"Start\",\"End\"]],\n",
    "                                                ref_introns.head(1)[[\"Start\",\"End\"]]) \n",
    "                                 for ref_id,ref_introns in ref_txipts]\n",
    "            else:\n",
    "                first_matches = [np.array_equal(novel_introns.head(1)[[\"Start\",\"End\"]],\n",
    "                                                ref_introns[::-1].reset_index(drop=True) # rev order so 1st row = 1st intron\n",
    "                                                .head(1)[[\"Start\",\"End\"]]\n",
    "                                               )\n",
    "                                 for ref_id,ref_introns in ref_txipts]\n",
    "\n",
    "\n",
    "            if not sum(first_matches) > 0:\n",
    "                print(\"{0} does not match any reference transcripts in its first intron. Skipping\".format(novel_id))\n",
    "                continue\n",
    "        \n",
    "            # Compare full introns chains of novel transcript against each ref transcript with match in first intron.\n",
    "        \n",
    "            for ref_tr, first_match in zip(ref_txipts, first_matches):\n",
    "            \n",
    "                if not first_match:\n",
    "                    continue\n",
    "                \n",
    "            \n",
    "                ref_id = ref_tr[0]\n",
    "#                 print(\"ref_id {}\".format(ref_id))\n",
    "                ref_introns = ref_tr[1]\n",
    "            \n",
    "                if strand == \"-\":\n",
    "                    ref_introns = ref_introns[::-1].reset_index(drop=True) # reverse so first row = first intron\n",
    "                        \n",
    "                # To avoid a slicing error for ref txipts shorter than novel\n",
    "                n_novel_introns = len(novel_introns)\n",
    "                n_ref_introns = len(ref_introns)\n",
    "                max_chain = min(n_novel_introns, n_ref_introns)\n",
    "#             print(max_chain)\n",
    "#             print(ref_tr[1].iloc[0,])\n",
    "#             print(n_novel_introns)\n",
    "#             print(len(ref_tr[1]))\n",
    "\n",
    "                # Row-wise, check whether match with corresponding intron of reference transcript\n",
    "                novel_chain_match = pd.DataFrame([(novel_introns.iloc[i,:][[\"Start\",\"End\"]]\n",
    "                                                   .eq(ref_introns.iloc[i,:][[\"Start\",\"End\"]]\n",
    "                                                      )\n",
    "                                                  )\n",
    "                                                  for i in range(max_chain)])\n",
    "    \n",
    "                # Collapse to single True/False per row - does intron completely match?\n",
    "                novel_chain_match = novel_chain_match.apply(np.all, axis=1, raw=True)\n",
    "            \n",
    "#             print(novel_chain_match)\n",
    "            \n",
    "                runs, starts, vals = rle(novel_chain_match)\n",
    "            #print(runs[0])\n",
    "            #print(starts)\n",
    "            #print(\"\\n {0}\".format(np.where(vals == False)))\n",
    "            \n",
    "                # Considered a valid match if a intron chain completely identical \n",
    "                # or matches at beginning of txipt but differs at the end\n",
    "                if np.all(vals) or np.array_equal(vals, [True,False]):\n",
    "                \n",
    "                    # Don't want to throw away all valid matches (yet)\n",
    "                    # Possible genuine match, should update dict with info\n",
    "                    # match (from start of ref Txipt)\n",
    "                \n",
    "                    if vals.size == 1:\n",
    "                        #i.e. all true/introns match (e.g. bleedthrough)\n",
    "                        terminal_non_match = 0\n",
    "                    else:\n",
    "                        # All valid = consective match & non-match, so non-match = 2nd in array \n",
    "                        terminal_non_match = runs[1]\n",
    "                \n",
    "#                 print(terminal_non_match)\n",
    "                \n",
    "                    if novel_id not in match_info_dict:\n",
    "                    \n",
    "                        match_info_dict[novel_id] = {\"matches\": [ref_id],\n",
    "                                                     \"chain_match\": [runs[0]], #Always starts with true, so take length of true\n",
    "                                                     \"terminal_non_match\": [terminal_non_match],\n",
    "                                                    }\n",
    "                \n",
    "                    else:\n",
    "                        # Append to dict\n",
    "                        match_info_dict[novel_id][\"matches\"].append(ref_id)\n",
    "                        match_info_dict[novel_id][\"chain_match\"].append(runs[0])\n",
    "                        match_info_dict[novel_id][\"terminal_non_match\"].append(terminal_non_match)\n",
    "                    \n",
    "                \n",
    "                else:\n",
    "                    continue\n",
    "    \n",
    "    # Output df for easier parsing\n",
    "#     match_df = pd.concat({key: pd.DataFrame.from_dict(d, orient = \"columns\") for key, d in match_info_dict.items()}, axis=0)\n",
    "    \n",
    "    return match_info_dict\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ca48baec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----processing chrom and strand pair 1 & +\n",
      "nov_tx_bl_f does not match any reference transcripts in its first intron. Skipping\n",
      "nov_tx_fi_f does not match any reference transcripts in its first intron. Skipping\n",
      "nov_tx_fi_p does not match any reference transcripts in its first intron. Skipping\n",
      "nov_tx_si_f does not match any reference transcripts in its first intron. Skipping\n",
      "----processing chrom and strand pair 2 & -\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'nov_tx_3ui_p': {'matches': ['ref_tr_1'],\n",
       "  'chain_match': [3],\n",
       "  'terminal_non_match': [0]},\n",
       " 'nov_tx_bl_p': {'matches': ['ref_tr_1'],\n",
       "  'chain_match': [2],\n",
       "  'terminal_non_match': [0]},\n",
       " 'nov_tx_exc_dist_p': {'matches': ['ref_tr_1'],\n",
       "  'chain_match': [2],\n",
       "  'terminal_non_match': [1]},\n",
       " 'nov_tx_le_dist_p': {'matches': ['ref_tr_1'],\n",
       "  'chain_match': [3],\n",
       "  'terminal_non_match': [0]},\n",
       " 'nov_tx_mult_p': {'matches': ['ref_tr_1'],\n",
       "  'chain_match': [1],\n",
       "  'terminal_non_match': [2]},\n",
       " 'nov_tx_si_p': {'matches': ['ref_tr_1'],\n",
       "  'chain_match': [1],\n",
       "  'terminal_non_match': [1]},\n",
       " 'nov_tx_le_dist_p_minus': {'matches': ['ref_tr_2'],\n",
       "  'chain_match': [1],\n",
       "  'terminal_non_match': [1]}}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "match_intron_chains(test_novel_introns, test_ref_introns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "cf33eff0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.array_equal(np.array([True, False]), [True, False])\n",
    "np.all([True,True])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fb031093",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test_introns = pr.from_dict(test_novel_3ui)#.subset(lambda df: df[\"transcript_id\"] == \"nov_tx_3ui_p\")\n",
    "\n",
    "# n_exons = len(test_introns)\n",
    "# # n exons = n -1 introns\n",
    "\n",
    "# # plus strand txipt - End of first exon = start of intron, Start of 3'exon = end coord\n",
    "\n",
    "# test_introns.as_df().iloc[0,lambda x: x.columns.get_loc(\"End\")]\n",
    "\n",
    "# def introns_from_df(df):\n",
    "#     '''\n",
    "#     '''\n",
    "    \n",
    "#     n_exons = len(df)\n",
    "    \n",
    "#     if n_exons < 2:\n",
    "#         raise Exception(\"at least two exons are required for transcript to have an intron\")\n",
    "#     # n exons = n-1 introns\n",
    "    \n",
    "#     strand = df[\"Strand\"].drop_duplicates().tolist()[0]\n",
    "# #     print(strand)\n",
    "#     chrom = df[\"Chromosome\"].drop_duplicates().tolist()[0]\n",
    "#     gene_id = df[\"gene_id\"].drop_duplicates().tolist()[0]\n",
    "#     tx_id = df[\"transcript_id\"].drop_duplicates().tolist()[0]\n",
    "#     feature = \"intron\"\n",
    "#     introns = {}\n",
    "#     for i in range(0, n_exons - 1):\n",
    "#         if strand == \"+\":\n",
    "#             intron_start = df.iloc[i, lambda x: x.columns.get_loc(\"End\")]\n",
    "#             intron_end = df.iloc[i+1, lambda x: x.columns.get_loc(\"Start\")]\n",
    "#             introns[str(i)] = {\"Chromosome\": chrom,\n",
    "#                                \"Start\": intron_start,\n",
    "#                                \"End\": intron_end,\n",
    "#                                \"Strand\": strand,\n",
    "#                                \"Feature\": feature,\n",
    "#                                \"gene_id\": gene_id,\n",
    "#                                \"transcript_id\": tx_id}\n",
    "#         elif strand == \"-\":\n",
    "#             intron_end = df.iloc[i, lambda x: x.columns.get_loc(\"Start\")]\n",
    "#             intron_start = df.iloc[i+1, lambda x: x.columns.get_loc(\"End\")]\n",
    "#             introns[str(i)] = {\"Chromosome\": chrom,\n",
    "#                                \"Start\": intron_start,\n",
    "#                                \"End\": intron_end,\n",
    "#                                \"Strand\": strand,\n",
    "#                                \"Feature\": feature,\n",
    "#                                \"gene_id\": gene_id,\n",
    "#                                \"transcript_id\": tx_id}\n",
    "#     return pd.DataFrame.from_dict(introns, orient = \"index\")\n",
    "        \n",
    "\n",
    "\n",
    "# def introns_by_tx(gr, by=\"transcript_id\", nb_cpu=1):\n",
    "#     '''\n",
    "#     '''\n",
    "    \n",
    "#     return gr.apply(lambda df: df.groupby(by).apply(introns_from_df), nb_cpu=1)\n",
    "    \n",
    "    \n",
    "# test_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "758737f1",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'test_introns' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-22-e041933c69d8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mintrons_by_tx\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtest_introns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'test_introns' is not defined"
     ]
    }
   ],
   "source": [
    "introns_by_tx(test_introns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf44cbed",
   "metadata": {},
   "outputs": [],
   "source": [
    "introns_by_tx(pr.data.ensembl_gtf().subset(lambda df: (df.gene_id == \"ENSG00000205231\") & (df.Feature == \"exon\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5b71b28",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "pr.data.ensembl_gtf().subset(lambda df: df.transcript_id == \"ENST00000450305\")[[\"transcript_id\",\"Feature\"]].features.introns(by=\"transcript\").dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7e4891a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test_ref.features.introns(by=\"transcript\")\n",
    "(pr.data.ensembl_gtf().subset(lambda df: df.transcript_id == \"ENST00000450305\")\n",
    "#  [[\"gene_id\",\"transcript_id\"]]\n",
    " .features.introns(by=\"transcript\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c86eba9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pr.data.ensembl_gtf().subset(lambda df: (df.gene_id == \"ENSG00000205231\") & (df.Feature == \"exon\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fa672ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "pr.data.ensembl_gtf().subset(lambda df: (df.gene_id == \"ENSG00000205231\")).features.introns(by=\"transcript\")\n",
    "#Start of first feature, end of next along (if properly sorted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8578c48f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# path_stie_gtf_chr1_nov = \"../two_sample_example_output/stringtie/chr1.no_ref_id.TDP43-F_S6.assembled.gtf\"\n",
    "# path_ref_gtf = \"../data/annotation/gencode.v34.annotation.gtf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a2eb6cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# stie_chr1 = pr.read_gtf(path_stie_gtf_chr1_nov)\n",
    "# stie_chr1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c443d1a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ref = pr.read_gtf(path_ref_gtf)\n",
    "# ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f516593",
   "metadata": {},
   "outputs": [],
   "source": [
    "# stie_chr1_exons = stie_chr1.subset(lambda df: df[\"Feature\"] == \"exon\")\n",
    "# stie_chr1_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23078b88",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ref_exons = ref.subset(lambda df: df[\"Feature\"] == \"exon\")\n",
    "# ref_chr1_exons = ref_exons.subset(lambda df: df[\"Chromosome\"] == \"chr1\")\n",
    "# ref_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44b602ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "#prot coding / lncRNA\n",
    "# print(ref_chr1_exons.gene_type.value_counts())\n",
    "# ref_chr1_pc_exons = ref_chr1_exons.subset(lambda df: df.gene_type.isin([\"protein_coding\", \"lncRNA\"]))\n",
    "# ref_chr1_pc_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46be9ffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# stie_chr1_introns = introns_by_tx(stie_chr1_exons,nb_cpu=1)\n",
    "# stie_chr1_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f8edcbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ref_all_introns = introns_by_tx(ref_exons, nb_cpu=4)\n",
    "# ref_all_introns\n",
    "# ref_chr1_introns = introns_by_tx(ref_chr1_pc_exons, nb_cpu=2)\n",
    "# ref_chr1_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29d16f4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr1_chain_matching = match_intron_chains(stie_chr1_introns, ref_chr1_introns)\n",
    "# chr1_chain_matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c6c96ea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9d6e3ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ref_chr1_introns.subset(lambda df: df.transcript_id == \"ENST00000674495.1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a19a84a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ref_chr1_exons.subset(lambda df: df.transcript_id == \"ENST00000674495.1\").sort()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16bf7805",
   "metadata": {},
   "source": [
    "### Alternative (hopefully more scalable approach)\n",
    "\n",
    "Try and use pyranges internals as much as possible, and avoid manual looping and comparisons\n",
    "\n",
    "Essentially, Find overlapping introns, then filter for those that are identical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "100294cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def intron_id(gr):\n",
    "\n",
    "    return gr.assign(\"intron_id\",\n",
    "                                         lambda df: pd.Series([\":\".join([tx_id, str(start), str(end)])\n",
    "                                                     for tx_id, start, end in zip(df[\"transcript_id\"],\n",
    "                                                                                  df[\"Start\"],\n",
    "                                                                                  df[\"End\"])\n",
    "                                                    ])\n",
    "                                        )\n",
    "\n",
    "def sort_introns_by_strand(df):\n",
    "    '''\n",
    "    '''\n",
    "    # first reset_index call removes the original index of the group (e.g. row 4005 in df)\n",
    "    # second reset_index call adds the sorted index as a column to the dataframe (the order along exon in each transcript)\n",
    "    if (df.Strand == '+').all():\n",
    "        return df.sort_values(by=['End']).reset_index(drop=True).reset_index()\n",
    "    elif (df.Strand == '-').all():\n",
    "        return df.sort_values(by=['Start'], ascending=False).reset_index(drop=True).reset_index()\n",
    "\n",
    "\n",
    "def filter_transcripts_by_chain(novel_exons,ref_exons, match_type = \"transcript\",nb_cpu = 1):\n",
    "    '''\n",
    "    '''\n",
    "\n",
    "    novel_cols_to_keep = [\"Feature\",\"transcript_id\"]\n",
    "    ref_cols_to_keep = [\"Feature\", \"transcript_id\", \"gene_id\", \"gene_name\"]\n",
    "\n",
    "    assert match_type in [\"transcript\", \"any\"], \"match_type must be one of 'transcript' or 'any'. value passed - {}\".format(str(match_type))\n",
    "\n",
    "    #1. Find introns by transcript & give each intron a unique ID\n",
    "    novel_introns = introns_by_tx(novel_exons).sort()\n",
    "    ref_introns = introns_by_tx(ref_exons).sort()\n",
    "\n",
    "    novel_introns = intron_id(novel_introns)\n",
    "    ref_introns = intron_id(ref_introns)\n",
    "\n",
    "\n",
    "    #2. Track number of introns in each novel transcript\n",
    "    novel_tx_intron_counts = (novel_introns.as_df()\n",
    "                              .groupby(\"transcript_id\").size())\n",
    "\n",
    "\n",
    "    # novel_introns, ref_introns\n",
    "\n",
    "    # 3. Store intron_ids for each transcript, sorted by intron_number (where 1 = first intron regardless of strand) in a df/Series\n",
    "    novel_intron_ids_ordered = (novel_introns.as_df()\n",
    "                                .groupby(\"transcript_id\")\n",
    "                                .apply(sort_introns_by_strand)\n",
    "                                .reset_index(drop=True)\n",
    "                                .rename({\"index\": \"intron_number\"}, axis=\"columns\")\n",
    "                               )\n",
    "    novel_intron_ids_ordered[\"intron_number\"] = novel_intron_ids_ordered[\"intron_number\"].add(1)\n",
    "\n",
    "    # df of txipt_id | intron_id | intron_number\n",
    "    novel_intron_ids_ordered = novel_intron_ids_ordered.loc[:,[\"transcript_id\",\"intron_id\",\"intron_number\"]]\n",
    "#     print(novel_intron_ids_ordered.dtypes)\n",
    "\n",
    "\n",
    "    #4. Find novel introns with any overlap with reference introns\n",
    "    # Inner join to add ref_rows to novel gr\n",
    "    joined = novel_introns.join(ref_introns, strandedness=\"same\",how=None, suffix =\"_ref\",nb_cpu=nb_cpu)\n",
    "\n",
    "    #5. Filter for overlaps that exactly match (or differ by given tolerance)\n",
    "    joined = joined.subset(lambda df: abs(df.Start - df.Start_ref) + abs(df.End - df.End_ref) <= 0, nb_cpu=nb_cpu)\n",
    "\n",
    "    # Minimal info needed on matches between novel and reference introns\n",
    "    joined = joined.as_df()[[\"transcript_id\",\"intron_id\",\"transcript_id_ref\",\"intron_id_ref\"]]\n",
    "\n",
    "#     print(joined.dtypes)\n",
    "\n",
    "    #6. Join ordered novel introns with match info\n",
    "\n",
    "    if match_type == \"transcript\":\n",
    "        # Looking for intron to match any annotated intron, regardless of reference transcript\n",
    "        novel_ref_match_info = novel_intron_ids_ordered.merge(joined,\n",
    "                                                              how=\"left\",\n",
    "                                                              on=\"intron_id\",\n",
    "                                                              suffixes=[\"_novel\",\"_match\"]\n",
    "                                                             )\n",
    "\n",
    "    else:\n",
    "        # Looking for introns (except last) to match the same reference transcript\n",
    "        # merge_ordered can do a 'grouped merge' filling in empty rows (introns) for each transcript_id\n",
    "        # This is especially useful if want to do transcript-specific intron matching\n",
    "        # For each reference transcript, all novel introns will be filled with NaN if no overlap for given transcript_id\n",
    "        # (i.e. novel txipt matches all but last intron of reference transcript)\n",
    "\n",
    "        novel_ref_match_info = (pd.merge_ordered(novel_intron_ids_ordered,\n",
    "                                    joined,\n",
    "                                    how=\"left\",\n",
    "                                    on=\"intron_id\",\n",
    "                                    right_by=\"transcript_id_ref\", # group matches by ref tx & join tx by tx\n",
    "                                    suffixes=[\"_novel\",\"_match\"],\n",
    "                                    fill_method=None)\n",
    "                   .sort_values(by=[\"transcript_id_novel\",\"intron_number\"])\n",
    "#                    .dropna(axis=\"columns\", subset=[\"intron_id_ref\"]) # This fills for\n",
    "              )\n",
    "\n",
    "\n",
    "    return novel_ref_match_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "48e6da55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id_novel</th>\n",
       "      <th>intron_id</th>\n",
       "      <th>intron_number</th>\n",
       "      <th>transcript_id_match</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>intron_id_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>nov_tx_3ui_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:30:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>nov_tx_3ui_p:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:120:200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>nov_tx_3ui_p:220:300</td>\n",
       "      <td>3</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:220:300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>nov_tx_3ui_p:310:330</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "      <td>nov_tx_bl_f:70:100</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "      <td>nov_tx_bl_f:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:120:200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>nov_tx_bl_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:30:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>nov_tx_bl_p:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:120:200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>nov_tx_exc_dist_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:30:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>nov_tx_exc_dist_p:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:120:200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>nov_tx_exc_dist_p:220:360</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "      <td>nov_tx_fi_f:35:50</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "      <td>nov_tx_fi_p:30:50</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>nov_tx_le_dist_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:30:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>nov_tx_le_dist_p:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:120:200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>nov_tx_le_dist_p:220:300</td>\n",
       "      <td>3</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:220:300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>nov_tx_le_dist_p:310:360</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "      <td>nov_tx_le_dist_p_minus:90:100</td>\n",
       "      <td>1</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>ref_tr_2:90:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "      <td>nov_tx_le_dist_p_minus:50:80</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>nov_tx_mult_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:30:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>nov_tx_mult_p:120:130</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>nov_tx_mult_p:140:150</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>nov_tx_si_f</td>\n",
       "      <td>nov_tx_si_f:70:100</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>nov_tx_si_f</td>\n",
       "      <td>nov_tx_si_f:120:140</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>nov_tx_si_p</td>\n",
       "      <td>nov_tx_si_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:30:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>nov_tx_si_p</td>\n",
       "      <td>nov_tx_si_p:120:140</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id_novel                      intron_id  intron_number  \\\n",
       "0             nov_tx_3ui_p            nov_tx_3ui_p:30:100              1   \n",
       "1             nov_tx_3ui_p           nov_tx_3ui_p:120:200              2   \n",
       "2             nov_tx_3ui_p           nov_tx_3ui_p:220:300              3   \n",
       "3             nov_tx_3ui_p           nov_tx_3ui_p:310:330              4   \n",
       "4              nov_tx_bl_f             nov_tx_bl_f:70:100              1   \n",
       "5              nov_tx_bl_f            nov_tx_bl_f:120:200              2   \n",
       "6              nov_tx_bl_p             nov_tx_bl_p:30:100              1   \n",
       "7              nov_tx_bl_p            nov_tx_bl_p:120:200              2   \n",
       "8        nov_tx_exc_dist_p       nov_tx_exc_dist_p:30:100              1   \n",
       "9        nov_tx_exc_dist_p      nov_tx_exc_dist_p:120:200              2   \n",
       "10       nov_tx_exc_dist_p      nov_tx_exc_dist_p:220:360              3   \n",
       "11             nov_tx_fi_f              nov_tx_fi_f:35:50              1   \n",
       "12             nov_tx_fi_p              nov_tx_fi_p:30:50              1   \n",
       "13        nov_tx_le_dist_p        nov_tx_le_dist_p:30:100              1   \n",
       "14        nov_tx_le_dist_p       nov_tx_le_dist_p:120:200              2   \n",
       "15        nov_tx_le_dist_p       nov_tx_le_dist_p:220:300              3   \n",
       "16        nov_tx_le_dist_p       nov_tx_le_dist_p:310:360              4   \n",
       "17  nov_tx_le_dist_p_minus  nov_tx_le_dist_p_minus:90:100              1   \n",
       "18  nov_tx_le_dist_p_minus   nov_tx_le_dist_p_minus:50:80              2   \n",
       "19           nov_tx_mult_p           nov_tx_mult_p:30:100              1   \n",
       "20           nov_tx_mult_p          nov_tx_mult_p:120:130              2   \n",
       "21           nov_tx_mult_p          nov_tx_mult_p:140:150              3   \n",
       "22             nov_tx_si_f             nov_tx_si_f:70:100              1   \n",
       "23             nov_tx_si_f            nov_tx_si_f:120:140              2   \n",
       "24             nov_tx_si_p             nov_tx_si_p:30:100              1   \n",
       "25             nov_tx_si_p            nov_tx_si_p:120:140              2   \n",
       "\n",
       "       transcript_id_match transcript_id_ref     intron_id_ref  \n",
       "0             nov_tx_3ui_p          ref_tr_1   ref_tr_1:30:100  \n",
       "1             nov_tx_3ui_p          ref_tr_1  ref_tr_1:120:200  \n",
       "2             nov_tx_3ui_p          ref_tr_1  ref_tr_1:220:300  \n",
       "3                      NaN               NaN               NaN  \n",
       "4                      NaN               NaN               NaN  \n",
       "5              nov_tx_bl_f          ref_tr_1  ref_tr_1:120:200  \n",
       "6              nov_tx_bl_p          ref_tr_1   ref_tr_1:30:100  \n",
       "7              nov_tx_bl_p          ref_tr_1  ref_tr_1:120:200  \n",
       "8        nov_tx_exc_dist_p          ref_tr_1   ref_tr_1:30:100  \n",
       "9        nov_tx_exc_dist_p          ref_tr_1  ref_tr_1:120:200  \n",
       "10                     NaN               NaN               NaN  \n",
       "11                     NaN               NaN               NaN  \n",
       "12                     NaN               NaN               NaN  \n",
       "13        nov_tx_le_dist_p          ref_tr_1   ref_tr_1:30:100  \n",
       "14        nov_tx_le_dist_p          ref_tr_1  ref_tr_1:120:200  \n",
       "15        nov_tx_le_dist_p          ref_tr_1  ref_tr_1:220:300  \n",
       "16                     NaN               NaN               NaN  \n",
       "17  nov_tx_le_dist_p_minus          ref_tr_2   ref_tr_2:90:100  \n",
       "18                     NaN               NaN               NaN  \n",
       "19           nov_tx_mult_p          ref_tr_1   ref_tr_1:30:100  \n",
       "20                     NaN               NaN               NaN  \n",
       "21                     NaN               NaN               NaN  \n",
       "22                     NaN               NaN               NaN  \n",
       "23                     NaN               NaN               NaN  \n",
       "24             nov_tx_si_p          ref_tr_1   ref_tr_1:30:100  \n",
       "25                     NaN               NaN               NaN  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filter_transcripts_by_chain(test_novel_gr, test_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6aec5e9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id_novel</th>\n",
       "      <th>intron_id</th>\n",
       "      <th>intron_number</th>\n",
       "      <th>transcript_id_match</th>\n",
       "      <th>transcript_id_ref</th>\n",
       "      <th>intron_id_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>nov_tx_3ui_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:30:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>nov_tx_3ui_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>nov_tx_3ui_p:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:120:200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>nov_tx_3ui_p:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>nov_tx_3ui_p:220:300</td>\n",
       "      <td>3</td>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:220:300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>nov_tx_3ui_p:220:300</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>nov_tx_3ui_p:310:330</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>nov_tx_3ui_p</td>\n",
       "      <td>nov_tx_3ui_p:310:330</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "      <td>nov_tx_bl_f:70:100</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "      <td>nov_tx_bl_f:70:100</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "      <td>nov_tx_bl_f:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:120:200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>nov_tx_bl_f</td>\n",
       "      <td>nov_tx_bl_f:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>nov_tx_bl_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:30:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>nov_tx_bl_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>nov_tx_bl_p:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:120:200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>nov_tx_bl_p</td>\n",
       "      <td>nov_tx_bl_p:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>nov_tx_exc_dist_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:30:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>nov_tx_exc_dist_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>nov_tx_exc_dist_p:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:120:200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>nov_tx_exc_dist_p:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>nov_tx_exc_dist_p:220:360</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>nov_tx_exc_dist_p</td>\n",
       "      <td>nov_tx_exc_dist_p:220:360</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "      <td>nov_tx_fi_f:35:50</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>nov_tx_fi_f</td>\n",
       "      <td>nov_tx_fi_f:35:50</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "      <td>nov_tx_fi_p:30:50</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>nov_tx_fi_p</td>\n",
       "      <td>nov_tx_fi_p:30:50</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>nov_tx_le_dist_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:30:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>nov_tx_le_dist_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>nov_tx_le_dist_p:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:120:200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>nov_tx_le_dist_p:120:200</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>nov_tx_le_dist_p:220:300</td>\n",
       "      <td>3</td>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:220:300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>nov_tx_le_dist_p:220:300</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>nov_tx_le_dist_p:310:360</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>nov_tx_le_dist_p</td>\n",
       "      <td>nov_tx_le_dist_p:310:360</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "      <td>nov_tx_le_dist_p_minus:90:100</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "      <td>nov_tx_le_dist_p_minus:90:100</td>\n",
       "      <td>1</td>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>ref_tr_2:90:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "      <td>nov_tx_le_dist_p_minus:50:80</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>nov_tx_le_dist_p_minus</td>\n",
       "      <td>nov_tx_le_dist_p_minus:50:80</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>nov_tx_mult_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:30:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>nov_tx_mult_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>nov_tx_mult_p:120:130</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>nov_tx_mult_p:120:130</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>nov_tx_mult_p:140:150</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>nov_tx_mult_p</td>\n",
       "      <td>nov_tx_mult_p:140:150</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>nov_tx_si_f</td>\n",
       "      <td>nov_tx_si_f:70:100</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>nov_tx_si_f</td>\n",
       "      <td>nov_tx_si_f:70:100</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>nov_tx_si_f</td>\n",
       "      <td>nov_tx_si_f:120:140</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>nov_tx_si_f</td>\n",
       "      <td>nov_tx_si_f:120:140</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>nov_tx_si_p</td>\n",
       "      <td>nov_tx_si_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>nov_tx_si_p</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>ref_tr_1:30:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>nov_tx_si_p</td>\n",
       "      <td>nov_tx_si_p:30:100</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>nov_tx_si_p</td>\n",
       "      <td>nov_tx_si_p:120:140</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>nov_tx_si_p</td>\n",
       "      <td>nov_tx_si_p:120:140</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ref_tr_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id_novel                      intron_id  intron_number  \\\n",
       "2             nov_tx_3ui_p            nov_tx_3ui_p:30:100              1   \n",
       "28            nov_tx_3ui_p            nov_tx_3ui_p:30:100              1   \n",
       "0             nov_tx_3ui_p           nov_tx_3ui_p:120:200              2   \n",
       "26            nov_tx_3ui_p           nov_tx_3ui_p:120:200              2   \n",
       "1             nov_tx_3ui_p           nov_tx_3ui_p:220:300              3   \n",
       "27            nov_tx_3ui_p           nov_tx_3ui_p:220:300              3   \n",
       "3             nov_tx_3ui_p           nov_tx_3ui_p:310:330              4   \n",
       "29            nov_tx_3ui_p           nov_tx_3ui_p:310:330              4   \n",
       "5              nov_tx_bl_f             nov_tx_bl_f:70:100              1   \n",
       "31             nov_tx_bl_f             nov_tx_bl_f:70:100              1   \n",
       "4              nov_tx_bl_f            nov_tx_bl_f:120:200              2   \n",
       "30             nov_tx_bl_f            nov_tx_bl_f:120:200              2   \n",
       "7              nov_tx_bl_p             nov_tx_bl_p:30:100              1   \n",
       "33             nov_tx_bl_p             nov_tx_bl_p:30:100              1   \n",
       "6              nov_tx_bl_p            nov_tx_bl_p:120:200              2   \n",
       "32             nov_tx_bl_p            nov_tx_bl_p:120:200              2   \n",
       "10       nov_tx_exc_dist_p       nov_tx_exc_dist_p:30:100              1   \n",
       "36       nov_tx_exc_dist_p       nov_tx_exc_dist_p:30:100              1   \n",
       "8        nov_tx_exc_dist_p      nov_tx_exc_dist_p:120:200              2   \n",
       "34       nov_tx_exc_dist_p      nov_tx_exc_dist_p:120:200              2   \n",
       "9        nov_tx_exc_dist_p      nov_tx_exc_dist_p:220:360              3   \n",
       "35       nov_tx_exc_dist_p      nov_tx_exc_dist_p:220:360              3   \n",
       "11             nov_tx_fi_f              nov_tx_fi_f:35:50              1   \n",
       "37             nov_tx_fi_f              nov_tx_fi_f:35:50              1   \n",
       "12             nov_tx_fi_p              nov_tx_fi_p:30:50              1   \n",
       "38             nov_tx_fi_p              nov_tx_fi_p:30:50              1   \n",
       "15        nov_tx_le_dist_p        nov_tx_le_dist_p:30:100              1   \n",
       "41        nov_tx_le_dist_p        nov_tx_le_dist_p:30:100              1   \n",
       "13        nov_tx_le_dist_p       nov_tx_le_dist_p:120:200              2   \n",
       "39        nov_tx_le_dist_p       nov_tx_le_dist_p:120:200              2   \n",
       "14        nov_tx_le_dist_p       nov_tx_le_dist_p:220:300              3   \n",
       "40        nov_tx_le_dist_p       nov_tx_le_dist_p:220:300              3   \n",
       "16        nov_tx_le_dist_p       nov_tx_le_dist_p:310:360              4   \n",
       "42        nov_tx_le_dist_p       nov_tx_le_dist_p:310:360              4   \n",
       "18  nov_tx_le_dist_p_minus  nov_tx_le_dist_p_minus:90:100              1   \n",
       "44  nov_tx_le_dist_p_minus  nov_tx_le_dist_p_minus:90:100              1   \n",
       "17  nov_tx_le_dist_p_minus   nov_tx_le_dist_p_minus:50:80              2   \n",
       "43  nov_tx_le_dist_p_minus   nov_tx_le_dist_p_minus:50:80              2   \n",
       "21           nov_tx_mult_p           nov_tx_mult_p:30:100              1   \n",
       "47           nov_tx_mult_p           nov_tx_mult_p:30:100              1   \n",
       "19           nov_tx_mult_p          nov_tx_mult_p:120:130              2   \n",
       "45           nov_tx_mult_p          nov_tx_mult_p:120:130              2   \n",
       "20           nov_tx_mult_p          nov_tx_mult_p:140:150              3   \n",
       "46           nov_tx_mult_p          nov_tx_mult_p:140:150              3   \n",
       "23             nov_tx_si_f             nov_tx_si_f:70:100              1   \n",
       "49             nov_tx_si_f             nov_tx_si_f:70:100              1   \n",
       "22             nov_tx_si_f            nov_tx_si_f:120:140              2   \n",
       "48             nov_tx_si_f            nov_tx_si_f:120:140              2   \n",
       "25             nov_tx_si_p             nov_tx_si_p:30:100              1   \n",
       "51             nov_tx_si_p             nov_tx_si_p:30:100              1   \n",
       "24             nov_tx_si_p            nov_tx_si_p:120:140              2   \n",
       "50             nov_tx_si_p            nov_tx_si_p:120:140              2   \n",
       "\n",
       "       transcript_id_match transcript_id_ref     intron_id_ref  \n",
       "2             nov_tx_3ui_p          ref_tr_1   ref_tr_1:30:100  \n",
       "28                     NaN          ref_tr_2               NaN  \n",
       "0             nov_tx_3ui_p          ref_tr_1  ref_tr_1:120:200  \n",
       "26                     NaN          ref_tr_2               NaN  \n",
       "1             nov_tx_3ui_p          ref_tr_1  ref_tr_1:220:300  \n",
       "27                     NaN          ref_tr_2               NaN  \n",
       "3                      NaN          ref_tr_1               NaN  \n",
       "29                     NaN          ref_tr_2               NaN  \n",
       "5                      NaN          ref_tr_1               NaN  \n",
       "31                     NaN          ref_tr_2               NaN  \n",
       "4              nov_tx_bl_f          ref_tr_1  ref_tr_1:120:200  \n",
       "30                     NaN          ref_tr_2               NaN  \n",
       "7              nov_tx_bl_p          ref_tr_1   ref_tr_1:30:100  \n",
       "33                     NaN          ref_tr_2               NaN  \n",
       "6              nov_tx_bl_p          ref_tr_1  ref_tr_1:120:200  \n",
       "32                     NaN          ref_tr_2               NaN  \n",
       "10       nov_tx_exc_dist_p          ref_tr_1   ref_tr_1:30:100  \n",
       "36                     NaN          ref_tr_2               NaN  \n",
       "8        nov_tx_exc_dist_p          ref_tr_1  ref_tr_1:120:200  \n",
       "34                     NaN          ref_tr_2               NaN  \n",
       "9                      NaN          ref_tr_1               NaN  \n",
       "35                     NaN          ref_tr_2               NaN  \n",
       "11                     NaN          ref_tr_1               NaN  \n",
       "37                     NaN          ref_tr_2               NaN  \n",
       "12                     NaN          ref_tr_1               NaN  \n",
       "38                     NaN          ref_tr_2               NaN  \n",
       "15        nov_tx_le_dist_p          ref_tr_1   ref_tr_1:30:100  \n",
       "41                     NaN          ref_tr_2               NaN  \n",
       "13        nov_tx_le_dist_p          ref_tr_1  ref_tr_1:120:200  \n",
       "39                     NaN          ref_tr_2               NaN  \n",
       "14        nov_tx_le_dist_p          ref_tr_1  ref_tr_1:220:300  \n",
       "40                     NaN          ref_tr_2               NaN  \n",
       "16                     NaN          ref_tr_1               NaN  \n",
       "42                     NaN          ref_tr_2               NaN  \n",
       "18                     NaN          ref_tr_1               NaN  \n",
       "44  nov_tx_le_dist_p_minus          ref_tr_2   ref_tr_2:90:100  \n",
       "17                     NaN          ref_tr_1               NaN  \n",
       "43                     NaN          ref_tr_2               NaN  \n",
       "21           nov_tx_mult_p          ref_tr_1   ref_tr_1:30:100  \n",
       "47                     NaN          ref_tr_2               NaN  \n",
       "19                     NaN          ref_tr_1               NaN  \n",
       "45                     NaN          ref_tr_2               NaN  \n",
       "20                     NaN          ref_tr_1               NaN  \n",
       "46                     NaN          ref_tr_2               NaN  \n",
       "23                     NaN          ref_tr_1               NaN  \n",
       "49                     NaN          ref_tr_2               NaN  \n",
       "22                     NaN          ref_tr_1               NaN  \n",
       "48                     NaN          ref_tr_2               NaN  \n",
       "25             nov_tx_si_p          ref_tr_1   ref_tr_1:30:100  \n",
       "51                     NaN          ref_tr_2               NaN  \n",
       "24                     NaN          ref_tr_1               NaN  \n",
       "50                     NaN          ref_tr_2               NaN  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filter_transcripts_by_chain(test_novel_gr, test_ref,match_type=\"any\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:bioinfo] *",
   "language": "python",
   "name": "conda-env-bioinfo-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
